// devextreme/esm/__internal/grids/new/grid_core/columns_controller/utils.js
/**
 * DevExtreme (esm/__internal/grids/new/grid_core/columns_controller/utils.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    compileGetter,
    getPathParts
} from "../../../../../core/utils/data";
import {
    captionize
} from "../../../../../core/utils/inflector";
import {
    isDefined,
    isString,
    type
} from "../../../../../core/utils/type";
import {
    getTreeNodeByPath,
    setTreeNodeByPath
} from "../../../../grids/new/grid_core/utils/tree/index";
import {
    defaultColumnProperties,
    defaultColumnPropertiesByDataType
} from "./options";
export function normalizeColumn(column, templateNormalizationFunc, columnFromDataOptions) {
    const dataType = column.dataType ?? (null === columnFromDataOptions || void 0 === columnFromDataOptions ? void 0 : columnFromDataOptions.dataType) ?? defaultColumnProperties.dataType;
    const columnDataTypeDefaultOptions = defaultColumnPropertiesByDataType[dataType];
    const columnFormat = column.format ?? (null === columnDataTypeDefaultOptions || void 0 === columnDataTypeDefaultOptions ? void 0 : columnDataTypeDefaultOptions.format) ?? (null === columnFromDataOptions || void 0 === columnFromDataOptions ? void 0 : columnFromDataOptions.format);
    const caption = captionize(column.name);
    const colWithDefaults = Object.assign({}, defaultColumnProperties, columnDataTypeDefaultOptions, {
        caption: caption
    }, column);
    const normalizedColumn = Object.assign({}, colWithDefaults, {
        dataType: dataType
    }, !!columnFormat && {
        format: columnFormat
    }, {
        calculateDisplayValue: isString(colWithDefaults.calculateDisplayValue) ? compileGetter(colWithDefaults.calculateDisplayValue) : colWithDefaults.calculateDisplayValue,
        headerItemTemplate: null === templateNormalizationFunc || void 0 === templateNormalizationFunc ? void 0 : templateNormalizationFunc(colWithDefaults.headerItemTemplate),
        fieldTemplate: null === templateNormalizationFunc || void 0 === templateNormalizationFunc ? void 0 : templateNormalizationFunc(colWithDefaults.fieldTemplate),
        fieldCaptionTemplate: null === templateNormalizationFunc || void 0 === templateNormalizationFunc ? void 0 : templateNormalizationFunc(colWithDefaults.fieldCaptionTemplate),
        fieldValueTemplate: null === templateNormalizationFunc || void 0 === templateNormalizationFunc ? void 0 : templateNormalizationFunc(colWithDefaults.fieldValueTemplate),
        calculateCellValue: colWithDefaults.calculateFieldValue,
        allowFiltering: colWithDefaults.allowFiltering ?? !!colWithDefaults.dataField,
        allowHeaderFiltering: colWithDefaults.allowHeaderFiltering ?? colWithDefaults.allowFiltering ?? !!colWithDefaults.dataField,
        allowSearch: colWithDefaults.allowSearch ?? colWithDefaults.allowFiltering ?? !!colWithDefaults.dataField,
        allowSorting: colWithDefaults.allowSorting ?? !!colWithDefaults.dataField
    });
    normalizedColumn.selector ?? (normalizedColumn.selector = data => normalizedColumn.calculateFieldValue(data));
    return normalizedColumn
}
export function getVisibleIndexes(indexes) {
    const newIndexes = [...indexes];
    let minNonExistingIndex = 0;
    indexes.forEach(((visibleIndex, index) => {
        while (newIndexes.includes(minNonExistingIndex)) {
            minNonExistingIndex += 1
        }
        newIndexes[index] = visibleIndex ?? minNonExistingIndex
    }));
    return newIndexes
}
export function normalizeVisibleIndexes(indexes, forceIndex) {
    const indexMap = indexes.map(((visibleIndex, index) => [index, visibleIndex]));
    const sortedIndexMap = new Array(indexes.length);
    if (isDefined(forceIndex)) {
        sortedIndexMap[indexes[forceIndex]] = forceIndex
    }
    let j = 0;
    indexMap.sort(((a, b) => a[1] - b[1])).forEach((_ref => {
        let [index] = _ref;
        if (index === forceIndex) {
            return
        }
        if (isDefined(sortedIndexMap[j])) {
            j += 1
        }
        sortedIndexMap[j] = index;
        j += 1
    }));
    const returnIndexes = new Array(indexes.length);
    sortedIndexMap.forEach(((index, visibleIndex) => {
        returnIndexes[index] = visibleIndex
    }));
    return returnIndexes
}
export function normalizeColumnsVisibleIndexes(columns, forceIndex) {
    const result = [...columns];
    const visibleIndexes = normalizeVisibleIndexes(columns.map((c => c.visibleIndex)), forceIndex);
    visibleIndexes.forEach(((visibleIndex, i) => {
        result[i].visibleIndex = visibleIndex
    }));
    return result
}
export function normalizeColumns(columns, templateNormalizationFunc, columnsFromData) {
    return columns.map((col => {
        const columnFromDataOptions = null === columnsFromData || void 0 === columnsFromData ? void 0 : columnsFromData[col.name];
        return normalizeColumn(col, templateNormalizationFunc, columnFromDataOptions)
    }))
}
export function preNormalizeColumns(columns) {
    const normalizedColumns = null === columns || void 0 === columns ? void 0 : columns.map((column => {
        if ("string" === typeof column) {
            return {
                dataField: column
            }
        }
        return column
    })).map(((column, index) => Object.assign({}, column, {
        name: column.name ?? column.dataField ?? `column-${index}`
    })));
    const visibleIndexes = getVisibleIndexes(null === normalizedColumns || void 0 === normalizedColumns ? void 0 : normalizedColumns.map((c => c.visibleIndex)));
    null === normalizedColumns || void 0 === normalizedColumns || normalizedColumns.forEach(((_, i) => {
        normalizedColumns[i].visibleIndex = visibleIndexes[i]
    }));
    return normalizedColumns
}
export function normalizeStringColumn(column) {
    if ("string" === typeof column) {
        return {
            dataField: column
        }
    }
    return column
}
export function getColumnIndexByName(columns, name) {
    return columns.findIndex((c => c.name === name))
}
export function getColumnByIndexOrName(columns, columnNameOrIndex) {
    const column = columns.find(((c, i) => {
        if (isString(columnNameOrIndex)) {
            return c.name === columnNameOrIndex
        }
        return i === columnNameOrIndex
    }));
    return column
}
export const getValueDataType = value => {
    const dataType = type(value);
    const isUnknownDataType = "string" !== dataType && "boolean" !== dataType && "number" !== dataType && "date" !== dataType && "object" !== dataType;
    return isUnknownDataType ? void 0 : dataType
};
export const getColumnFormat = column => {
    if (column.format) {
        return column.format
    }
    if ("date" === column.dataType || "datetime" === column.dataType) {
        return "shortDate"
    }
    return
};
export const getColumnOptionsFromDataItem = dataItem => {
    const dataFields = Object.keys(dataItem);
    return {
        dataFields: dataFields,
        columns: Object.entries(dataItem).reduce(((result, _ref2) => {
            let [key, value] = _ref2;
            const dataType = getValueDataType(value);
            const format = getColumnFormat({
                dataType: dataType
            });
            result[key] = {
                dataType: dataType,
                format: format
            };
            return result
        }), {})
    }
};
export const columnOptionUpdate = (settings, columnIdx, updatePath, value) => {
    const newSettings = [...settings];
    const updatePathParts = getPathParts(updatePath);
    const columnTreeNode = getTreeNodeByPath(newSettings[columnIdx], updatePathParts);
    if (columnTreeNode === value) {
        return settings
    }
    newSettings[columnIdx] = setTreeNodeByPath(settings[columnIdx], value, updatePathParts);
    return normalizeColumnsVisibleIndexes(newSettings, columnIdx)
};
export function addDataFieldToComputedColumns(columns) {
    return columns.map((column => {
        if (column.dataField) {
            return column
        }
        return Object.assign({}, column, {
            dataField: column.name
        })
    }))
}
