// devextreme/cjs/localization.js
/**
 * DevExtreme (cjs/localization.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "formatDate", {
    enumerable: true,
    get: function() {
        return _localization.formatDate
    }
});
Object.defineProperty(exports, "formatMessage", {
    enumerable: true,
    get: function() {
        return _localization.formatMessage
    }
});
Object.defineProperty(exports, "formatNumber", {
    enumerable: true,
    get: function() {
        return _localization.formatNumber
    }
});
Object.defineProperty(exports, "loadMessages", {
    enumerable: true,
    get: function() {
        return _localization.loadMessages
    }
});
Object.defineProperty(exports, "locale", {
    enumerable: true,
    get: function() {
        return _localization.locale
    }
});
Object.defineProperty(exports, "parseDate", {
    enumerable: true,
    get: function() {
        return _localization.parseDate
    }
});
Object.defineProperty(exports, "parseNumber", {
    enumerable: true,
    get: function() {
        return _localization.parseNumber
    }
});
var _localization = require("./__internal/core/localization");
