// devextreme/cjs/__internal/scheduler/r1/components/base/date_header_text.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/components/base/date_header_text.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.DateHeaderText = void 0;
var _inferno = require("inferno");
var _index = require("../../../../core/r1/runtime/inferno/index");
const DateHeaderTextDefaultProps = {
    text: "",
    splitText: false
};
class DateHeaderText extends _index.BaseInfernoComponent {
    constructor() {
        super(...arguments);
        this._textCache = null
    }
    getTextParts() {
        if (null !== this._textCache) {
            return this._textCache
        }
        const {
            text: text
        } = this.props;
        this._textCache = text ? text.split(" ") : [""];
        return this._textCache
    }
    componentWillUpdate(nextProps) {
        if (this.props.text !== nextProps.text) {
            this._textCache = null
        }
    }
    render() {
        const {
            splitText: splitText,
            text: text
        } = this.props;
        const textParts = this.getTextParts();
        return (0, _inferno.createFragment)(splitText ? textParts.map((part => (0, _inferno.createVNode)(1, "div", "dx-scheduler-header-panel-cell-date", (0, _inferno.createVNode)(1, "span", null, part, 0), 2))) : text, 0)
    }
}
exports.DateHeaderText = DateHeaderText;
DateHeaderText.defaultProps = DateHeaderTextDefaultProps;
