// devextreme/cjs/__internal/scheduler/r1/components/base/group_panel_vertical_cell.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/components/base/group_panel_vertical_cell.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.GroupPanelVerticalCell = void 0;
var _inferno = require("inferno");
var _index = require("../../../../core/r1/runtime/inferno/index");
var _index2 = require("../../../../scheduler/r1/components/templates/index");
var _group_panel_props = require("./group_panel_props");
class GroupPanelVerticalCell extends _index.BaseInfernoComponent {
    render() {
        const {
            className: className,
            data: data,
            id: id,
            color: color,
            text: text,
            index: index,
            cellTemplate: cellTemplate
        } = this.props;
        return (0, _inferno.createVNode)(1, "div", `dx-scheduler-group-header ${className}`, cellTemplate ? (0, _inferno.createComponentVNode)(2, _index2.PublicTemplate, {
            template: cellTemplate,
            templateProps: {
                data: {
                    data: data,
                    id: id,
                    color: color,
                    text: text
                },
                index: index
            }
        }) : (0, _inferno.createVNode)(1, "div", "dx-scheduler-group-header-content", text, 0), 0)
    }
}
exports.GroupPanelVerticalCell = GroupPanelVerticalCell;
GroupPanelVerticalCell.defaultProps = _group_panel_props.GroupPanelCellDefaultProps;
