// devextreme/cjs/__internal/grids/grid_core/ai_column/controllers/m_ai_column_controller.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/ai_column/controllers/m_ai_column_controller.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.AIColumnController = void 0;
var _m_type = require("../../../../core/utils/m_type");
var _m_modules = require("../../m_modules");
var _m_utils = _interopRequireDefault(require("../../m_utils"));
var _utils = require("../utils");
var _m_ai_column_integration_controller = require("./m_ai_column_integration_controller");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
class AIColumnController extends _m_modules.Controller {
    getDefaultCellValue(column, cellValue) {
        var _column$ai2;
        if (void 0 === cellValue) {
            var _column$ai;
            return (null === (_column$ai = column.ai) || void 0 === _column$ai ? void 0 : _column$ai.emptyText) ?? null
        }
        return (null === (_column$ai2 = column.ai) || void 0 === _column$ai2 ? void 0 : _column$ai2.noDataText) ?? null
    }
    _endCustomLoadingIfNoPendingRequests() {
        if (!this.aiColumnIntegrationController.isAnyRequestAwaitingCompletion()) {
            this.dataController.endCustomLoading()
        }
    }
    addAICommandColumn() {
        const that = this;
        const {
            dataController: dataController,
            aiColumnIntegrationController: aiColumnIntegrationController
        } = this;
        this.columnsController.addCommandColumn(Object.assign({}, (0, _utils.getAICommandColumnDefaultOptions)(), {
            calculateCellValue(data) {
                const key = dataController.keyOf(data);
                const cellValue = aiColumnIntegrationController.getAIColumnText(this.name, key);
                const defaultValue = that.getDefaultCellValue(this, cellValue);
                return cellValue || defaultValue
            }
        }))
    }
    subscribeToDataSourceChanged() {
        var _this$dataController$;
        this.dataSourceChangedHandler = this.handleDataSourceChanged.bind(this);
        null === (_this$dataController$ = this.dataController.dataSource()) || void 0 === _this$dataController$ || _this$dataController$.changed.add(this.dataSourceChangedHandler)
    }
    unsubscribeFromDataControllerChanged() {
        if (!this.dataControllerChangedHandler) {
            return
        }
        this.dataController.changed.remove(this.dataControllerChangedHandler)
    }
    subscribeToDataControllerChanged() {
        if (!this.getAIColumns().length || !_m_utils.default.isVirtualRowRendering(this)) {
            return
        }
        this.dataControllerChangedHandler = this.dataControllerChangedHandler ?? this.handleDataControllerChanged.bind(this);
        this.dataController.changed.add(this.dataControllerChangedHandler)
    }
    handleDataControllerChanged() {
        if (this.dataController.isViewportChanging()) {
            this.sendRequests()
        }
    }
    unsubscribeFromStoreEvents() {
        const store = this.dataController.store();
        if (this.storeUpdatedHandler) {
            null === store || void 0 === store || store.off("updated", this.storeUpdatedHandler)
        }
        if (this.storeRemovedHandler) {
            null === store || void 0 === store || store.off("removed", this.storeRemovedHandler)
        }
        if (this.storeBeforePushHandler) {
            null === store || void 0 === store || store.off("beforePush", this.storeBeforePushHandler)
        }
    }
    subscribeToStoreEvents() {
        const store = this.dataController.store();
        if (!store) {
            return
        }
        this.storeUpdatedHandler = this.storeUpdatedHandler ?? this.handleStoreUpdated.bind(this);
        this.storeRemovedHandler = this.storeRemovedHandler ?? this.handleStoreRemoved.bind(this);
        this.storeBeforePushHandler = this.storeBeforePushHandler ?? this.handleStoreBeforePush.bind(this);
        store.on("updated", this.storeUpdatedHandler);
        store.on("removed", this.storeRemovedHandler);
        store.on("beforePush", this.storeBeforePushHandler)
    }
    handleStoreUpdated(key) {
        this.clearAIColumnsByKey(key)
    }
    handleStoreRemoved(key) {
        this.clearAIColumnsByKey(key)
    }
    handleStoreBeforePush(_ref) {
        let {
            changes: changes
        } = _ref;
        changes.forEach((_ref2 => {
            let {
                key: key
            } = _ref2;
            if ((0, _m_type.isDefined)(key)) {
                this.clearAIColumnsByKey(key)
            }
        }))
    }
    updateAICells() {
        this.dataController.updateItems({
            repaintChangesOnly: this.option("repaintChangesOnly")
        })
    }
    checkStoreKey() {
        const store = this.dataController.store();
        if (store && !store.key()) {
            this.dataController.fireError("E1042", "AI Column");
            return false
        }
        return true
    }
    clearAIColumnsByKey(key) {
        const aiColumns = this.getAIColumns();
        aiColumns.forEach((col => {
            this.aiColumnIntegrationController.clearAIColumnByKey(col.name, key)
        }))
    }
    sendRequests() {
        const aiColumns = this.getAIColumns();
        if (!aiColumns.length || !this.checkStoreKey()) {
            return
        }
        for (const col of aiColumns) {
            if ((0, _utils.isAIColumnAutoMode)(col)) {
                this.sendRequest(col.name, true)
            }
        }
    }
    handleDataSourceChanged(args) {
        if ("loadError" === (null === args || void 0 === args ? void 0 : args.changeType)) {
            return
        }
        this.sendRequests()
    }
    callbackNames() {
        return ["aiRequestCompleted", "aiRequestRejected"]
    }
    init() {
        this.columnsController = this.getController("columns");
        this.dataController = this.getController("data");
        this.aiColumnIntegrationController = new _m_ai_column_integration_controller.AIColumnIntegrationController(this.component);
        this.aiColumnIntegrationController.init();
        this.aiColumnOptionChangedHandler = this.aiColumnOptionChanged.bind(this);
        this.columnsController.aiColumnOptionChanged.add(this.aiColumnOptionChangedHandler);
        this.subscribeToDataSourceChanged();
        this.unsubscribeFromStoreEvents();
        this.subscribeToStoreEvents();
        this.unsubscribeFromDataControllerChanged();
        this.subscribeToDataControllerChanged();
        this.addAICommandColumn()
    }
    getAIColumns() {
        return this.columnsController.getColumns().filter((col => "ai" === col.type))
    }
    publicMethods() {
        return ["abortAIColumnRequest", "sendAIColumnRequest", "refreshAIColumn", "clearAIColumn", "getAIColumnText"]
    }
    abortAIColumnRequest(columnName) {
        this.aiColumnIntegrationController.abortRequest(columnName);
        this._endCustomLoadingIfNoPendingRequests()
    }
    sendRequest(columnName, useCache) {
        let needToShowLoadPanel = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : true;
        if (!this.checkStoreKey()) {
            return
        }
        const callbacks = this.getRequestCallbacks();
        this.aiColumnIntegrationController.sendRequestCore({
            columnName: columnName,
            useCache: useCache,
            needToShowLoadPanel: needToShowLoadPanel,
            callbacks: callbacks
        })
    }
    sendAIColumnRequest(columnName) {
        this.sendRequest(columnName, false)
    }
    refreshAIColumn(columnName) {
        this.sendRequest(columnName, false)
    }
    getRequestCallbacks() {
        return {
            onRequestSending: needToShowLoadPanel => {
                if (needToShowLoadPanel) {
                    this.dataController.beginCustomLoading()
                }
            },
            onComplete: data => {
                this._endCustomLoadingIfNoPendingRequests();
                this.aiRequestCompleted.fire(data);
                this.updateAICells()
            },
            onError: error => {
                this._endCustomLoadingIfNoPendingRequests();
                this.aiRequestRejected.fire(error)
            },
            onRequestCanceled: () => {
                this._endCustomLoadingIfNoPendingRequests()
            }
        }
    }
    clearAIColumn(columnName) {
        this.abortAIColumnRequest(columnName);
        this.aiColumnIntegrationController.clearAIColumn(columnName);
        this.columnsController.columnOption(columnName, "ai.prompt", "");
        this.updateAICells()
    }
    getAIColumnText(columnName, key) {
        return this.aiColumnIntegrationController.getAIColumnText(columnName, key)
    }
    aiColumnOptionChanged(column, optionName, value) {
        const isPromptOptionName = (0, _utils.isPromptOption)(optionName, value);
        if (isPromptOptionName && column.name) {
            var _column$ai3;
            this.aiColumnIntegrationController.clearAIColumn(column.name);
            if (!(null !== (_column$ai3 = column.ai) && void 0 !== _column$ai3 && _column$ai3.prompt)) {
                this.updateAICells()
            }
        }
    }
    dispose() {
        var _this$dataController$2;
        super.dispose();
        null === (_this$dataController$2 = this.dataController.dataSource()) || void 0 === _this$dataController$2 || _this$dataController$2.changed.remove(this.dataSourceChangedHandler);
        this.unsubscribeFromStoreEvents();
        this.unsubscribeFromDataControllerChanged()
    }
}
exports.AIColumnController = AIColumnController;
