// devextreme/cjs/__internal/ui/button_group.js
/**
 * DevExtreme (cjs/__internal/ui/button_group.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = exports.BUTTON_GROUP_CLASS = void 0;
var _component_registrator = _interopRequireDefault(require("../../core/component_registrator"));
var _renderer = _interopRequireDefault(require("../../core/renderer"));
var _bindable_template = require("../../core/templates/bindable_template");
var _extend = require("../../core/utils/extend");
var _type = require("../../core/utils/type");
var _button = _interopRequireDefault(require("../../ui/button"));
var _uiCollection_widget = _interopRequireDefault(require("../../ui/collection/ui.collection_widget.edit"));
var _widget = _interopRequireDefault(require("../core/widget/widget"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const BUTTON_GROUP_CLASS = exports.BUTTON_GROUP_CLASS = "dx-buttongroup";
const BUTTON_GROUP_WRAPPER_CLASS = `${BUTTON_GROUP_CLASS}-wrapper`;
const BUTTON_GROUP_ITEM_CLASS = `${BUTTON_GROUP_CLASS}-item`;
const BUTTON_GROUP_FIRST_ITEM_CLASS = `${BUTTON_GROUP_CLASS}-first-item`;
const BUTTON_GROUP_LAST_ITEM_CLASS = `${BUTTON_GROUP_CLASS}-last-item`;
const BUTTON_GROUP_ITEM_HAS_WIDTH = `${BUTTON_GROUP_ITEM_CLASS}-has-width`;
const SHAPE_STANDARD_CLASS = "dx-shape-standard";
const BUTTON_GROUP_STYLING_MODE_CLASS = {
    contained: "dx-buttongroup-mode-contained",
    outlined: "dx-buttongroup-mode-outlined",
    text: "dx-buttongroup-mode-text"
};
class ButtonCollection extends _uiCollection_widget.default {
    _initTemplates() {
        super._initTemplates();
        this._templateManager.addDefaultTemplates({
            item: new _bindable_template.BindableTemplate((($container, data, model) => {
                this._prepareItemStyles($container);
                const {
                    buttonTemplate: buttonTemplate
                } = this.option();
                this._createComponent($container, _button.default, (0, _extend.extend)({}, model, data, this._getBasicButtonOptions(), {
                    _templateData: this._hasCustomTemplate(buttonTemplate) ? model : {},
                    template: model.template || buttonTemplate
                }))
            }), ["text", "type", "icon", "disabled", "visible", "hint"], this.option("integrationOptions.watchMethod"))
        })
    }
    _getBasicButtonOptions() {
        const {
            hoverStateEnabled: hoverStateEnabled,
            activeStateEnabled: activeStateEnabled,
            stylingMode: stylingMode
        } = this.option();
        return {
            focusStateEnabled: false,
            onClick: null,
            hoverStateEnabled: hoverStateEnabled,
            activeStateEnabled: activeStateEnabled,
            stylingMode: stylingMode
        }
    }
    _getDefaultOptions() {
        return Object.assign({}, super._getDefaultOptions(), {
            itemTemplateProperty: null
        })
    }
    _hasCustomTemplate(template) {
        return (0, _type.isFunction)(template) || this.option("integrationOptions.templates")[template]
    }
    _selectedItemClass() {
        return "dx-item-selected dx-state-selected"
    }
    _prepareItemStyles($item) {
        const itemIndex = $item.data("dxItemIndex");
        if (0 === itemIndex) {
            $item.addClass(BUTTON_GROUP_FIRST_ITEM_CLASS)
        }
        const {
            items: items
        } = this.option();
        if (items && itemIndex === items.length - 1) {
            $item.addClass(BUTTON_GROUP_LAST_ITEM_CLASS)
        }
        $item.addClass("dx-shape-standard")
    }
    _renderItemContent(args) {
        args.container = (0, _renderer.default)(args.container).parent();
        return super._renderItemContent(args)
    }
    _setAriaSelectionAttribute($target, value) {
        this.setAria("pressed", value, $target)
    }
    _renderItemContentByNode(args, $node) {
        args.container = (0, _renderer.default)(args.container).children().first();
        return super._renderItemContentByNode(args, $node)
    }
    _focusTarget() {
        return this.$element().parent()
    }
    _keyboardEventBindingTarget() {
        return this._focusTarget()
    }
    _enterKeyHandler(e) {
        e.preventDefault();
        super._enterKeyHandler(e)
    }
    _refreshContent() {
        this._prepareContent();
        this._renderContent()
    }
    _itemClass() {
        return BUTTON_GROUP_ITEM_CLASS
    }
    _itemSelectHandler(e) {
        const {
            selectionMode: selectionMode
        } = this.option();
        if ("single" === selectionMode && this.isItemSelected(e.currentTarget)) {
            return
        }
        super._itemSelectHandler(e)
    }
}
class ButtonGroup extends _widget.default {
    _getDefaultOptions() {
        return Object.assign({}, super._getDefaultOptions(), {
            hoverStateEnabled: true,
            focusStateEnabled: true,
            selectionMode: "single",
            selectedItems: [],
            selectedItemKeys: [],
            stylingMode: "contained",
            keyExpr: "text",
            items: [],
            buttonTemplate: "content",
            onSelectionChanged: null,
            onItemClick: null
        })
    }
    _init() {
        super._init();
        this._createItemClickAction()
    }
    _createItemClickAction() {
        this._itemClickAction = this._createActionByOption("onItemClick")
    }
    _initMarkup() {
        this.setAria("role", "group");
        this.$element().addClass(BUTTON_GROUP_CLASS);
        this._renderStylingMode();
        this._renderButtons();
        this._syncSelectionOptions();
        super._initMarkup()
    }
    _renderStylingMode() {
        const {
            stylingMode: stylingMode
        } = this.option();
        for (const key in BUTTON_GROUP_STYLING_MODE_CLASS) {
            this.$element().removeClass(BUTTON_GROUP_STYLING_MODE_CLASS[key])
        }
        this.$element().addClass(BUTTON_GROUP_STYLING_MODE_CLASS[stylingMode ?? "contained"])
    }
    _fireSelectionChangeEvent(addedItems, removedItems) {
        this._createActionByOption("onSelectionChanged", {
            excludeValidators: ["disabled", "readOnly"]
        })({
            addedItems: addedItems,
            removedItems: removedItems
        })
    }
    _renderButtons() {
        const $buttons = (0, _renderer.default)("<div>").addClass(BUTTON_GROUP_WRAPPER_CLASS).appendTo(this.$element());
        const {
            selectedItems: selectedItems,
            selectionMode: selectionMode,
            items: items,
            keyExpr: keyExpr,
            buttonTemplate: buttonTemplate,
            selectedItemKeys: selectedItemKeys,
            focusStateEnabled: focusStateEnabled,
            hoverStateEnabled: hoverStateEnabled,
            activeStateEnabled: activeStateEnabled,
            stylingMode: stylingMode,
            accessKey: accessKey,
            tabIndex: tabIndex
        } = this.option();
        const options = {
            selectionMode: selectionMode,
            items: items,
            keyExpr: keyExpr,
            buttonTemplate: buttonTemplate,
            selectedItemKeys: selectedItemKeys,
            focusStateEnabled: focusStateEnabled,
            hoverStateEnabled: hoverStateEnabled,
            activeStateEnabled: activeStateEnabled,
            stylingMode: stylingMode,
            accessKey: accessKey,
            tabIndex: tabIndex,
            noDataText: "",
            selectionRequired: false,
            onItemRendered: e => {
                const {
                    width: width
                } = this.option();
                if ((0, _type.isDefined)(width)) {
                    (0, _renderer.default)(e.itemElement).addClass(BUTTON_GROUP_ITEM_HAS_WIDTH)
                }
            },
            onSelectionChanged: e => {
                this._syncSelectionOptions();
                this._fireSelectionChangeEvent(e.addedItems, e.removedItems)
            },
            onItemClick: e => {
                this._itemClickAction(e)
            }
        };
        if ((0, _type.isDefined)(selectedItems) && selectedItems.length) {
            options.selectedItems = selectedItems
        }
        this._buttonsCollection = this._createComponent($buttons, ButtonCollection, options)
    }
    _syncSelectionOptions() {
        this._setOptionWithoutOptionChange("selectedItems", this._buttonsCollection.option("selectedItems"));
        this._setOptionWithoutOptionChange("selectedItemKeys", this._buttonsCollection.option("selectedItemKeys"))
    }
    _optionChanged(args) {
        const {
            name: name,
            value: value
        } = args;
        switch (name) {
            case "stylingMode":
            case "selectionMode":
            case "keyExpr":
            case "buttonTemplate":
            case "items":
            case "activeStateEnabled":
            case "focusStateEnabled":
            case "hoverStateEnabled":
            case "tabIndex":
                this._invalidate();
                break;
            case "selectedItemKeys":
            case "selectedItems":
                this._buttonsCollection.option(name, value);
                break;
            case "onItemClick":
                this._createItemClickAction();
                break;
            case "onSelectionChanged":
                break;
            case "width":
                super._optionChanged(args);
                this._buttonsCollection.itemElements().toggleClass(BUTTON_GROUP_ITEM_HAS_WIDTH, !!value);
                break;
            default:
                super._optionChanged(args)
        }
    }
}(0, _component_registrator.default)("dxButtonGroup", ButtonGroup);
var _default = exports.default = ButtonGroup;
