// devextreme/cjs/__internal/ui/m_tooltip.js
/**
 * DevExtreme (cjs/__internal/ui/m_tooltip.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _component_registrator = _interopRequireDefault(require("../../core/component_registrator"));
var _guid = _interopRequireDefault(require("../../core/guid"));
var _renderer = _interopRequireDefault(require("../../core/renderer"));
var _type = require("../../core/utils/type");
var _ui = _interopRequireDefault(require("../../ui/popover/ui.popover"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const TOOLTIP_CLASS = "dx-tooltip";
const TOOLTIP_WRAPPER_CLASS = "dx-tooltip-wrapper";
class Tooltip extends _ui.default {
    _getDefaultOptions() {
        return Object.assign({}, super._getDefaultOptions(), {
            toolbarItems: [],
            showCloseButton: false,
            enableBodyScroll: true,
            showTitle: false,
            title: null,
            titleTemplate: null,
            onTitleRendered: null,
            bottomTemplate: null,
            preventScrollEvents: false,
            propagateOutsideClick: true
        })
    }
    _render() {
        var _this$$wrapper;
        this.$element().addClass("dx-tooltip");
        null === (_this$$wrapper = this.$wrapper()) || void 0 === _this$$wrapper || _this$$wrapper.addClass("dx-tooltip-wrapper");
        super._render()
    }
    _renderContent() {
        super._renderContent();
        this._toggleAriaAttributes()
    }
    _toggleAriaDescription(showing) {
        const {
            target: target
        } = this.option();
        const $target = (0, _renderer.default)(target);
        const label = showing ? this._contentId : void 0;
        if (!(0, _type.isWindow)($target.get(0))) {
            this.setAria("describedby", label, $target)
        }
    }
    _toggleAriaAttributes() {
        this._contentId = `dx-${new _guid.default}`;
        this.$overlayContent().attr({
            id: this._contentId
        });
        this._toggleAriaDescription(true)
    }
}(0, _component_registrator.default)("dxTooltip", Tooltip);
var _default = exports.default = Tooltip;
