// devextreme/cjs/viz/core/export.js
/**
 * DevExtreme (cjs/viz/core/export.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "ExportMenu", {
    enumerable: true,
    get: function() {
        return _export.ExportMenu
    }
});
Object.defineProperty(exports, "combineMarkups", {
    enumerable: true,
    get: function() {
        return _export.combineMarkups
    }
});
Object.defineProperty(exports, "exportFromMarkup", {
    enumerable: true,
    get: function() {
        return _export.exportFromMarkup
    }
});
Object.defineProperty(exports, "exportWidgets", {
    enumerable: true,
    get: function() {
        return _export.exportWidgets
    }
});
Object.defineProperty(exports, "getMarkup", {
    enumerable: true,
    get: function() {
        return _export.getMarkup
    }
});
Object.defineProperty(exports, "plugin", {
    enumerable: true,
    get: function() {
        return _export.plugin
    }
});
var _export = require("../../__internal/viz/core/export");
