// devextreme/cjs/viz/core/utils.js
/**
 * DevExtreme (cjs/viz/core/utils.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "PANE_PADDING", {
    enumerable: true,
    get: function() {
        return _utils.PANE_PADDING
    }
});
Object.defineProperty(exports, "adjustVisualRange", {
    enumerable: true,
    get: function() {
        return _utils.adjustVisualRange
    }
});
Object.defineProperty(exports, "convertAngleToRendererSpace", {
    enumerable: true,
    get: function() {
        return _utils.convertAngleToRendererSpace
    }
});
Object.defineProperty(exports, "convertPolarToXY", {
    enumerable: true,
    get: function() {
        return _utils.convertPolarToXY
    }
});
Object.defineProperty(exports, "convertVisualRangeObject", {
    enumerable: true,
    get: function() {
        return _utils.convertVisualRangeObject
    }
});
Object.defineProperty(exports, "convertXYToPolar", {
    enumerable: true,
    get: function() {
        return _utils.convertXYToPolar
    }
});
Object.defineProperty(exports, "decreaseGaps", {
    enumerable: true,
    get: function() {
        return _utils.decreaseGaps
    }
});
Object.defineProperty(exports, "degreesToRadians", {
    enumerable: true,
    get: function() {
        return _utils.degreesToRadians
    }
});
Object.defineProperty(exports, "enumParser", {
    enumerable: true,
    get: function() {
        return _utils.enumParser
    }
});
Object.defineProperty(exports, "extractColor", {
    enumerable: true,
    get: function() {
        return _utils.extractColor
    }
});
Object.defineProperty(exports, "getAddFunction", {
    enumerable: true,
    get: function() {
        return _utils.getAddFunction
    }
});
Object.defineProperty(exports, "getAdjustedLog10", {
    enumerable: true,
    get: function() {
        return _utils.getAdjustedLog10
    }
});
Object.defineProperty(exports, "getAppropriateFormat", {
    enumerable: true,
    get: function() {
        return _utils.getAppropriateFormat
    }
});
Object.defineProperty(exports, "getCategoriesInfo", {
    enumerable: true,
    get: function() {
        return _utils.getCategoriesInfo
    }
});
Object.defineProperty(exports, "getCosAndSin", {
    enumerable: true,
    get: function() {
        return _utils.getCosAndSin
    }
});
Object.defineProperty(exports, "getDecimalOrder", {
    enumerable: true,
    get: function() {
        return _utils.getDecimalOrder
    }
});
Object.defineProperty(exports, "getDistance", {
    enumerable: true,
    get: function() {
        return _utils.getDistance
    }
});
Object.defineProperty(exports, "getLog", {
    enumerable: true,
    get: function() {
        return _utils.getLog
    }
});
Object.defineProperty(exports, "getLogExt", {
    enumerable: true,
    get: function() {
        return _utils.getLogExt
    }
});
Object.defineProperty(exports, "getNextDefsSvgId", {
    enumerable: true,
    get: function() {
        return _utils.getNextDefsSvgId
    }
});
Object.defineProperty(exports, "getPower", {
    enumerable: true,
    get: function() {
        return _utils.getPower
    }
});
Object.defineProperty(exports, "getVerticallyShiftedAngularCoords", {
    enumerable: true,
    get: function() {
        return _utils.getVerticallyShiftedAngularCoords
    }
});
Object.defineProperty(exports, "getVizRangeObject", {
    enumerable: true,
    get: function() {
        return _utils.getVizRangeObject
    }
});
Object.defineProperty(exports, "isRelativeHeightPane", {
    enumerable: true,
    get: function() {
        return _utils.isRelativeHeightPane
    }
});
Object.defineProperty(exports, "map", {
    enumerable: true,
    get: function() {
        return _utils.map
    }
});
Object.defineProperty(exports, "mergeMarginOptions", {
    enumerable: true,
    get: function() {
        return _utils.mergeMarginOptions
    }
});
Object.defineProperty(exports, "normalizeAngle", {
    enumerable: true,
    get: function() {
        return _utils.normalizeAngle
    }
});
Object.defineProperty(exports, "normalizeArcParams", {
    enumerable: true,
    get: function() {
        return _utils.normalizeArcParams
    }
});
Object.defineProperty(exports, "normalizeBBox", {
    enumerable: true,
    get: function() {
        return _utils.normalizeBBox
    }
});
Object.defineProperty(exports, "normalizeEnum", {
    enumerable: true,
    get: function() {
        return _utils.normalizeEnum
    }
});
Object.defineProperty(exports, "normalizePanesHeight", {
    enumerable: true,
    get: function() {
        return _utils.normalizePanesHeight
    }
});
Object.defineProperty(exports, "parseScalar", {
    enumerable: true,
    get: function() {
        return _utils.parseScalar
    }
});
Object.defineProperty(exports, "patchFontOptions", {
    enumerable: true,
    get: function() {
        return _utils.patchFontOptions
    }
});
Object.defineProperty(exports, "pointInCanvas", {
    enumerable: true,
    get: function() {
        return _utils.pointInCanvas
    }
});
Object.defineProperty(exports, "processSeriesTemplate", {
    enumerable: true,
    get: function() {
        return _utils.processSeriesTemplate
    }
});
Object.defineProperty(exports, "raiseTo", {
    enumerable: true,
    get: function() {
        return _utils.raiseTo
    }
});
Object.defineProperty(exports, "raiseToExt", {
    enumerable: true,
    get: function() {
        return _utils.raiseToExt
    }
});
Object.defineProperty(exports, "rangesAreEqual", {
    enumerable: true,
    get: function() {
        return _utils.rangesAreEqual
    }
});
Object.defineProperty(exports, "rotateBBox", {
    enumerable: true,
    get: function() {
        return _utils.rotateBBox
    }
});
Object.defineProperty(exports, "roundValue", {
    enumerable: true,
    get: function() {
        return _utils.roundValue
    }
});
Object.defineProperty(exports, "setCanvasValues", {
    enumerable: true,
    get: function() {
        return _utils.setCanvasValues
    }
});
Object.defineProperty(exports, "unique", {
    enumerable: true,
    get: function() {
        return _utils.unique
    }
});
Object.defineProperty(exports, "updatePanesCanvases", {
    enumerable: true,
    get: function() {
        return _utils.updatePanesCanvases
    }
});
Object.defineProperty(exports, "valueOf", {
    enumerable: true,
    get: function() {
        return _utils.valueOf
    }
});
var _utils = require("../../__internal/viz/core/utils");
