// devextreme/cjs/__internal/core/r1/runtime/inferno/mocked/shared.js
/**
 * DevExtreme (cjs/__internal/core/r1/runtime/inferno/mocked/shared.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ERROR_MSG = void 0;
exports.isFunction = isFunction;
exports.isInvalid = isInvalid;
exports.isNull = isNull;
exports.isNullOrUndef = isNullOrUndef;
exports.throwError = throwError;
const ERROR_MSG = exports.ERROR_MSG = "a runtime error occured! Use Inferno in development environment to find the error.";

function isNullOrUndef(o) {
    return void 0 === o || null === o
}

function isInvalid(o) {
    return null === o || false === o || true === o || void 0 === o
}

function isFunction(o) {
    return "function" === typeof o
}

function isNull(o) {
    return null === o
}

function throwError(message) {
    if (!message) {
        message = ERROR_MSG
    }
    throw new Error(`Inferno Error: ${message}`)
}
