// devextreme/esm/__internal/grids/grid_core/__tests__/__mock__/model/cell/data_cell.js
/**
 * DevExtreme (esm/__internal/grids/grid_core/__tests__/__mock__/model/cell/data_cell.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
const SELECTORS = {
    editCell: "dx-editor-cell",
    invalidCell: "invalid"
};
export class DataCellModel {
    constructor(root, addWidgetPrefix) {
        var _this$root, _this$root2;
        this.root = root;
        this.addWidgetPrefix = addWidgetPrefix;
        this.isEditCell = !!(null !== (_this$root = this.root) && void 0 !== _this$root && _this$root.classList.contains(SELECTORS.editCell));
        this.isValidCell = !(null !== (_this$root2 = this.root) && void 0 !== _this$root2 && _this$root2.classList.contains(addWidgetPrefix(SELECTORS.invalidCell)))
    }
    getElement() {
        return this.root
    }
    getText() {
        var _this$root3;
        return (null === (_this$root3 = this.root) || void 0 === _this$root3 ? void 0 : _this$root3.textContent) ?? ""
    }
    getHTML() {
        var _this$root4;
        return (null === (_this$root4 = this.root) || void 0 === _this$root4 ? void 0 : _this$root4.innerHTML) ?? ""
    }
}
