// devextreme/cjs/__internal/ui/radio_group/m_radio_button.js
/**
 * DevExtreme (cjs/__internal/ui/radio_group/m_radio_button.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _click = require("../../../common/core/events/click");
var _events_engine = _interopRequireDefault(require("../../../common/core/events/core/events_engine"));
var _utils = require("../../../common/core/events/utils");
var _component_registrator = _interopRequireDefault(require("../../../core/component_registrator"));
var _devices = _interopRequireDefault(require("../../../core/devices"));
var _renderer = _interopRequireDefault(require("../../../core/renderer"));
var _editor = _interopRequireDefault(require("../../ui/editor/editor"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const RADIO_BUTTON_CLASS = "dx-radiobutton";
const RADIO_BUTTON_ICON_CLASS = "dx-radiobutton-icon";
const RADIO_BUTTON_ICON_DOT_CLASS = "dx-radiobutton-icon-dot";
const RADIO_BUTTON_CHECKED_CLASS = "dx-radiobutton-checked";
const RADIO_BUTTON_ICON_CHECKED_CLASS = "dx-radiobutton-icon-checked";
class RadioButton extends _editor.default {
    _supportedKeys() {
        return Object.assign({}, super._supportedKeys(), {
            space: function(e) {
                e.preventDefault();
                this._clickAction({
                    event: e
                })
            }
        })
    }
    _getDefaultOptions() {
        return Object.assign({}, super._getDefaultOptions(), {
            hoverStateEnabled: true,
            activeStateEnabled: true,
            value: false
        })
    }
    _canValueBeChangedByClick() {
        return true
    }
    _defaultOptionsRules() {
        return super._defaultOptionsRules().concat([{
            device: () => "desktop" === _devices.default.real().deviceType && !_devices.default.isSimulator(),
            options: {
                focusStateEnabled: true
            }
        }])
    }
    _init() {
        super._init();
        this.$element().addClass("dx-radiobutton")
    }
    _initMarkup() {
        super._initMarkup();
        this._renderIcon();
        this._renderCheckedState(this.option("value"));
        this._renderClick();
        this.setAria("role", "radio")
    }
    _renderIcon() {
        this._$icon = (0, _renderer.default)("<div>").addClass("dx-radiobutton-icon");
        (0, _renderer.default)("<div>").addClass("dx-radiobutton-icon-dot").appendTo(this._$icon);
        this.$element().append(this._$icon)
    }
    _renderCheckedState(checked) {
        this.$element().toggleClass("dx-radiobutton-checked", checked).find(".dx-radiobutton-icon").toggleClass("dx-radiobutton-icon-checked", checked);
        this.setAria("checked", checked)
    }
    _renderClick() {
        const eventName = (0, _utils.addNamespace)(_click.name, this.NAME);
        this._clickAction = this._createAction((args => {
            this._clickHandler(args.event)
        }));
        _events_engine.default.off(this.$element(), eventName);
        _events_engine.default.on(this.$element(), eventName, (e => {
            var _this$_clickAction;
            null === (_this$_clickAction = this._clickAction) || void 0 === _this$_clickAction || _this$_clickAction.call(this, {
                event: e
            })
        }))
    }
    _clickHandler(e) {
        this._saveValueChangeEvent(e);
        this.option("value", true);
        this._saveValueChangeEvent(void 0)
    }
    _optionChanged(args) {
        const {
            name: name,
            value: value
        } = args;
        if ("value" === name) {
            this._renderCheckedState(value);
            super._optionChanged(args)
        } else {
            super._optionChanged(args)
        }
    }
}(0, _component_registrator.default)("dxRadioButton", RadioButton);
var _default = exports.default = RadioButton;
