// devextreme/cjs/__internal/scheduler/r1/components/templates/public_template.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/components/templates/public_template.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PublicTemplate = void 0;
var _inferno = require("inferno");
var _common = require("../../../../../core/utils/common");
var _index = require("../../../../core/r1/utils/index");
const PublicTemplate = _ref => {
    let {
        template: template,
        templateProps: templateProps,
        renderOptions: renderOptions
    } = _ref;
    if (void 0 === template) {
        return (0, _inferno.createFragment)()
    }
    const templateFn = (0, _index.getTemplate)(template);
    const templatePropsWithComparer = Object.assign({}, templateProps, {
        data: Object.assign({}, templateProps.data ?? {}, {
            isEqual: (null === renderOptions || void 0 === renderOptions ? void 0 : renderOptions.propsComparer) ?? _common.equalByValue
        })
    });
    return null === templateFn || void 0 === templateFn ? void 0 : templateFn(templatePropsWithComparer)
};
exports.PublicTemplate = PublicTemplate;
