// devextreme/cjs/__internal/grids/grid_core/__tests__/__mock__/model/ai_prompt_editor.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/__tests__/__mock__/model/ai_prompt_editor.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.AIPromptEditorModel = void 0;
var _popup = _interopRequireDefault(require("../../../../../../ui/popup"));
var _progress_bar = require("../../../../../ui/__tests__/__mock__/model/progress_bar");
var _text_area = require("../../../../../ui/__tests__/__mock__/model/text_area");
var _button = require("../../../../../ui/button/__tests__/__mock__/model/button");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const CLASSES = {
    wrapper: "dx-overlay-wrapper",
    toolbarItem: "dx-toolbar-item",
    stateInvisible: "dx-state-invisible",
    stateDisabled: "dx-state-disabled",
    aiPromptEditor: "dx-ai-prompt-editor",
    aiPromptEditorTextArea: "dx-ai-prompt-editor__text-area",
    aiPromptEditorRefreshButton: "dx-ai-prompt-editor__refresh-button",
    aiPromptEditorApplyButton: "dx-ai-prompt-editor__apply-button",
    aiPromptEditorStopButton: "dx-ai-prompt-editor__stop-button",
    aiPromptEditorProgressBar: "dx-ai-prompt-editor__progressbar"
};
class AIPromptEditorModel {
    constructor(root) {
        this.root = root
    }
    isVisible() {
        return null !== this.getWrapperElement()
    }
    getPopupInstance() {
        return _popup.default.getInstance(this.root)
    }
    getTextArea() {
        return new _text_area.TextAreaModel(this.getWrapperElement().querySelector(`.${CLASSES.aiPromptEditorTextArea}`))
    }
    getWrapperElement() {
        return document.body.querySelector(`.${CLASSES.aiPromptEditor}.${CLASSES.wrapper}`)
    }
    getRefreshButton() {
        return new _button.ButtonModel(this.getWrapperElement().querySelector(`.${CLASSES.aiPromptEditorRefreshButton}`))
    }
    getApplyButton() {
        return new _button.ButtonModel(this.getWrapperElement().querySelector(`.${CLASSES.aiPromptEditorApplyButton}`))
    }
    getStopButton() {
        return new _button.ButtonModel(this.getWrapperElement().querySelector(`.${CLASSES.aiPromptEditorStopButton}`))
    }
    getProgressBar() {
        return new _progress_bar.ProgressBarModel(this.getWrapperElement().querySelector(`.${CLASSES.aiPromptEditorProgressBar}`))
    }
    isApplyToolbarItemVisible() {
        const applyButton = this.getApplyButton().getElement();
        const toolbarItem = applyButton.closest(`.${CLASSES.toolbarItem}`);
        return !!toolbarItem && !toolbarItem.classList.contains(CLASSES.stateInvisible)
    }
    isStopToolbarItemVisible() {
        const stopButton = this.getStopButton().getElement();
        const toolbarItem = stopButton.closest(`.${CLASSES.toolbarItem}`);
        return !!toolbarItem && !toolbarItem.classList.contains(CLASSES.stateInvisible)
    }
}
exports.AIPromptEditorModel = AIPromptEditorModel;
