// devextreme/cjs/__internal/grids/grid_core/__tests__/__mock__/model/column_chooser.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/__tests__/__mock__/model/column_chooser.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ColumnChooserModel = void 0;
var _tree_view = require("../../../../../ui/__tests__/__mock__/model/tree_view");
const CLASSES = {
    columnChooser: "dx-datagrid-column-chooser",
    columnChooserList: "dx-datagrid-column-chooser-list",
    popupWrapper: "dx-popup-wrapper"
};
class ColumnChooserModel {
    constructor(root) {
        this.root = root
    }
    getPopupWrapper() {
        return document.body.querySelector(`.${CLASSES.popupWrapper}.${CLASSES.columnChooser}`)
    }
    getOverlay() {
        const wrapper = this.getPopupWrapper();
        return (null === wrapper || void 0 === wrapper ? void 0 : wrapper.querySelector(".dx-overlay-content")) ?? null
    }
    getTreeView() {
        const overlay = this.getOverlay();
        if (!overlay) {
            return null
        }
        const treeViewElement = overlay.querySelector(`.${CLASSES.columnChooserList}`);
        return treeViewElement ? new _tree_view.TreeViewModel(treeViewElement) : null
    }
    isVisible() {
        return null !== this.getOverlay()
    }
    searchColumn(text) {
        const treeView = this.getTreeView();
        null === treeView || void 0 === treeView || treeView.setSearchValue(text)
    }
    toggleColumn(columnText) {
        const treeView = this.getTreeView();
        const checkBox = null === treeView || void 0 === treeView ? void 0 : treeView.getCheckboxByText(columnText);
        null === checkBox || void 0 === checkBox || checkBox.toggle()
    }
}
exports.ColumnChooserModel = ColumnChooserModel;
