// devextreme/cjs/__internal/grids/grid_core/__tests__/__mock__/model/edit_form.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/__tests__/__mock__/model/edit_form.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.EditFormModel = void 0;
const SELECTORS = {
    textEditorInput: "dx-texteditor-input",
    item: "dx-item"
};
class EditFormModel {
    constructor(root) {
        this.root = root
    }
    getElement() {
        return this.root
    }
    getItem(id) {
        var _this$root;
        return (null === (_this$root = this.root) || void 0 === _this$root ? void 0 : _this$root.querySelector(`.${SELECTORS.textEditorInput}[id*=_${id}]`)) ?? null
    }
    getItems() {
        var _this$root2;
        return (null === (_this$root2 = this.root) || void 0 === _this$root2 ? void 0 : _this$root2.querySelectorAll(`.${SELECTORS.item}`)) ?? null
    }
}
exports.EditFormModel = EditFormModel;
