// devextreme/cjs/core/element_data.js
/**
 * DevExtreme (cjs/core/element_data.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "afterCleanData", {
    enumerable: true,
    get: function() {
        return _m_element_data.afterCleanData
    }
});
Object.defineProperty(exports, "beforeCleanData", {
    enumerable: true,
    get: function() {
        return _m_element_data.beforeCleanData
    }
});
Object.defineProperty(exports, "cleanData", {
    enumerable: true,
    get: function() {
        return _m_element_data.cleanData
    }
});
Object.defineProperty(exports, "cleanDataRecursive", {
    enumerable: true,
    get: function() {
        return _m_element_data.cleanDataRecursive
    }
});
Object.defineProperty(exports, "data", {
    enumerable: true,
    get: function() {
        return _m_element_data.data
    }
});
Object.defineProperty(exports, "getDataStrategy", {
    enumerable: true,
    get: function() {
        return _m_element_data.getDataStrategy
    }
});
Object.defineProperty(exports, "removeData", {
    enumerable: true,
    get: function() {
        return _m_element_data.removeData
    }
});
Object.defineProperty(exports, "setDataStrategy", {
    enumerable: true,
    get: function() {
        return _m_element_data.setDataStrategy
    }
});
Object.defineProperty(exports, "strategyChanging", {
    enumerable: true,
    get: function() {
        return _m_element_data.strategyChanging
    }
});
var _m_element_data = require("../__internal/core/m_element_data");
