// devextreme/cjs/ui/themes.js
/**
 * DevExtreme (cjs/ui/themes.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.waitWebFont = exports.waitForThemeLoad = exports.setDefaultTimeout = exports.resetTheme = exports.ready = exports.isWebFontLoaded = exports.isPendingThemeLoaded = exports.isMaterialBased = exports.isMaterial = exports.isGeneric = exports.isFluent = exports.isDark = exports.isCompact = exports.initialized = exports.init = exports.detachCssClasses = exports.default = exports.current = exports.attachCssClasses = void 0;
var _themes = _interopRequireDefault(require("../__internal/ui/themes"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const {
    setDefaultTimeout: setDefaultTimeout,
    init: init,
    initialized: initialized,
    resetTheme: resetTheme,
    ready: ready,
    waitWebFont: waitWebFont,
    isWebFontLoaded: isWebFontLoaded,
    isCompact: isCompact,
    isDark: isDark,
    isGeneric: isGeneric,
    isMaterial: isMaterial,
    isFluent: isFluent,
    isMaterialBased: isMaterialBased,
    detachCssClasses: detachCssClasses,
    attachCssClasses: attachCssClasses,
    current: current,
    waitForThemeLoad: waitForThemeLoad,
    isPendingThemeLoaded: isPendingThemeLoaded
} = _themes.default;
exports.isPendingThemeLoaded = isPendingThemeLoaded;
exports.waitForThemeLoad = waitForThemeLoad;
exports.current = current;
exports.attachCssClasses = attachCssClasses;
exports.detachCssClasses = detachCssClasses;
exports.isMaterialBased = isMaterialBased;
exports.isFluent = isFluent;
exports.isMaterial = isMaterial;
exports.isGeneric = isGeneric;
exports.isDark = isDark;
exports.isCompact = isCompact;
exports.isWebFontLoaded = isWebFontLoaded;
exports.waitWebFont = waitWebFont;
exports.ready = ready;
exports.resetTheme = resetTheme;
exports.initialized = initialized;
exports.init = init;
exports.setDefaultTimeout = setDefaultTimeout;
var _default = exports.default = _themes.default;
