// devextreme/esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_adaptivity_geometry_inside_interval.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_adaptivity_geometry_inside_interval.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    getAbstractSizeByViewOrientation,
    getRealSizeByViewOrientation
} from "./swap_by_view_orientation";
const COLLECTOR_ADAPTIVE_BOTTOM_OFFSET = 40;
export const addAdaptivityGeometryInsideInterval = (entity, _ref) => {
    let {
        cellSize: cellSize,
        collectorSize: collectorSize,
        collectorWithMarginsSize: collectorWithMarginsSize,
        viewOrientation: viewOrientation
    } = _ref;
    const cellAbstractSize = getAbstractSizeByViewOrientation(cellSize, viewOrientation);
    const topInsideCell = entity.isAllDayPanelOccupied || "vertical" === viewOrientation ? (cellSize.height - collectorWithMarginsSize.height) / 2 : cellSize.height - 40;
    const leftInsideCell = (cellSize.width - collectorWithMarginsSize.width) / 2;
    const abstractGeometry = getAbstractSizeByViewOrientation({
        top: topInsideCell,
        left: leftInsideCell,
        width: collectorSize.width,
        height: collectorSize.height
    }, viewOrientation);
    abstractGeometry.offsetX += entity.columnIndex * cellAbstractSize.sizeX;
    const geometry = getRealSizeByViewOrientation(abstractGeometry, viewOrientation);
    const items = entity.items.map((item => Object.assign({}, item, {
        width: cellSize.width,
        height: cellSize.height
    })));
    return Object.assign({}, entity, geometry, {
        items: items
    })
};
