// devextreme/esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_geometry.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_geometry.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    addAdaptivityGeometryInsideInterval
} from "./add_adaptivity_geometry_inside_interval";
import {
    addGeometryInsideInterval
} from "./add_geometry_inside_interval";
import {
    addGroupingOffset
} from "./add_grouping_offset";
const RTLSwap = (entity, _ref) => {
    let {
        panelSize: panelSize,
        isRTLEnabled: isRTLEnabled
    } = _ref;
    if (isRTLEnabled) {
        const deltaWidth = entity.items.length ? 0 : entity.width;
        entity.left = panelSize.width - entity.left - deltaWidth
    }
};
const addPanelOffset = (entity, _ref2) => {
    let {
        cellSize: cellSize,
        viewOrientation: viewOrientation,
        isTimelineView: isTimelineView
    } = _ref2;
    switch (true) {
        case "horizontal" === viewOrientation && !isTimelineView:
            entity.top += entity.rowIndex * cellSize.height;
            break;
        case "horizontal" === viewOrientation && isTimelineView:
        case "vertical" === viewOrientation:
            entity.left += entity.rowIndex * cellSize.width
    }
};
export const addGeometry = (entities, options) => entities.map((rawEntity => {
    const {
        isAdaptivityEnabled: isAdaptivityEnabled,
        maxAppointmentsPerCell: maxAppointmentsPerCell
    } = options;
    const entity = isAdaptivityEnabled && 0 === maxAppointmentsPerCell ? addAdaptivityGeometryInsideInterval(rawEntity, options) : addGeometryInsideInterval(rawEntity, options);
    addPanelOffset(entity, options);
    addGroupingOffset(entity, options);
    RTLSwap(entity, options);
    return entity
}));
