// devextreme/esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/get_appointment_geometry.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/get_appointment_geometry.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    getAppointmentX,
    getAppointmentY
} from "./get_appointment_abstract_geometry";
import {
    getAbstractSizeByViewOrientation,
    getRealSizeByViewOrientation
} from "./swap_by_view_orientation";
export const getAppointmentGeometry = (entity, _ref) => {
    let {
        collectorPosition: collectorPosition,
        cellSize: cellSize,
        collectorWithMarginsSize: collectorWithMarginsSize,
        viewOrientation: viewOrientation,
        cells: cells
    } = _ref;
    const cellAbstractSize = getAbstractSizeByViewOrientation(cellSize, viewOrientation);
    const collectorFullAbstractSize = getAbstractSizeByViewOrientation(collectorWithMarginsSize, viewOrientation);
    const abstractGeometry = Object.assign({}, getAppointmentX(entity, cellAbstractSize, cells), getAppointmentY(entity, cellAbstractSize, collectorFullAbstractSize.sizeY, collectorPosition));
    const entityGeometry = getRealSizeByViewOrientation(abstractGeometry, viewOrientation);
    return entityGeometry
};
