// devextreme/esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/get_appointment_abstract_geometry.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/get_appointment_abstract_geometry.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
const getInsideCellX = (date, _ref, cellSizeX) => {
    let {
        min: min,
        max: max
    } = _ref;
    const cellDuration = max - min;
    const startTimeDelta = date - min;
    return 0 === cellDuration ? 0 : startTimeDelta * cellSizeX / cellDuration
};
export const getAppointmentX = (entity, cellSize, cells) => {
    const startX = getInsideCellX(entity.startDateUTC, cells[entity.cellIndex], cellSize.sizeX);
    const endX = getInsideCellX(entity.endDateUTC, cells[entity.endCellIndex], cellSize.sizeX);
    const offsetX = entity.columnIndex * cellSize.sizeX + startX;
    const sizeX = (entity.endCellIndex - entity.cellIndex) * cellSize.sizeX + endX - startX;
    return {
        offsetX: offsetX,
        sizeX: sizeX
    }
};
export const getAppointmentY = (entity, cellSize, collectorSizeY, collectorPosition) => {
    if (entity.isAllDayPanelOccupied && !entity.inStackWithCollector) {
        const sizeY = 0 === entity.maxLevel ? cellSize.sizeY - collectorSizeY : (cellSize.sizeY - collectorSizeY) / entity.maxLevel;
        const offsetY = entity.level * sizeY;
        return {
            sizeY: sizeY,
            offsetY: offsetY
        }
    }
    const maxSizeY = cellSize.sizeY - collectorSizeY;
    const sizeY = 0 === entity.maxLevel ? maxSizeY : maxSizeY / entity.maxLevel;
    let offsetY = entity.level * sizeY;
    if ("start" === collectorPosition) {
        offsetY += collectorSizeY
    }
    return {
        sizeY: sizeY,
        offsetY: offsetY
    }
};
