// devextreme/esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/swap_by_view_orientation.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/swap_by_view_orientation.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
export function getAbstractSizeByViewOrientation(size, viewOrientation) {
    const abstractSize = {
        sizeY: "horizontal" === viewOrientation ? size.height : size.width,
        sizeX: "horizontal" === viewOrientation ? size.width : size.height
    };
    if (!("top" in size && "left" in size)) {
        return abstractSize
    }
    return Object.assign({}, abstractSize, {
        offsetY: "horizontal" === viewOrientation ? size.top : size.left,
        offsetX: "horizontal" === viewOrientation ? size.left : size.top
    })
}
export function getRealSizeByViewOrientation(size, viewOrientation) {
    const realSize = {
        height: "horizontal" === viewOrientation ? size.sizeY : size.sizeX,
        width: "horizontal" === viewOrientation ? size.sizeX : size.sizeY
    };
    if (!("offsetY" in size && "offsetX" in size)) {
        return realSize
    }
    return Object.assign({}, realSize, {
        top: "horizontal" === viewOrientation ? size.offsetY : size.offsetX,
        left: "horizontal" === viewOrientation ? size.offsetX : size.offsetY
    })
}
