// devextreme/cjs/__internal/ui/multi_view/multi_view.animation.js
/**
 * DevExtreme (cjs/__internal/ui/multi_view/multi_view.animation.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.animation = exports._translator = void 0;
var _animation = require("../../../common/core/animation");
var _translator2 = require("../../../common/core/animation/translator");
const _translator = exports._translator = {
    move($element, position) {
        (0, _translator2.move)($element, {
            left: position
        })
    }
};
const animation = exports.animation = {
    moveTo($element, position, duration, completeAction) {
        _animation.fx.animate($element.get(0), {
            type: "slide",
            to: {
                left: position
            },
            duration: duration,
            complete: completeAction
        })
    },
    complete($element) {
        _animation.fx.stop($element.get(0), true)
    }
};
