// devextreme/cjs/__internal/scheduler/constants.js
/**
 * DevExtreme (cjs/__internal/scheduler/constants.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.VERTICAL_VIEW_TYPES = exports.VERTICAL_GROUP_ORIENTATION = exports.LIST_ITEM_DATA_KEY = exports.LIST_ITEM_CLASS = exports.HORIZONTAL_GROUP_ORIENTATION = exports.APPOINTMENT_SETTINGS_KEY = void 0;
const LIST_ITEM_DATA_KEY = exports.LIST_ITEM_DATA_KEY = "dxListItemData";
const LIST_ITEM_CLASS = exports.LIST_ITEM_CLASS = "dx-list-item";
const APPOINTMENT_SETTINGS_KEY = exports.APPOINTMENT_SETTINGS_KEY = "dxAppointmentSettings";
const VERTICAL_GROUP_ORIENTATION = exports.VERTICAL_GROUP_ORIENTATION = "vertical";
const HORIZONTAL_GROUP_ORIENTATION = exports.HORIZONTAL_GROUP_ORIENTATION = "horizontal";
const VERTICAL_VIEW_TYPES = exports.VERTICAL_VIEW_TYPES = ["day", "week", "workWeek"];
