// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/options/get_panel_collector_options.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/options/get_panel_collector_options.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getPanelCollectorOptions = void 0;
var _get_collector_size = require("./get_collector_size");
var _get_max_level = require("./get_max_level");
const UNLIMITED_COLLECTOR_SIZES = {
    collectorSize: {
        width: 0,
        height: 0
    },
    collectorWithMarginsSize: {
        width: 0,
        height: 0
    }
};
const ALL_DAY_COLLECTOR_WIDTH_FACTOR = .75;
const MIN_LEVEL_VERTICAL_VIEW = 1;
const getPanelCollectorOptions = (schedulerStore, _ref) => {
    var _DOMMetaData$dateTabl;
    let {
        alwaysReserveSpaceForCollector: alwaysReserveSpaceForCollector,
        isTimelineView: isTimelineView,
        viewOrientation: viewOrientation,
        isAdaptivityEnabled: isAdaptivityEnabled,
        collectorCSS: collectorCSS,
        DOMMetaData: DOMMetaData,
        panelName: panelName
    } = _ref;
    const allDayPanelCellDOM = DOMMetaData.allDayPanelCellsMeta[0] || DOMMetaData.dateTableCellsMeta[0][0];
    const regularPanelCellDOM = (null === (_DOMMetaData$dateTabl = DOMMetaData.dateTableCellsMeta[1]) || void 0 === _DOMMetaData$dateTabl ? void 0 : _DOMMetaData$dateTabl[0]) || DOMMetaData.dateTableCellsMeta[0][0];
    const cellDOM = "allDayPanel" === panelName ? allDayPanelCellDOM : regularPanelCellDOM;
    const allDayPanelCellSize = {
        width: allDayPanelCellDOM.width ?? 0,
        height: allDayPanelCellDOM.height ?? 0
    };
    const cellSize = {
        width: cellDOM.width ?? 0,
        height: cellDOM.height ?? 0
    };
    const maxAppointmentsPerCell = schedulerStore.getViewOption("maxAppointmentsPerCell");
    const collectorSizes = "unlimited" === maxAppointmentsPerCell && !alwaysReserveSpaceForCollector ? UNLIMITED_COLLECTOR_SIZES : (0, _get_collector_size.getCollectorSize)(cellSize, collectorCSS, !isAdaptivityEnabled && "allDayPanel" === panelName ? .75 * cellSize.width : 0);
    const maxLevelOptions = {
        maxAppointmentsPerCell: maxAppointmentsPerCell,
        cellSize: cellSize,
        collectorSize: collectorSizes.collectorWithMarginsSize,
        viewOrientation: viewOrientation,
        isTimelineView: isTimelineView,
        isAdaptivityEnabled: isAdaptivityEnabled
    };
    const maxLevel = (0, _get_max_level.getMaxLevel)(maxLevelOptions);
    const minLevel = "vertical" === viewOrientation ? 1 : (0, _get_max_level.getMaxLevel)(Object.assign({}, maxLevelOptions, {
        maxAppointmentsPerCell: "auto"
    }));
    return {
        allDayPanelCellSize: allDayPanelCellSize,
        cellSize: cellSize,
        collectorSizes: collectorSizes,
        maxLevel: maxLevel,
        minLevel: minLevel
    }
};
exports.getPanelCollectorOptions = getPanelCollectorOptions;
