// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/options/get_month_intervals.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/options/get_month_intervals.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getMonthIntervals = void 0;
var _shift_intervals = require("../../common/shift_intervals");
var _trim_interval = require("../../common/trim_interval");
var _get_one_day_cell_intervals = require("./get_one_day_cell_intervals");
const _excluded = ["startDayHour", "endDayHour", "skippedDays"];

function _objectWithoutPropertiesLoose(r, e) {
    if (null == r) {
        return {}
    }
    var t = {};
    for (var n in r) {
        if ({}.hasOwnProperty.call(r, n)) {
            if (e.includes(n)) {
                continue
            }
            t[n] = r[n]
        }
    }
    return t
}
const MONTH_INTERVAL_DAYS_COUNT = 7;
const splitBy7Days = intervals => {
    const result = [];
    const date = new Date(intervals.min);
    while (date.getTime() < intervals.max) {
        const min = date.getTime();
        date.setUTCDate(date.getUTCDate() + 7);
        result.push({
            min: min,
            max: date.getTime()
        })
    }
    return result
};
const cropIntervalsByDayHours = (intervals, startDayHour, endDayHour) => intervals.map((item => Object.assign({}, item, {
    min: new Date(item.min).setUTCHours(startDayHour, 0, 0, 0),
    max: new Date(item.max - 1).setUTCHours(endDayHour, 0, 0, 0)
})));
const getMonthIntervals = (_ref, viewOffset, isTimeline) => {
    let {
        startDayHour: startDayHour,
        endDayHour: endDayHour,
        skippedDays: skippedDays
    } = _ref, dateInterval = _objectWithoutPropertiesLoose(_ref, _excluded);
    const trimmedInterval = (0, _trim_interval.trimInterval)(dateInterval);
    const intervals = isTimeline ? [trimmedInterval] : splitBy7Days(trimmedInterval);
    const croppedIntervals = cropIntervalsByDayHours(intervals, startDayHour, endDayHour);
    const shiftedIntervals = (0, _shift_intervals.shiftIntervals)(croppedIntervals, viewOffset);
    const cells = (0, _get_one_day_cell_intervals.getOneDayCellIntervals)({
        intervals: intervals,
        startDayHour: startDayHour,
        endDayHour: endDayHour,
        skippedDays: skippedDays
    });
    const shiftedCells = (0, _shift_intervals.shiftIntervals)(cells, viewOffset);
    return {
        cells: shiftedCells,
        dayIntervals: shiftedCells,
        intervals: shiftedIntervals
    }
};
exports.getMonthIntervals = getMonthIntervals;
