// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/options/get_collector_size.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/options/get_collector_size.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getCollectorSize = void 0;
const DEFAULT_COLLECTOR_HEIGHT = 20;
const MIN_COLLECTOR_SIZE = 20;
const getPxValue = function() {
    let value = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : "";
    let defaultValue = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : 0;
    if (!value.endsWith("px")) {
        return defaultValue
    }
    return parseInt(value, 10) || defaultValue
};
const getCollectorSize = (cellSize, collectorCSS, defaultCollectorWidth) => {
    const parsedSize = {
        height: getPxValue(collectorCSS.height, 20),
        width: getPxValue(collectorCSS.width, defaultCollectorWidth),
        marginRight: getPxValue(collectorCSS.marginRight),
        marginLeft: getPxValue(collectorCSS.marginLeft),
        marginTop: getPxValue(collectorCSS.marginTop),
        marginBottom: getPxValue(collectorCSS.marginBottom)
    };
    const marginHeight = parsedSize.marginTop + parsedSize.marginBottom;
    const marginWidth = parsedSize.marginLeft + parsedSize.marginRight;
    const height = Math.max(20, parsedSize.height);
    const width = Math.max(20, parsedSize.width || cellSize.width - marginWidth);
    return {
        collectorSize: {
            width: width,
            height: height
        },
        collectorWithMarginsSize: {
            width: width + marginWidth,
            height: height + marginHeight
        }
    }
};
exports.getCollectorSize = getCollectorSize;
