// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/options/get_min_appointment_size.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/options/get_min_appointment_size.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getMinAppointmentSize = exports.getDefaultAppointmentSize = void 0;
var _themes = require("../../../../../ui/themes");
const COMPACT_THEME_APPOINTMENT_DEFAULT_HEIGHT = 18;
const APPOINTMENT_DEFAULT_HEIGHT = 20;
const DAY_VIEW_APPOINTMENT_MIN_HEIGHT = 12;
const APPOINTMENT_DEFAULT_HORIZONTAL_WIDTH = 40;
const APPOINTMENT_DEFAULT_VERTICAL_WIDTH = 50;
const APPOINTMENT_MIN_HEIGHT = 35;
const APPOINTMENT_MIN_WIDTH = 40;
const TIMELINE_APPOINTMENT_DEFAULT_HEIGHT = 60;
const ADAPTIVE_APPOINTMENT_DEFAULT_WIDTH = 30;
const getMinAppointmentHeightByTheme = () => (0, _themes.isCompact)((0, _themes.current)()) ? 18 : 20;
const getMinAppointmentSize = options => {
    const {
        isTimelineView: isTimelineView,
        isAdaptivityEnabled: isAdaptivityEnabled,
        isMonthView: isMonthView,
        isAllDayAppointment: isAllDayAppointment
    } = options;
    if (isAdaptivityEnabled) {
        return {
            width: 30,
            height: 30
        }
    }
    if (isTimelineView) {
        return {
            width: 40,
            height: 35
        }
    }
    const height = isMonthView || isAllDayAppointment ? getMinAppointmentHeightByTheme() : 12;
    return {
        width: 40,
        height: height
    }
};
exports.getMinAppointmentSize = getMinAppointmentSize;
const getDefaultAppointmentSize = _ref => {
    let {
        isTimelineView: isTimelineView,
        isAdaptivityEnabled: isAdaptivityEnabled,
        viewOrientation: viewOrientation
    } = _ref;
    if (isAdaptivityEnabled) {
        return {
            width: 30,
            height: 30
        }
    }
    return {
        width: "vertical" === viewOrientation ? 50 : 40,
        height: isTimelineView ? 60 : getMinAppointmentHeightByTheme()
    }
};
exports.getDefaultAppointmentSize = getDefaultAppointmentSize;
