// devextreme/esm/__internal/ui/diagram/ui.diagram.toolbox.js
/**
 * DevExtreme (esm/__internal/ui/diagram/ui.diagram.toolbox.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import messageLocalization from "../../../common/core/localization/message";
import $ from "../../../core/renderer";
import {
    Deferred
} from "../../../core/utils/deferred";
import {
    extend
} from "../../../core/utils/extend";
import {
    getHeight,
    getOuterHeight,
    setHeight
} from "../../../core/utils/size";
import {
    hasWindow
} from "../../../core/utils/window";
import Accordion from "../../../ui/accordion";
import {
    getDiagram
} from "../../ui/diagram/diagram.importer";
import DiagramFloatingPanel from "../../ui/diagram/ui.diagram.floating_panel";
import Tooltip from "../../ui/m_tooltip";
import ScrollView from "../../ui/scroll_view/scroll_view";
import TextBox from "../../ui/text_box/m_text_box";
const DIAGRAM_TOOLBOX_MIN_HEIGHT = 130;
const DIAGRAM_TOOLBOX_POPUP_CLASS = "dx-diagram-toolbox-popup";
const DIAGRAM_TOOLBOX_PANEL_CLASS = "dx-diagram-toolbox-panel";
const DIAGRAM_TOOLBOX_INPUT_CONTAINER_CLASS = "dx-diagram-toolbox-input-container";
const DIAGRAM_TOOLBOX_INPUT_CLASS = "dx-diagram-toolbox-input";
const DIAGRAM_TOOLTIP_DATATOGGLE = "shape-toolbox-tooltip";
const DIAGRAM_TOOLBOX_START_DRAG_CLASS = ".dxdi-tb-start-drag-flag";
class DiagramToolbox extends DiagramFloatingPanel {
    _init() {
        super._init();
        this._toolboxes = [];
        this._filterText = "";
        this._createOnShapeCategoryRenderedAction();
        this._createOnFilterChangedAction()
    }
    _getPopupClass() {
        return "dx-diagram-toolbox-popup"
    }
    _getPopupHeight() {
        return this.isMobileView() ? "100%" : super._getPopupHeight()
    }
    _getPopupMaxHeight() {
        return this.isMobileView() ? "100%" : super._getPopupMaxHeight()
    }
    _getPopupMinHeight() {
        return 130
    }
    _getPopupPosition() {
        const {
            offsetParent: offsetParent,
            offsetX: offsetX,
            offsetY: offsetY
        } = this.option();
        const position = {
            my: "left top",
            at: "left top",
            of: offsetParent
        };
        if (!this.isMobileView()) {
            return extend(position, {
                offset: `${offsetX} ${offsetY}`
            })
        }
        return position
    }
    _getPopupAnimation() {
        const $parent = this.option("offsetParent");
        if (this.isMobileView()) {
            return {
                hide: this._getPopupSlideAnimationObject({
                    direction: "left",
                    from: {
                        position: {
                            my: "left top",
                            at: "left top",
                            of: $parent
                        }
                    },
                    to: {
                        position: {
                            my: "right top",
                            at: "left top",
                            of: $parent
                        }
                    }
                }),
                show: this._getPopupSlideAnimationObject({
                    direction: "right",
                    from: {
                        position: {
                            my: "right top",
                            at: "left top",
                            of: $parent
                        }
                    },
                    to: {
                        position: {
                            my: "left top",
                            at: "left top",
                            of: $parent
                        }
                    }
                })
            }
        }
        return super._getPopupAnimation()
    }
    _getPopupOptions() {
        const options = super._getPopupOptions();
        if (!this.isMobileView()) {
            return extend(options, {
                showTitle: true,
                toolbarItems: [{
                    widget: "dxButton",
                    location: "center",
                    options: {
                        activeStateEnabled: false,
                        focusStateEnabled: false,
                        hoverStateEnabled: false,
                        icon: "diagram-toolbox-drag",
                        stylingMode: "outlined",
                        type: "normal"
                    }
                }]
            })
        }
        return options
    }
    _renderPopupContent($parent) {
        let panelHeight = "100%";
        if (this.option("showSearch")) {
            const $inputContainer = $("<div>").addClass("dx-diagram-toolbox-input-container").appendTo($parent);
            this._updateElementWidth($inputContainer);
            this._renderSearchInput($inputContainer);
            if (hasWindow()) {
                var _this$_searchInput;
                panelHeight = `calc(100% - ${getHeight(null===(_this$_searchInput=this._searchInput)||void 0===_this$_searchInput?void 0:_this$_searchInput.$element())}px)`
            }
        }
        const $panel = $("<div>").addClass("dx-diagram-toolbox-panel").appendTo($parent);
        setHeight($panel, panelHeight);
        this._updateElementWidth($panel);
        this._renderScrollView($panel)
    }
    _updateElementWidth($element) {
        if (void 0 !== this.option("toolboxWidth")) {
            const {
                toolboxWidth: toolboxWidth
            } = this.option();
            $element.css("width", toolboxWidth)
        }
    }
    updateMaxHeight() {
        if (this.isMobileView()) {
            return
        }
        let maxHeight = 6;
        if (this._popup) {
            const $title = this._getPopupTitle();
            maxHeight += getOuterHeight($title)
        }
        if (this._accordion) {
            maxHeight += getOuterHeight(this._accordion.$element())
        }
        if (this._searchInput) {
            maxHeight += getOuterHeight(this._searchInput.$element())
        }
        this.option("maxHeight", maxHeight)
    }
    _renderSearchInput($parent) {
        const $input = $("<div>").addClass("dx-diagram-toolbox-input").appendTo($parent);
        this._searchInput = this._createComponent($input, TextBox, {
            stylingMode: "outlined",
            placeholder: messageLocalization.format("dxDiagram-uiSearch"),
            onValueChanged: data => {
                this._onInputChanged(data.value)
            },
            valueChangeEvent: "keyup",
            buttons: [{
                name: "search",
                location: "after",
                options: {
                    activeStateEnabled: false,
                    focusStateEnabled: false,
                    hoverStateEnabled: false,
                    icon: "search",
                    stylingMode: "outlined",
                    type: "normal",
                    onClick: () => {
                        var _this$_searchInput2;
                        null === (_this$_searchInput2 = this._searchInput) || void 0 === _this$_searchInput2 || _this$_searchInput2.focus()
                    }
                }
            }]
        })
    }
    _renderScrollView($parent) {
        const $scrollViewWrapper = $("<div>").appendTo($parent);
        this._scrollView = this._createComponent($scrollViewWrapper, ScrollView);
        const _moveIsAllowed = this._scrollView._moveIsAllowed.bind(this._scrollView);
        this._scrollView._moveIsAllowed = e => {
            for (let i = 0; i < (null === (_this$_toolboxes = this._toolboxes) || void 0 === _this$_toolboxes ? void 0 : _this$_toolboxes.length); i += 1) {
                var _this$_toolboxes, _this$_toolboxes2;
                const $element = null === (_this$_toolboxes2 = this._toolboxes) || void 0 === _this$_toolboxes2 ? void 0 : _this$_toolboxes2[i];
                if ($($element).children(".dxdi-tb-start-drag-flag").length) {
                    return false
                }
            }
            return _moveIsAllowed(e)
        };
        const $accordion = $("<div>").appendTo(this._scrollView.content());
        this._updateElementWidth($accordion);
        this._renderAccordion($accordion)
    }
    _getAccordionDataSource() {
        const result = [];
        const toolboxGroups = this.option("toolboxGroups");
        for (let i = 0; i < toolboxGroups.length; i += 1) {
            const {
                category: category
            } = toolboxGroups[i];
            const {
                title: title
            } = toolboxGroups[i];
            const groupObj = {
                category: category,
                title: title || category,
                expanded: toolboxGroups[i].expanded,
                displayMode: toolboxGroups[i].displayMode,
                shapes: toolboxGroups[i].shapes,
                onTemplate: (widget, $element, data) => {
                    var _this$_toolboxes3;
                    const $toolboxElement = $($element);
                    this._onShapeCategoryRenderedAction({
                        category: data.category,
                        displayMode: data.displayMode,
                        dataToggle: "shape-toolbox-tooltip",
                        shapes: data.shapes,
                        $element: $toolboxElement
                    });
                    null === (_this$_toolboxes3 = this._toolboxes) || void 0 === _this$_toolboxes3 || _this$_toolboxes3.push($toolboxElement);
                    if ("" !== this._filterText) {
                        var _this$_toolboxes4;
                        this._onFilterChangedAction({
                            text: this._filterText,
                            filteringToolboxes: (null === (_this$_toolboxes4 = this._toolboxes) || void 0 === _this$_toolboxes4 ? void 0 : _this$_toolboxes4.length) - 1
                        })
                    }
                    this._createTooltips($toolboxElement)
                }
            };
            result.push(groupObj)
        }
        return result
    }
    _createTooltips($toolboxElement) {
        if (this._isTouchMode()) {
            return
        }
        const targets = $toolboxElement.find('[data-toggle="shape-toolbox-tooltip"]');
        const $container = this.$element();
        targets.each(((_, element) => {
            const $target = $(element);
            const title = $target.attr("title");
            if (title) {
                const $tooltip = $("<div>").text(title).appendTo($container);
                this._createComponent($tooltip, Tooltip, {
                    target: $target.get(0),
                    showEvent: "mouseenter",
                    hideEvent: "mouseleave",
                    position: "top",
                    animation: {
                        show: {
                            type: "fade",
                            from: 0,
                            to: 1,
                            delay: 500
                        },
                        hide: {
                            type: "fade",
                            from: 1,
                            to: 0,
                            delay: 100
                        }
                    }
                })
            }
        }))
    }
    _isTouchMode() {
        const {
            Browser: Browser
        } = getDiagram();
        return Browser.TouchUI
    }
    _renderAccordion($container) {
        const {
            disabled: disabled
        } = this.option();
        this._accordion = this._createComponent($container, Accordion, {
            multiple: true,
            animationDuration: 0,
            activeStateEnabled: false,
            focusStateEnabled: false,
            hoverStateEnabled: false,
            collapsible: true,
            displayExpr: "title",
            dataSource: this._getAccordionDataSource(),
            disabled: disabled,
            itemTemplate: (data, index, $element) => {
                data.onTemplate(this, $element, data)
            },
            onSelectionChanged: e => {
                this._updateScrollAnimateSubscription(e.component)
            },
            onContentReady: e => {
                e.component.option("selectedItems", []);
                const items = e.component.option("dataSource");
                for (let i = 0; i < (null === items || void 0 === items ? void 0 : items.length); i += 1) {
                    if (false === (null === items || void 0 === items ? void 0 : items[i].expanded)) {
                        e.component.collapseItem(i)
                    } else if (true === (null === items || void 0 === items ? void 0 : items[i].expanded)) {
                        e.component.expandItem(i)
                    }
                }
                if (null !== items && void 0 !== items && items.length && void 0 === items[0].expanded) {
                    e.component.expandItem(0)
                }
                this._updateScrollAnimateSubscription(e.component)
            }
        })
    }
    _updateScrollAnimateSubscription(component) {
        component._deferredAnimate = new Deferred;
        component._deferredAnimate.done((() => {
            this.updateMaxHeight();
            this._scrollView.update();
            this._updateScrollAnimateSubscription(component)
        }))
    }
    _onInputChanged(text) {
        var _this$_toolboxes5;
        this._filterText = text;
        this._onFilterChangedAction({
            text: this._filterText,
            filteringToolboxes: null === (_this$_toolboxes5 = this._toolboxes) || void 0 === _this$_toolboxes5 ? void 0 : _this$_toolboxes5.map((($element, index) => index))
        });
        this.updateTooltips();
        this.updateMaxHeight();
        this._scrollView.update()
    }
    updateFilter() {
        this._onInputChanged(this._filterText)
    }
    updateTooltips() {
        var _this$_toolboxes6;
        null === (_this$_toolboxes6 = this._toolboxes) || void 0 === _this$_toolboxes6 || _this$_toolboxes6.forEach(($element => {
            const $tooltipContainer = $($element);
            this._createTooltips($tooltipContainer)
        }))
    }
    _createOnShapeCategoryRenderedAction() {
        this._onShapeCategoryRenderedAction = this._createActionByOption("onShapeCategoryRendered")
    }
    _createOnFilterChangedAction() {
        this._onFilterChangedAction = this._createActionByOption("onFilterChanged")
    }
    _optionChanged(args) {
        var _this$_accordion;
        switch (args.name) {
            case "onShapeCategoryRendered":
                this._createOnShapeCategoryRenderedAction();
                break;
            case "onFilterChanged":
                this._createOnFilterChangedAction();
                break;
            case "showSearch":
            case "toolboxWidth":
                this._invalidate();
                break;
            case "toolboxGroups":
                null === (_this$_accordion = this._accordion) || void 0 === _this$_accordion || _this$_accordion.option("dataSource", this._getAccordionDataSource());
                break;
            default:
                super._optionChanged(args)
        }
    }
}
export default DiagramToolbox;
