// devextreme/esm/__internal/scheduler/r1/components/wrappers/group_panel.js
/**
 * DevExtreme (esm/__internal/scheduler/r1/components/wrappers/group_panel.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import registerComponent from "../../../../../core/component_registrator";
import {
    ComponentWrapper
} from "../../../../core/r1/index";
import {
    GroupPanel
} from "../base/group_panel";
export class GroupPanelComponent extends ComponentWrapper {
    _setOptionsByReference() {
        super._setOptionsByReference();
        this._optionsByReference = Object.assign({}, this._optionsByReference, {
            resourceCellTemplate: true
        })
    }
    get _propsInfo() {
        return {
            twoWay: [],
            allowNull: [],
            elements: [],
            templates: ["resourceCellTemplate"],
            props: ["viewContext", "groups", "groupOrientation", "groupPanelData", "groupByDate", "height", "className", "resourceCellTemplate"]
        }
    }
    get _viewComponent() {
        return GroupPanel
    }
}
registerComponent("dxGroupPanel", GroupPanelComponent);
