// devextreme/esm/__internal/scheduler/r1/components/wrappers/time_panel.js
/**
 * DevExtreme (esm/__internal/scheduler/r1/components/wrappers/time_panel.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import registerComponent from "../../../../../core/component_registrator";
import {
    ComponentWrapper
} from "../../../../core/r1/index";
import {
    TimePanelTable
} from "../base/time_panel_table";
export class TimePanelComponent extends ComponentWrapper {
    _setOptionsByReference() {
        super._setOptionsByReference();
        this._optionsByReference = Object.assign({}, this._optionsByReference, {
            timeCellTemplate: true
        })
    }
    get _propsInfo() {
        return {
            twoWay: [],
            allowNull: [],
            elements: [],
            templates: ["timeCellTemplate"],
            props: ["viewContext", "groupOrientation", "timePanelData", "timeCellTemplate"]
        }
    }
    get _viewComponent() {
        return TimePanelTable
    }
}
registerComponent("dxTimePanelTableLayout", TimePanelComponent);
