// devextreme/cjs/__internal/core/ai_integration/commands/index.js
/**
 * DevExtreme (cjs/__internal/core/ai_integration/commands/index.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BaseCommand", {
    enumerable: true,
    get: function() {
        return _base.BaseCommand
    }
});
Object.defineProperty(exports, "ChangeStyleCommand", {
    enumerable: true,
    get: function() {
        return _changeStyle.ChangeStyleCommand
    }
});
Object.defineProperty(exports, "ChangeToneCommand", {
    enumerable: true,
    get: function() {
        return _changeTone.ChangeToneCommand
    }
});
Object.defineProperty(exports, "ExecuteCommand", {
    enumerable: true,
    get: function() {
        return _execute.ExecuteCommand
    }
});
Object.defineProperty(exports, "ExpandCommand", {
    enumerable: true,
    get: function() {
        return _expand.ExpandCommand
    }
});
Object.defineProperty(exports, "ProofreadCommand", {
    enumerable: true,
    get: function() {
        return _proofread.ProofreadCommand
    }
});
Object.defineProperty(exports, "ShortenCommand", {
    enumerable: true,
    get: function() {
        return _shorten.ShortenCommand
    }
});
Object.defineProperty(exports, "SmartPasteCommand", {
    enumerable: true,
    get: function() {
        return _smartPaste.SmartPasteCommand
    }
});
Object.defineProperty(exports, "SummarizeCommand", {
    enumerable: true,
    get: function() {
        return _summarize.SummarizeCommand
    }
});
Object.defineProperty(exports, "TranslateCommand", {
    enumerable: true,
    get: function() {
        return _translate.TranslateCommand
    }
});
var _base = require("../../../core/ai_integration/commands/base");
var _changeStyle = require("../../../core/ai_integration/commands/changeStyle");
var _changeTone = require("../../../core/ai_integration/commands/changeTone");
var _execute = require("../../../core/ai_integration/commands/execute");
var _expand = require("../../../core/ai_integration/commands/expand");
var _proofread = require("../../../core/ai_integration/commands/proofread");
var _shorten = require("../../../core/ai_integration/commands/shorten");
var _smartPaste = require("../../../core/ai_integration/commands/smartPaste");
var _summarize = require("../../../core/ai_integration/commands/summarize");
var _translate = require("../../../core/ai_integration/commands/translate");
