// devextreme/esm/__internal/scheduler/__tests__/__mock__/m_mock_scheduler.js
/**
 * DevExtreme (esm/__internal/scheduler/__tests__/__mock__/m_mock_scheduler.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    jest
} from "@jest/globals";
import {
    logger
} from "../../../core/utils/m_console";
import DOMComponent from "../../../core/widget/dom_component";
import SchedulerWorkSpace from "../../workspaces/m_work_space";
export const DEFAULT_CELL_WIDTH = 250;
export const DEFAULT_CELL_HEIGHT = 80;
export const DEFAULT_TIMELINE_CELL_HEIGHT = 450;
export const setupSchedulerTestEnvironment = function() {
    let {
        width: width = DEFAULT_CELL_WIDTH,
        height: height = DEFAULT_CELL_HEIGHT
    } = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {};
    jest.spyOn(logger, "warn").mockImplementation((() => {}));
    DOMComponent.prototype._isVisible = jest.fn((() => true));
    SchedulerWorkSpace.prototype._createCrossScrollingConfig = () => ({
        direction: "both",
        onScroll: jest.fn(),
        onEnd: jest.fn()
    });
    const {
        getComputedStyle: getComputedStyle
    } = window;
    window.getComputedStyle = jest.fn((function(element, pseudoElement) {
        const styles = getComputedStyle.call(this, element, pseudoElement);
        if (element.classList.contains("dx-scheduler-appointment-collector")) {
            styles.setProperty("width", "24px");
            styles.setProperty("height", "20px");
            styles.setProperty("margin", "3px")
        }
        return styles
    }));
    Element.prototype.getBoundingClientRect = jest.fn((function() {
        const classList = Array.from(this.classList);
        if (true === (classList.includes("dx-scheduler-date-table-cell") || classList.includes("dx-scheduler-all-day-table-cell"))) {
            return {
                width: width,
                height: height,
                top: 0,
                left: 0,
                bottom: height,
                right: width,
                x: 0,
                y: 0,
                toJSON: () => {}
            }
        } else {
            return {
                width: 0,
                height: 0,
                top: 0,
                left: 0,
                bottom: 0,
                right: 0,
                x: 0,
                y: 0,
                toJSON: () => {}
            }
        }
    }))
};
