// devextreme/esm/__internal/grids/new/grid_core/options.js
/**
 * DevExtreme (esm/__internal/grids/new/grid_core/options.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import browser from "../../../../core/utils/browser";
import {
    isMaterialBased
} from "../../../../ui/themes";
import * as columnChooser from "./column_chooser/index";
import * as columnsController from "./columns_controller/index";
import * as contentView from "./content_view/index";
import * as dataController from "./data_controller/index";
import * as editing from "./editing/index";
import * as filterSync from "./filtering/filter_sync/index";
import * as headerFilter from "./filtering/header_filter/index";
import * as filterController from "./filtering/index";
import {
    filterPanel
} from "./filtering/index";
import * as keyboardNavigation from "./keyboard_navigation/index";
import * as pager from "./pager/index";
import * as searchPanel from "./search/index";
import * as selection from "./selection/index";
import * as sortingController from "./sorting_controller/index";
import * as toolbar from "./toolbar/index";
export const defaultOptions = Object.assign({}, dataController.defaultOptions, sortingController.defaultOptions, columnsController.defaultOptions, pager.defaultOptions, filterPanel.defaultOptions, filterController.defaultOptions, headerFilter.defaultOptions, filterSync.defaultOptions, contentView.defaultOptions, searchPanel.defaultOptions, columnChooser.defaultOptions, selection.defaultOptions, toolbar.defaultOptions, editing.defaultOptions, keyboardNavigation.defaultOptions);
export const defaultOptionsRules = [{
    device: () => isMaterialBased(),
    options: {
        headerFilter: {
            height: 315
        },
        editing: {
            useIcons: true
        },
        selection: {
            showCheckBoxesMode: "always"
        }
    }
}, {
    device: () => browser.webkit,
    options: {
        loadingTimeout: 30,
        loadPanel: {
            animation: {
                show: {
                    easing: "cubic-bezier(1, 0, 1, 0)",
                    duration: 500,
                    from: {
                        opacity: 0
                    },
                    to: {
                        opacity: 1
                    }
                }
            }
        }
    }
}];
