// devextreme/cjs/__internal/pagination/common/base_pagination_props.js
/**
 * DevExtreme (cjs/__internal/pagination/common/base_pagination_props.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.BasePaginationDefaultProps = void 0;
var _base_props = require("../../core/r1/base_props");
var _message = _interopRequireDefault(require("../../../common/core/localization/message"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const BasePaginationDefaultProps = exports.BasePaginationDefaultProps = Object.assign({}, _base_props.BaseWidgetDefaultProps, {
    isGridCompatibilityMode: false,
    showInfo: false,
    displayMode: "adaptive",
    maxPagesCount: 10,
    pageCount: 1,
    visible: true,
    hasKnownLastPage: true,
    pagesNavigatorVisible: "auto",
    showPageSizeSelector: "auto",
    allowedPageSizes: [5, 10],
    showNavigationButtons: false,
    itemCount: 1,
    label: _message.default.format("dxPagination-ariaLabel")
});
