// devextreme/cjs/__internal/pagination/common/consts.js
/**
 * DevExtreme (cjs/__internal/pagination/common/consts.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PAGINATION_SELECTION_CLASS = exports.PAGINATION_SELECTED_PAGE_SIZE_CLASS = exports.PAGINATION_PAGE_SIZE_CLASS = exports.PAGINATION_PAGE_SIZES_CLASS = exports.PAGINATION_PAGE_INDEXES_CLASS = exports.PAGINATION_PAGE_CLASS = exports.PAGINATION_PAGES_CLASS = exports.PAGINATION_CLASS = exports.PAGER_CLASS = exports.LIGHT_MODE_CLASS = exports.FIRST_CHILD_CLASS = void 0;
const PAGER_CLASS = exports.PAGER_CLASS = "dx-pager";
const PAGINATION_CLASS = exports.PAGINATION_CLASS = "dx-pagination";
const LIGHT_MODE_CLASS = exports.LIGHT_MODE_CLASS = "dx-light-mode";
const PAGINATION_PAGES_CLASS = exports.PAGINATION_PAGES_CLASS = "dx-pages";
const PAGINATION_PAGE_INDEXES_CLASS = exports.PAGINATION_PAGE_INDEXES_CLASS = "dx-page-indexes";
const PAGINATION_PAGE_CLASS = exports.PAGINATION_PAGE_CLASS = "dx-page";
const PAGINATION_SELECTION_CLASS = exports.PAGINATION_SELECTION_CLASS = "dx-selection";
const PAGINATION_PAGE_SIZE_CLASS = exports.PAGINATION_PAGE_SIZE_CLASS = "dx-page-size";
const PAGINATION_PAGE_SIZES_CLASS = exports.PAGINATION_PAGE_SIZES_CLASS = "dx-page-sizes";
const PAGINATION_SELECTED_PAGE_SIZE_CLASS = exports.PAGINATION_SELECTED_PAGE_SIZE_CLASS = `${PAGINATION_PAGE_SIZE_CLASS} ${PAGINATION_SELECTION_CLASS}`;
const FIRST_CHILD_CLASS = exports.FIRST_CHILD_CLASS = "dx-first-child";
