// devextreme/cjs/__internal/pagination/common/pagination_config_provider.js
/**
 * DevExtreme (cjs/__internal/pagination/common/pagination_config_provider.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PaginationConfigProviderDefaultProps = exports.PaginationConfigProvider = void 0;
var _index = require("../../core/r1/runtime/inferno/index");
var _pagination_config_context = require("./pagination_config_context");
const PaginationConfigProviderDefaultProps = exports.PaginationConfigProviderDefaultProps = {};
class PaginationConfigProvider extends _index.BaseInfernoComponent {
    constructor() {
        super(...arguments);
        this.state = {}
    }
    getConfig() {
        return {
            isGridCompatibilityMode: this.props.isGridCompatibilityMode
        }
    }
    getChildContext() {
        return Object.assign({}, this.context, {
            [_pagination_config_context.PaginationConfigContext.id]: this.getConfig() || _pagination_config_context.PaginationConfigContext.defaultValue
        })
    }
    render() {
        return this.props.children
    }
}
exports.PaginationConfigProvider = PaginationConfigProvider;
PaginationConfigProvider.defaultProps = PaginationConfigProviderDefaultProps;
