// devextreme/cjs/__internal/grids/new/grid_core/data_controller/public_methods.js
/**
 * DevExtreme (cjs/__internal/grids/new/grid_core/data_controller/public_methods.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PublicMethods = PublicMethods;
var _m_utils = require("../../../../data/m_utils");

function PublicMethods(GridCore) {
    return class extends GridCore {
        getDataSource() {
            return this.dataController.dataSource.peek()
        }
        byKey(key) {
            const items = this.getDataSource().items();
            const store = this.getDataSource().store();
            const keyExpr = store.key();
            const foundItem = items.find((item => (0, _m_utils.keysEqual)(keyExpr, key, this.keyOf(item))));
            if (foundItem) {
                return Promise.resolve(foundItem)
            }
            return store.byKey(key)
        }
        getCombinedFilter() {
            return this.dataController.getCombinedFilter()
        }
        keyOf(obj) {
            return this.dataController.getDataKey(obj)
        }
        pageCount() {
            return this.dataController.pageCount.peek()
        }
        pageSize(value) {
            if (void 0 === value) {
                return this.dataController.pageSize.peek()
            }
            this.dataController.pageSize.value = value
        }
        pageIndex(newIndex) {
            if (void 0 === newIndex) {
                return this.dataController.pageIndex.peek()
            }
            this.dataController.pageIndex.value = newIndex;
            return this.dataController.waitLoaded()
        }
        totalCount() {
            return this.dataController.totalCount.peek()
        }
    }
}
