// devextreme/esm/__internal/ui/gantt/ui.gantt.model_changes_listener.js
/**
 * DevExtreme (esm/__internal/ui/gantt/ui.gantt.model_changes_listener.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
const GANTT_TASKS = "tasks";
const GANTT_DEPENDENCIES = "dependencies";
const GANTT_RESOURCES = "resources";
const GANTT_RESOURCE_ASSIGNMENTS = "resourceAssignments";
export const ModelChangesListener = {
    create: gantt => ({
        NotifyTaskCreated: (task, callback, errorCallback) => {
            gantt._onRecordInserted("tasks", task, callback)
        },
        NotifyTaskRemoved: (taskId, errorCallback, task) => {
            gantt._onRecordRemoved("tasks", taskId, task)
        },
        NotifyTaskUpdated: (taskId, newValues, errorCallback) => {
            gantt._onRecordUpdated("tasks", taskId, newValues)
        },
        NotifyParentTaskUpdated: (task, errorCallback) => {
            gantt._onParentTaskUpdated(task)
        },
        NotifyDependencyInserted: (dependency, callback, errorCallback) => {
            gantt._onRecordInserted("dependencies", dependency, callback)
        },
        NotifyDependencyRemoved: (dependencyId, errorCallback, dependency) => {
            gantt._onRecordRemoved("dependencies", dependencyId, dependency)
        },
        NotifyResourceCreated: (resource, callback, errorCallback) => {
            gantt._onRecordInserted("resources", resource, callback)
        },
        NotifyResourceRemoved: (resourceId, errorCallback, resource) => {
            gantt._onRecordRemoved("resources", resourceId, resource)
        },
        NotifyResourceAssigned: (assignment, callback, errorCallback) => {
            gantt._onRecordInserted("resourceAssignments", assignment, callback)
        },
        NotifyResourceUnassigned: (assignmentId, errorCallback, assignment) => {
            gantt._onRecordRemoved("resourceAssignments", assignmentId, assignment)
        },
        NotifyParentDataRecalculated: data => {
            gantt._onParentTasksRecalculated(data)
        },
        NotifyTaskCreating: args => {
            var _gantt$_actionsManage;
            null === (_gantt$_actionsManage = gantt._actionsManager) || void 0 === _gantt$_actionsManage || _gantt$_actionsManage.raiseInsertingAction("tasks", args)
        },
        NotifyTaskRemoving: args => {
            var _gantt$_actionsManage2;
            null === (_gantt$_actionsManage2 = gantt._actionsManager) || void 0 === _gantt$_actionsManage2 || _gantt$_actionsManage2.raiseDeletingAction("tasks", args)
        },
        NotifyTaskUpdating: args => {
            var _gantt$_actionsManage3;
            null === (_gantt$_actionsManage3 = gantt._actionsManager) || void 0 === _gantt$_actionsManage3 || _gantt$_actionsManage3.raiseUpdatingAction("tasks", args)
        },
        NotifyTaskMoving: args => {
            var _gantt$_actionsManage4, _gantt$_actionsManage5;
            null === (_gantt$_actionsManage4 = gantt._actionsManager) || void 0 === _gantt$_actionsManage4 || _gantt$_actionsManage4.raiseUpdatingAction("tasks", args, null === (_gantt$_actionsManage5 = gantt._actionsManager) || void 0 === _gantt$_actionsManage5 ? void 0 : _gantt$_actionsManage5.getTaskMovingAction())
        },
        NotifyTaskEditDialogShowing: args => {
            var _gantt$_actionsManage6;
            null === (_gantt$_actionsManage6 = gantt._actionsManager) || void 0 === _gantt$_actionsManage6 || _gantt$_actionsManage6.raiseTaskEditDialogShowingAction(args)
        },
        NotifyResourceManagerDialogShowing: args => {
            var _gantt$_actionsManage7;
            null === (_gantt$_actionsManage7 = gantt._actionsManager) || void 0 === _gantt$_actionsManage7 || _gantt$_actionsManage7.raiseResourceManagerDialogShowingAction(args)
        },
        NotifyDependencyInserting: args => {
            var _gantt$_actionsManage8;
            null === (_gantt$_actionsManage8 = gantt._actionsManager) || void 0 === _gantt$_actionsManage8 || _gantt$_actionsManage8.raiseInsertingAction("dependencies", args)
        },
        NotifyDependencyRemoving: args => {
            var _gantt$_actionsManage9;
            null === (_gantt$_actionsManage9 = gantt._actionsManager) || void 0 === _gantt$_actionsManage9 || _gantt$_actionsManage9.raiseDeletingAction("dependencies", args)
        },
        NotifyResourceCreating: args => {
            var _gantt$_actionsManage10;
            null === (_gantt$_actionsManage10 = gantt._actionsManager) || void 0 === _gantt$_actionsManage10 || _gantt$_actionsManage10.raiseInsertingAction("resources", args)
        },
        NotifyResourceRemoving: args => {
            var _gantt$_actionsManage11;
            null === (_gantt$_actionsManage11 = gantt._actionsManager) || void 0 === _gantt$_actionsManage11 || _gantt$_actionsManage11.raiseDeletingAction("resources", args)
        },
        NotifyResourceAssigning: args => {
            var _gantt$_actionsManage12;
            null === (_gantt$_actionsManage12 = gantt._actionsManager) || void 0 === _gantt$_actionsManage12 || _gantt$_actionsManage12.raiseInsertingAction("resourceAssignments", args)
        },
        NotifyResourceUnassigning: args => {
            var _gantt$_actionsManage13;
            null === (_gantt$_actionsManage13 = gantt._actionsManager) || void 0 === _gantt$_actionsManage13 || _gantt$_actionsManage13.raiseDeletingAction("resourceAssignments", args)
        },
        NotifyScaleCellPrepared: args => {
            var _gantt$_actionsManage14;
            null === (_gantt$_actionsManage14 = gantt._actionsManager) || void 0 === _gantt$_actionsManage14 || _gantt$_actionsManage14.raiseScaleCellPreparedAction(args)
        },
        NotifyGanttViewUpdated: () => {
            gantt._onGanttViewCoreUpdated()
        }
    })
};
