// devextreme/cjs/__internal/grids/new/grid_core/column_chooser/column_chooser.js
/**
 * DevExtreme (cjs/__internal/grids/new/grid_core/column_chooser/column_chooser.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ColumnChooser = exports.CLASS = void 0;
var _inferno = require("inferno");
var _message = _interopRequireDefault(require("../../../../../localization/message"));
var _column_sortable = require("../../card_view/header_panel/column_sortable");
var _item = require("../../card_view/header_panel/item");
var _const = require("../const");
var _popup = require("../inferno_wrappers/popup");
var _tree_view = require("../inferno_wrappers/tree_view");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const CLASS = exports.CLASS = {
    excludeFlexBox: _const.CLASSES.excludeFlexBox,
    root: "column-chooser",
    toolbarBtn: "column-chooser-button",
    list: "column-chooser-list",
    plain: "column-chooser-plain",
    dragMode: "column-chooser-mode-drag",
    selectMode: "column-chooser-mode-select",
    treeviewItem: "dx-treeview-item",
    treeviewExpanderIcon: "dx-treeview-expander-icon-stub"
};
class ColumnChooser extends _inferno.Component {
    constructor() {
        super(...arguments);
        this.onShowing = e => {
            const popup = e.component;
            if (void 0 === this.props.popupConfig.position) {
                popup.option("position", {
                    my: "right top",
                    at: "right bottom",
                    of: ".dx-cardview-column-chooser-button",
                    collision: "fit",
                    offset: "-2 -2",
                    boundaryOffset: "2 2"
                })
            }
            this.setPopupAttributes(popup)
        };
        this.getColumnByIndex = index => {
            const treeView = this.props.treeViewRef.current;
            const column = treeView.getNodes()[index].itemData.column;
            return column
        }
    }
    render() {
        const {
            visible: visible,
            popupConfig: popupConfig,
            popupRef: popupRef,
            sortableConfig: sortableConfig
        } = this.props;
        if (!visible) {
            return (0, _inferno.createFragment)()
        }
        const treeView = this.getTreeView();
        return (0, _inferno.createVNode)(1, "div", CLASS.excludeFlexBox, (0, _inferno.createComponentVNode)(2, _popup.Popup, {
            componentRef: popupRef,
            visible: true,
            shading: false,
            dragEnabled: true,
            resizeEnabled: true,
            showCloseButton: true,
            _loopFocus: true,
            toolbarItems: popupConfig.toolbarItems,
            wrapperAttr: {
                class: this.getPopupWrapperClass()
            },
            width: popupConfig.width,
            height: popupConfig.height,
            container: popupConfig.container,
            position: popupConfig.position,
            onHidden: popupConfig.onHidden,
            onShowing: this.onShowing,
            children: (0, _inferno.createComponentVNode)(2, _column_sortable.ColumnSortable, {
                height: "100%",
                source: "column-chooser",
                filter: `.${CLASS.treeviewItem}`,
                getColumnByIndex: this.getColumnByIndex,
                isColumnDraggable: sortableConfig.isColumnDraggable,
                visibleColumns: this.props.visibleColumns,
                allowDragging: !this.isSelectMode(),
                columnDragTemplate: _item.Item,
                onColumnMove: this.props.onColumnMove,
                onDragStart: sortableConfig.onDragStart,
                onDragEnd: sortableConfig.onDragEnd,
                onPlaceholderPrepared: sortableConfig.onPlaceholderPrepared,
                children: treeView
            })
        }), 2)
    }
    isSelectMode() {
        return "select" === this.props.mode
    }
    addWidgetPrefix(cssClass) {
        return `dx-cardview-${cssClass}`
    }
    getPopupWrapperClass() {
        const modeSpecificClass = this.isSelectMode() ? CLASS.selectMode : CLASS.dragMode;
        return [this.addWidgetPrefix(CLASS.root), this.addWidgetPrefix(modeSpecificClass)].join(" ")
    }
    setPopupAttributes(popup) {
        popup.setAria({
            label: _message.default.format("dxDataGrid-columnChooserTitle")
        });
        popup.$content().addClass(this.addWidgetPrefix(CLASS.list));
        popup.$content().toggleClass(this.addWidgetPrefix(CLASS.plain), !this.props.isBandColumnsUsed)
    }
    getTreeView() {
        const {
            treeViewRef: treeViewRef,
            treeViewConfig: treeViewConfig,
            treeViewSelectModeConfig: treeViewSelectModeConfig,
            treeViewDragAndDropModeConfig: treeViewDragAndDropModeConfig
        } = this.props;
        return (0, _inferno.normalizeProps)((0, _inferno.createComponentVNode)(2, _tree_view.TreeView, Object.assign({
            componentRef: treeViewRef,
            dataStructure: "plain",
            activeStateEnabled: true,
            focusStateEnabled: true,
            hoverStateEnabled: true,
            rootValue: null,
            searchEditorOptions: treeViewConfig.searchEditorOptions,
            searchEnabled: treeViewConfig.searchEnabled,
            searchTimeout: treeViewConfig.searchTimeout,
            noDataText: treeViewConfig.noDataText,
            items: treeViewConfig.items
        }, this.isSelectMode() ? treeViewSelectModeConfig : treeViewDragAndDropModeConfig)))
    }
}
exports.ColumnChooser = ColumnChooser;
