// devextreme/cjs/__internal/grids/new/grid_core/column_chooser/view.test.js
/**
 * DevExtreme (cjs/__internal/grids/new/grid_core/column_chooser/view.test.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.renderColumnChooserWithToolbar = exports.renderColumnChooser = void 0;
var _globals = require("@jest/globals");
var _inferno = require("inferno");
var _columns_controller = require("../columns_controller");
var _di = require("../di.test_utils");
var _options_controller = require("../options_controller/options_controller.mock");
var _controller = require("../toolbar/controller");
var _view = require("../toolbar/view");
var _view2 = require("./view");
const CLASS = {
    root: "dx-cardview-column-chooser",
    selectMode: "dx-cardview-column-chooser-mode-select",
    dragMode: "dx-cardview-column-chooser-mode-drag",
    list: "dx-cardview-column-chooser-list",
    plain: "dx-cardview-column-chooser-plain",
    item: "dx-column-chooser-item",
    treeViewItem: "dx-treeview-item-content"
};
const renderColumnChooser = function() {
    let options = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {};
    let open = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : true;
    const context = (0, _di.getContext)(options);
    const element = document.createElement("div");
    const columnChooserView = context.get(_view2.ColumnChooserView);
    const columnsController = context.get(_columns_controller.ColumnsController);
    columnChooserView.render(element);
    if (open) {
        var _columnChooserView$po;
        columnChooserView.show();
        (0, _inferno.rerender)();
        null === (_columnChooserView$po = columnChooserView.popupRef.current) || void 0 === _columnChooserView$po || _columnChooserView$po.option("onShowing")({
            component: columnChooserView.popupRef.current
        })
    }
    return {
        element: element,
        columnChooserView: columnChooserView,
        columnsController: columnsController
    }
};
exports.renderColumnChooser = renderColumnChooser;
const renderColumnChooserWithToolbar = function() {
    let options = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : {};
    const context = (0, _di.getContext)(options);
    const rootElement = document.createElement("div");
    const optionsController = context.get(_options_controller.OptionsControllerMock);
    const columnChooserView = context.get(_view2.ColumnChooserView);
    const toolbarView = context.get(_view.ToolbarView);
    const toolbarController = context.get(_controller.ToolbarController);
    const toolbarElement = document.createElement("div");
    const columnChooserElement = document.createElement("div");
    rootElement.append(toolbarElement, columnChooserElement);
    toolbarView.render(toolbarElement);
    columnChooserView.render(columnChooserElement);
    return {
        element: rootElement,
        optionsController: optionsController,
        toolbarController: toolbarController
    }
};
exports.renderColumnChooserWithToolbar = renderColumnChooserWithToolbar;
(0, _globals.describe)("ColumnChooser", (() => {
    (0, _globals.describe)("View", (() => {
        (0, _globals.it)("toolbar button", (() => {
            const {
                toolbarController: toolbarController,
                optionsController: optionsController
            } = renderColumnChooserWithToolbar({
                toolbar: {
                    visible: true
                },
                columnChooser: {
                    enabled: true
                }
            });
            let toolbarItems = toolbarController.items.peek();
            (0, _globals.expect)(toolbarItems).toHaveLength(1);
            (0, _globals.expect)(toolbarItems[0].name).toEqual("columnChooserButton");
            optionsController.option("columnChooser.enabled", false);
            toolbarItems = toolbarController.items.peek();
            (0, _globals.expect)(toolbarItems).toHaveLength(0)
        }));
        _globals.it.each([{
            mode: "dragAndDrop",
            opened: true,
            result: true
        }, {
            mode: "dragAndDrop",
            opened: false,
            result: false
        }, {
            mode: "select",
            opened: true,
            result: false
        }, {
            mode: "select",
            opened: false,
            result: false
        }])("'dragModeOpened' state is correct when mode: '$mode', column chooser opened: $opened", (_ref => {
            let {
                mode: mode,
                opened: opened,
                result: result
            } = _ref;
            const {
                columnChooserView: columnChooserView
            } = renderColumnChooser({
                columns: ["Column 1", "Column 2"],
                columnChooser: {
                    enabled: true,
                    mode: mode
                }
            }, false);
            if (opened) {
                columnChooserView.show()
            }(0, _globals.expect)(columnChooserView.dragModeOpened.peek()).toEqual(result)
        }));
        (0, _globals.describe)("Select mode", (() => {
            (0, _globals.it)("popup has correct css classes", (() => {
                const {
                    columnChooserView: columnChooserView
                } = renderColumnChooser({
                    columns: ["Column 1", "Column 2"],
                    columnChooser: {
                        enabled: true,
                        mode: "select"
                    }
                });
                const popup = columnChooserView.popupRef.current;
                const wrapperElement = null === popup || void 0 === popup ? void 0 : popup._$wrapper.get(0);
                (0, _globals.expect)(null === popup || void 0 === popup ? void 0 : popup.content().classList).toContain(CLASS.list);
                (0, _globals.expect)(null === popup || void 0 === popup ? void 0 : popup.content().classList).toContain(CLASS.plain);
                (0, _globals.expect)(null === wrapperElement || void 0 === wrapperElement ? void 0 : wrapperElement.classList).toContain(CLASS.root);
                (0, _globals.expect)(null === wrapperElement || void 0 === wrapperElement ? void 0 : wrapperElement.classList).toContain(CLASS.selectMode)
            }))
        }));
        (0, _globals.describe)("Drag mode", (() => {
            (0, _globals.it)("popup has correct css classes", (() => {
                const {
                    columnChooserView: columnChooserView
                } = renderColumnChooser({
                    columns: ["Column 1", "Column 2"],
                    columnChooser: {
                        enabled: true,
                        mode: "dragAndDrop"
                    }
                });
                const popup = columnChooserView.popupRef.current;
                const overlayElement = null === popup || void 0 === popup ? void 0 : popup._$wrapper.get(0);
                (0, _globals.expect)(null === popup || void 0 === popup ? void 0 : popup.content().classList).toContain(CLASS.list);
                (0, _globals.expect)(null === overlayElement || void 0 === overlayElement ? void 0 : overlayElement.classList).toContain(CLASS.root);
                (0, _globals.expect)(null === overlayElement || void 0 === overlayElement ? void 0 : overlayElement.classList).toContain(CLASS.dragMode)
            }))
        }))
    }))
}));
