// devextreme/cjs/__internal/grids/data_grid/grouping/const.js
/**
 * DevExtreme (cjs/__internal/grids/data_grid/grouping/const.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.CONTEXT_MENU_UNGROUP_COLUMN_ICON_NAME = exports.CONTEXT_MENU_UNGROUP_ALL_COLUMNS_ICON_NAME = exports.CONTEXT_MENU_GROUP_BY_COLUMN_ICON_NAME = exports.CLASSES = void 0;
const CLASSES = exports.CLASSES = {
    groupPanel: "dx-datagrid-group-panel",
    groupPanelMessage: "dx-group-panel-message",
    groupPanelItem: "dx-group-panel-item",
    groupPanelLabel: "dx-toolbar-label",
    groupPanelContainer: "dx-toolbar-item"
};
const CONTEXT_MENU_GROUP_BY_COLUMN_ICON_NAME = exports.CONTEXT_MENU_GROUP_BY_COLUMN_ICON_NAME = "groupbycolumn";
const CONTEXT_MENU_UNGROUP_COLUMN_ICON_NAME = exports.CONTEXT_MENU_UNGROUP_COLUMN_ICON_NAME = "ungroupcolumn";
const CONTEXT_MENU_UNGROUP_ALL_COLUMNS_ICON_NAME = exports.CONTEXT_MENU_UNGROUP_ALL_COLUMNS_ICON_NAME = "ungroupallcolumns";
