// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/steps/expand_all_day.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/steps/expand_all_day.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.expandAllDayRegularPanel = exports.expandAllDayAllDayPanel = void 0;
var _m_date = require("../../../../core/utils/m_date");
const toMs = _m_date.dateUtils.dateToMilliseconds;
const MINUTE_MS = toMs("minute");
const DAY_MS = toMs("day");
const getDayInterval = (date, viewOffsetMs) => {
    const trimmedDate = new Date(date).setUTCHours(0, 0, 0, 0);
    const startOfDay = trimmedDate + viewOffsetMs;
    const endOfDay = trimmedDate + DAY_MS + viewOffsetMs;
    return {
        min: startOfDay,
        max: endOfDay
    }
};
const getShiftedStartDate = (startDate, viewOffsetMs) => {
    const {
        min: min,
        max: max
    } = getDayInterval(startDate, viewOffsetMs);
    switch (true) {
        case startDate > max - MINUTE_MS:
            return max;
        case startDate < min:
            return min - DAY_MS;
        default:
            return min
    }
};
const getShiftedEndDate = (endDate, viewOffsetMs) => {
    const {
        min: min,
        max: max
    } = getDayInterval(endDate, viewOffsetMs);
    switch (true) {
        case endDate >= max:
            return max + DAY_MS - MINUTE_MS;
        case endDate < min:
            return min - MINUTE_MS;
        default:
            return max - MINUTE_MS
    }
};
const expandAllDayAllDayPanel = (entities, endDayHour, viewOffsetMs) => entities.map((entity => {
    if (!entity.allDay) {
        return entity
    }
    if (0 === viewOffsetMs) {
        const minStartDate = new Date(entity.startDateUTC).setUTCHours(endDayHour, 0, 0, 0) - MINUTE_MS;
        const maxEndDate = new Date(entity.endDateUTC).setUTCHours(endDayHour, 0, 0, 0) - MINUTE_MS;
        return Object.assign({}, entity, {
            startDateUTC: Math.min(entity.startDateUTC, minStartDate),
            endDateUTC: maxEndDate
        })
    }
    return Object.assign({}, entity, {
        startDateUTC: getShiftedStartDate(entity.startDateUTC, viewOffsetMs),
        endDateUTC: getShiftedEndDate(entity.endDateUTC, viewOffsetMs)
    })
}));
exports.expandAllDayAllDayPanel = expandAllDayAllDayPanel;
const expandAllDayRegularPanel = entities => entities.map((entity => {
    if (!entity.allDay) {
        return entity
    }
    const startDate = new Date(entity.startDateUTC);
    const endDate = new Date(entity.endDateUTC);
    endDate.setDate(endDate.getDate() + 1);
    return Object.assign({}, entity, {
        endDateUTC: endDate.setUTCHours(startDate.getUTCHours(), startDate.getUTCMinutes(), startDate.getUTCSeconds(), startDate.getUTCMilliseconds())
    })
}));
exports.expandAllDayRegularPanel = expandAllDayRegularPanel;
