// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/steps/snap_to_cells.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/steps/snap_to_cells.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.snapToCells = void 0;
const snapToCells = function(entities, cells) {
    let isSnapToCell = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : true;
    if (!isSnapToCell) {
        return entities
    }
    return entities.map((entity => {
        const {
            cellIndex: cellIndex,
            endCellIndex: endCellIndex
        } = entity;
        return Object.assign({}, entity, {
            startDateUTC: cells[cellIndex].min,
            endDateUTC: cells[endCellIndex].max,
            duration: cells[endCellIndex].max - cells[cellIndex].min
        })
    }))
};
exports.snapToCells = snapToCells;
