// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/steps/virtual_screen_crop.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/steps/virtual_screen_crop.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.cropByVirtualScreen = void 0;
const cropByVirtualScreen = (entities, _ref) => {
    let {
        isVirtualScrolling: isVirtualScrolling,
        getVirtualScreen: getVirtualScreen
    } = _ref;
    if (!isVirtualScrolling) {
        return entities
    }
    return entities.reduce(((acc, item) => {
        const screen = getVirtualScreen(item.groupIndex);
        const isInsideVirtualScreen = !(item.left + item.width < screen.left || item.left > screen.right || item.top + item.height < screen.top || item.top > screen.bottom);
        if (isInsideVirtualScreen) {
            const right = item.left + item.width;
            const bottom = item.top + item.height;
            const left = Math.max(screen.left, item.left);
            const top = Math.max(screen.top, item.top);
            const width = Math.min(screen.right, right) - left;
            const height = Math.min(screen.bottom, bottom) - top;
            acc.push(Object.assign({}, item, {
                left: left,
                width: width,
                top: top,
                height: height
            }))
        }
        return acc
    }), [])
};
exports.cropByVirtualScreen = cropByVirtualScreen;
