// devextreme/esm/viz/core/utils.js
/**
 * DevExtreme (esm/viz/core/utils.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
export {
    PANE_PADDING,
    getLog,
    getAdjustedLog10,
    raiseTo,
    normalizeAngle,
    convertAngleToRendererSpace,
    degreesToRadians,
    getCosAndSin,
    getDistance,
    getDecimalOrder,
    getAppropriateFormat,
    roundValue,
    getPower,
    map,
    normalizeEnum,
    setCanvasValues,
    normalizeBBox,
    rotateBBox,
    decreaseGaps,
    parseScalar,
    enumParser,
    patchFontOptions,
    convertPolarToXY,
    convertXYToPolar,
    processSeriesTemplate,
    getCategoriesInfo,
    isRelativeHeightPane,
    normalizePanesHeight,
    updatePanesCanvases,
    unique,
    getVerticallyShiftedAngularCoords,
    mergeMarginOptions,
    getVizRangeObject,
    normalizeArcParams,
    convertVisualRangeObject,
    getAddFunction,
    adjustVisualRange,
    getLogExt,
    raiseToExt,
    rangesAreEqual,
    valueOf,
    pointInCanvas,
    getNextDefsSvgId,
    extractColor
}
from "../../__internal/viz/core/utils";
