// devextreme/esm/__internal/scheduler/view_model/generate_view_model/generate_agenda_view_model.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/generate_agenda_view_model.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    getCompareOptions
} from "../common/get_compare_options";
import {
    splitIntervalByDay
} from "../common/split_interval_by_days";
import {
    addLastInGroup
} from "./steps/add_last_in_group";
import {
    addSortedIndex
} from "./steps/add_sorted_index";
import {
    sortByGroupIndex,
    sortByStartDate
} from "./steps/sorting";
import {
    splitByParts
} from "./steps/split_by_parts/split_by_parts";
const saveDatesAfterSplit = entities => entities.map((entity => Object.assign({}, entity, {
    datesAfterSplit: {
        startDateUTC: entity.startDateUTC,
        endDateUTC: entity.endDateUTC
    }
})));
const addAgendaGeometry = (entities, height) => entities.map((entity => Object.assign({}, entity, {
    height: height,
    width: "100%"
})));
export const generateAgendaViewModel = (schedulerStore, items) => {
    const height = schedulerStore.fire("getAgendaVerticalStepHeight");
    const compareOptions = getCompareOptions(schedulerStore);
    const intervals = splitIntervalByDay(Object.assign({}, compareOptions, {
        startDayHour: 0,
        endDayHour: 24
    }));
    let entities = splitByParts(items, intervals);
    entities = saveDatesAfterSplit(entities);
    entities = addAgendaGeometry(entities, height);
    entities = sortByStartDate(entities);
    entities = sortByGroupIndex(entities);
    entities = addLastInGroup(entities);
    return addSortedIndex(entities)
};
