// devextreme/cjs/__internal/scheduler/view_model/common/get_compare_options.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/common/get_compare_options.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getCompareOptions = void 0;
var _m_utils_time_zone = _interopRequireDefault(require("../../m_utils_time_zone"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const getCompareOptions = schedulerStore => {
    const workspace = schedulerStore.getWorkSpace();
    const dateRange = workspace.getDateRange();
    const compareOptions = {
        startDayHour: schedulerStore.getViewOption("startDayHour"),
        endDayHour: schedulerStore.getViewOption("endDayHour"),
        min: _m_utils_time_zone.default.createUTCDateWithLocalOffset(dateRange[0]).getTime(),
        max: _m_utils_time_zone.default.createUTCDateWithLocalOffset(dateRange[1]).getTime(),
        skippedDays: schedulerStore.currentView.skippedDays
    };
    return compareOptions
};
exports.getCompareOptions = getCompareOptions;
