// devextreme/cjs/__internal/grids/new/grid_core/editing/popup/component.js
/**
 * DevExtreme (cjs/__internal/grids/new/grid_core/editing/popup/component.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.EditPopup = void 0;
var _inferno = require("inferno");
var _const = require("../../const");
var _form = require("../../inferno_wrappers/form");
var _popup = require("../../inferno_wrappers/popup");
var _buttons = require("./buttons");
class EditPopup extends _inferno.Component {
    render() {
        if (!this.props.visible) {
            this.props.formRef.current = null;
            return (0, _inferno.createFragment)()
        }
        const toolbarItems = [(0, _buttons.getSaveButtonConfig)({
            onSave: this.props.onSave,
            text: this.props.texts.saveCard
        }), (0, _buttons.getCancelButtonConfig)({
            onCancel: this.props.onCancel,
            text: this.props.texts.cancel
        })];
        return (0, _inferno.createVNode)(1, "div", _const.CLASSES.excludeFlexBox, (0, _inferno.normalizeProps)((0, _inferno.createComponentVNode)(2, _popup.Popup, Object.assign({
            visible: true,
            toolbarItems: toolbarItems,
            onHidden: this.props.onHide,
            showTitle: false
        }, this.props.popupProps, {
            children: (0, _inferno.normalizeProps)((0, _inferno.createComponentVNode)(2, _form.Form, Object.assign({
                componentRef: this.props.formRef,
                colCount: 2,
                labelLocation: "top",
                customizeItem: this.props.customizeItem,
                items: this.props.items
            }, this.props.formProps)))
        }))), 2)
    }
}
exports.EditPopup = EditPopup;
