// devextreme/esm/__internal/pagination/common/pagination_config_provider.js
/**
 * DevExtreme (esm/__internal/pagination/common/pagination_config_provider.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    BaseInfernoComponent
} from "../../core/r1/runtime/inferno/index";
import {
    PaginationConfigContext
} from "./pagination_config_context";
export const PaginationConfigProviderDefaultProps = {};
export class PaginationConfigProvider extends BaseInfernoComponent {
    constructor() {
        super(...arguments);
        this.state = {}
    }
    getConfig() {
        return {
            isGridCompatibilityMode: this.props.isGridCompatibilityMode
        }
    }
    getChildContext() {
        return Object.assign({}, this.context, {
            [PaginationConfigContext.id]: this.getConfig() || PaginationConfigContext.defaultValue
        })
    }
    render() {
        return this.props.children
    }
}
PaginationConfigProvider.defaultProps = PaginationConfigProviderDefaultProps;
