// devextreme/cjs/__internal/ui/diagram/ui.diagram.dialog_manager.js
/**
 * DevExtreme (cjs/__internal/ui/diagram/ui.diagram.dialog_manager.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _message = _interopRequireDefault(require("../../../common/core/localization/message"));
var _renderer = _interopRequireDefault(require("../../../core/renderer"));
var _window = require("../../../core/utils/window");
var _file_uploader = _interopRequireDefault(require("../../../ui/file_uploader"));
var _diagram = require("../../ui/diagram/diagram.importer");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const DiagramDialogManager = {
    getConfigurations() {
        const {
            DiagramCommand: DiagramCommand
        } = (0, _diagram.getDiagram)();
        return this.dialogList || (this.dialogList = [{
            command: DiagramCommand.InsertShapeImage,
            title: _message.default.format("dxDiagram-dialogInsertShapeImageTitle"),
            onGetContent: this.getChangeImageDialogContent
        }, {
            command: DiagramCommand.EditShapeImage,
            title: _message.default.format("dxDiagram-dialogEditShapeImageTitle"),
            onGetContent: this.getChangeImageDialogContent
        }])
    },
    getChangeImageDialogContent(args) {
        const $uploader = (0, _renderer.default)("<div>");
        args.component._createComponent($uploader, _file_uploader.default, {
            selectButtonText: _message.default.format("dxDiagram-dialogEditShapeImageSelectButton"),
            accept: "image/*",
            uploadMode: "useForm",
            onValueChanged(e) {
                const window = (0, _window.getWindow)();
                const reader = new window.FileReader;
                reader.onload = () => {
                    args.component._commandParameter = e.target.result
                };
                reader.readAsDataURL(e.value[0])
            }
        });
        return $uploader
    },
    getDialogParameters(command) {
        const commandIndex = this.getConfigurations().map((c => c.command)).indexOf(command);
        if (commandIndex >= 0) {
            return this.getConfigurations()[commandIndex]
        }
        return null
    }
};
var _default = exports.default = DiagramDialogManager;
