// devextreme/esm/__internal/scheduler/view_model/filtration/utils/filter_by_intervals/filter_by_intervals.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/filtration/utils/filter_by_intervals/filter_by_intervals.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    isAppointmentMatchedIntervals
} from "../../../common/is_appointment_matched_intervals";
const getIntervals = (appointment, _ref) => {
    let {
        allDayIntervals: allDayIntervals,
        regularIntervals: regularIntervals,
        isDateTimeView: isDateTimeView
    } = _ref;
    if (isDateTimeView && appointment.allDay && !appointment.isAllDayPanelOccupied) {
        return regularIntervals.map((interval => ({
            min: new Date(interval.min).setUTCHours(0, 0, 0, 0),
            max: interval.max
        })))
    }
    return appointment.allDay || appointment.isAllDayPanelOccupied ? allDayIntervals : regularIntervals
};
export const filterByIntervals = (entities, options) => entities.filter((appointment => {
    const intervals = getIntervals(appointment, options);
    const fixedAppointment = Object.assign({}, appointment);
    if (appointment.allDay) {
        fixedAppointment.endDateUTC += 1
    }
    return isAppointmentMatchedIntervals({
        startDate: fixedAppointment.startDateUTC,
        endDate: fixedAppointment.endDateUTC
    }, intervals)
}));
