// devextreme/cjs/__internal/grids/new/grid_core/content_view/load_panel.js
/**
 * DevExtreme (cjs/__internal/grids/new/grid_core/content_view/load_panel.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.LoadPanel = void 0;
var _inferno = require("inferno");
var _window = require("../../../../../core/utils/window");
var _index = require("../../../../core/r1/runtime/inferno/index");
var _common_props_context = require("../core/common_props_context");
var _load_panel = require("../inferno_wrappers/load_panel");
class LoadPanel extends _index.BaseInfernoComponent {
    calculatePosition(rootElement) {
        const window = (0, _window.getWindow)();
        if (rootElement.offsetHeight > window.innerHeight) {
            return {
                of: window,
                boundary: rootElement,
                collision: "fit"
            }
        }
        return {
            of: rootElement
        }
    }
    render() {
        const {
            rootElementRef: rootElementRef
        } = this.context[_common_props_context.CommonPropsContext.id];
        const loadPanelProperties = Object.assign({
            container: rootElementRef.current,
            position: this.calculatePosition(rootElementRef.current)
        }, this.props);
        return (0, _inferno.normalizeProps)((0, _inferno.createComponentVNode)(2, _load_panel.LoadPanel, Object.assign({}, loadPanelProperties)))
    }
}
exports.LoadPanel = LoadPanel;
