// devextreme/cjs/__internal/core/ai_integration/core/prompt_manager.js
/**
 * DevExtreme (cjs/__internal/core/ai_integration/core/prompt_manager.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.PromptManager = exports.ERROR_MESSAGES = void 0;
var _index = require("../../../core/ai_integration/templates/index");
const ERROR_MESSAGES = exports.ERROR_MESSAGES = {
    TEMPLATE_NOT_FOUND: "Template not found"
};
class PromptManager {
    constructor() {
        this.templates = new Map(Object.entries(_index.templates))
    }
    buildPrompt(templateName, data) {
        const template = this.templates.get(templateName);
        if (!template) {
            throw new Error(ERROR_MESSAGES.TEMPLATE_NOT_FOUND)
        }
        const system = this.generateMessage(template.system, data.system);
        const user = this.generateMessage(template.user, data.user);
        const prompt = {
            system: system,
            user: user
        };
        return prompt
    }
    generateMessage(promptTemplate, placeholders) {
        if (!placeholders && !promptTemplate) {
            return
        }
        if (!promptTemplate && placeholders) {
            return Object.keys(placeholders).reduce(((acc, key) => `${acc} ${placeholders[key]}`), "").trim()
        }
        if (!placeholders && promptTemplate) {
            return promptTemplate
        }
        const result = this.replacePlaceholders(promptTemplate, placeholders);
        return result
    }
    replacePlaceholders(promptTemplate, placeholders) {
        const result = Object.entries(placeholders).reduce(((acc, _ref) => {
            let [key, value] = _ref;
            return acc.replaceAll(`{{${key}}}`, value)
        }), promptTemplate);
        return result
    }
}
exports.PromptManager = PromptManager;
