// devextreme/cjs/__internal/grids/new/grid_core/inferno_wrappers/popup.js
/**
 * DevExtreme (cjs/__internal/grids/new/grid_core/inferno_wrappers/popup.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Popup = void 0;
var _inferno = require("inferno");
var _popup = _interopRequireDefault(require("../../../../../ui/popup"));
var _utils = require("./utils");
var _widget_wrapper = require("./widget_wrapper");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
class Popup extends _widget_wrapper.InfernoWrapper {
    constructor() {
        super(...arguments);
        this.contentRef = {}
    }
    render() {
        return (0, _inferno.createFragment)([super.render(), this.contentRef.current && (0, _inferno.createPortal)(this.props.children, this.contentRef.current)], 0)
    }
    transformRef(props) {
        var _props;
        if (null !== (_props = props) && void 0 !== _props && null !== (_props = _props.position) && void 0 !== _props && null !== (_props = _props.of) && void 0 !== _props && _props.current) {
            props = Object.assign({}, props, {
                position: Object.assign({}, props.position, {
                    of: (0, _utils.wrapRef)(props.position.of)
                })
            })
        }
        return props
    }
    createComponent(ref, props) {
        return super.createComponent(ref, this.transformRef(props))
    }
    updateComponentOptions(prevProps, props) {
        super.updateComponentOptions(prevProps, this.transformRef(props))
    }
    getComponentFabric() {
        return _popup.default
    }
    componentDidMount() {
        super.componentDidMount();
        this.contentRef.current = this.component.$content().get(0);
        this.setState({})
    }
}
exports.Popup = Popup;
