// devextreme/cjs/__internal/core/environment/hide_callback.js
/**
 * DevExtreme (cjs/__internal/core/environment/hide_callback.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.hideCallback = exports.fireCallback = void 0;
const hideCallback = exports.hideCallback = function() {
    let callbacks = [];
    return {
        add(callback) {
            if (!callbacks.includes(callback)) {
                callbacks.push(callback)
            }
        },
        remove(callback) {
            const indexOfCallback = callbacks.indexOf(callback);
            if (-1 !== indexOfCallback) {
                callbacks.splice(indexOfCallback, 1)
            }
        },
        fire() {
            const callback = callbacks.pop();
            const result = !!callback;
            if (result) {
                callback()
            }
            return result
        },
        hasCallback: () => callbacks.length > 0
    }
}();
const fireCallback = () => hideCallback.fire();
exports.fireCallback = fireCallback;
