// devextreme/cjs/__internal/scheduler/r1/components/wrappers/all_day_table.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/components/wrappers/all_day_table.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.AllDayTableComponent = void 0;
var _component_registrator = _interopRequireDefault(require("../../../../../core/component_registrator"));
var _all_day_panel_table = require("../../../../scheduler/r1/components/base/all_day_panel_table");
var _date_table = require("./date_table");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
class AllDayTableComponent extends _date_table.DateTableComponent {
    get _propsInfo() {
        return {
            twoWay: [],
            allowNull: [],
            elements: [],
            templates: ["dataCellTemplate"],
            props: ["viewData", "viewContext", "groupOrientation", "leftVirtualCellWidth", "rightVirtualCellWidth", "topVirtualRowHeight", "bottomVirtualRowHeight", "addDateTableClass", "addVerticalSizesClassToRows", "width", "dataCellTemplate"]
        }
    }
    get _viewComponent() {
        return _all_day_panel_table.AllDayTable
    }
}
exports.AllDayTableComponent = AllDayTableComponent;
(0, _component_registrator.default)("dxAllDayTable", AllDayTableComponent);
