// devextreme/cjs/__internal/scheduler/r1/components/wrappers/group_panel.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/components/wrappers/group_panel.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.GroupPanelComponent = void 0;
var _component_registrator = _interopRequireDefault(require("../../../../../core/component_registrator"));
var _index = require("../../../../core/r1/index");
var _group_panel = require("../base/group_panel");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
class GroupPanelComponent extends _index.ComponentWrapper {
    _setOptionsByReference() {
        super._setOptionsByReference();
        this._optionsByReference = Object.assign({}, this._optionsByReference, {
            resourceCellTemplate: true
        })
    }
    get _propsInfo() {
        return {
            twoWay: [],
            allowNull: [],
            elements: [],
            templates: ["resourceCellTemplate"],
            props: ["viewContext", "groups", "groupOrientation", "groupPanelData", "groupByDate", "height", "className", "resourceCellTemplate"]
        }
    }
    get _viewComponent() {
        return _group_panel.GroupPanel
    }
}
exports.GroupPanelComponent = GroupPanelComponent;
(0, _component_registrator.default)("dxGroupPanel", GroupPanelComponent);
