// devextreme/cjs/__internal/scheduler/r1/components/wrappers/time_panel.js
/**
 * DevExtreme (cjs/__internal/scheduler/r1/components/wrappers/time_panel.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TimePanelComponent = void 0;
var _component_registrator = _interopRequireDefault(require("../../../../../core/component_registrator"));
var _index = require("../../../../core/r1/index");
var _time_panel_table = require("../base/time_panel_table");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
class TimePanelComponent extends _index.ComponentWrapper {
    _setOptionsByReference() {
        super._setOptionsByReference();
        this._optionsByReference = Object.assign({}, this._optionsByReference, {
            timeCellTemplate: true
        })
    }
    get _propsInfo() {
        return {
            twoWay: [],
            allowNull: [],
            elements: [],
            templates: ["timeCellTemplate"],
            props: ["viewContext", "groupOrientation", "timePanelData", "timeCellTemplate"]
        }
    }
    get _viewComponent() {
        return _time_panel_table.TimePanelTable
    }
}
exports.TimePanelComponent = TimePanelComponent;
(0, _component_registrator.default)("dxTimePanelTableLayout", TimePanelComponent);
