// devextreme/cjs/__internal/ui/file_manager/ui.file_manager.files_tree_view.js
/**
 * DevExtreme (cjs/__internal/ui/file_manager/ui.file_manager.files_tree_view.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;
var _renderer = _interopRequireDefault(require("../../../core/renderer"));
var _common = require("../../../core/utils/common");
var _deferred = require("../../../core/utils/deferred");
var _icon = require("../../../core/utils/icon");
var _type = require("../../../core/utils/type");
var _window = require("../../../core/utils/window");
var _tree_view = _interopRequireDefault(require("../../../ui/tree_view"));
var _widget = _interopRequireDefault(require("../../core/widget/widget"));
var _uiFile_manager = _interopRequireDefault(require("../../ui/file_manager/ui.file_manager.file_actions_button"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const FILE_MANAGER_DIRS_TREE_CLASS = "dx-filemanager-dirs-tree";
const FILE_MANAGER_DIRS_TREE_FOCUSED_ITEM_CLASS = "dx-filemanager-focused-item";
const FILE_MANAGER_DIRS_TREE_ITEM_TEXT_CLASS = "dx-filemanager-dirs-tree-item-text";
const TREE_VIEW_ITEM_CLASS = "dx-treeview-item";
class FileManagerFilesTreeView extends _widget.default {
    _initMarkup() {
        this._initActions();
        const {
            getCurrentDirectory: getCurrentDirectory,
            storeExpandedState: storeExpandedState
        } = this.option();
        this._getCurrentDirectory = getCurrentDirectory;
        this._createFileActionsButton = _common.noop;
        this._storeExpandedState = storeExpandedState || false;
        const $treeView = (0, _renderer.default)("<div>").addClass("dx-filemanager-dirs-tree").appendTo(this.$element());
        const treeViewOptions = {
            dataStructure: "plain",
            rootValue: "",
            createChildren: this._onFilesTreeViewCreateSubDirectories.bind(this),
            itemTemplate: this._createFilesTreeViewItemTemplate.bind(this),
            keyExpr: "getInternalKey",
            parentIdExpr: "parentDirectory.getInternalKey",
            displayExpr: itemInfo => itemInfo.getDisplayName(),
            hasItemsExpr: "fileItem.hasSubDirectories",
            onItemClick: e => {
                var _this$_actions$onDire, _this$_actions;
                return null === (_this$_actions$onDire = (_this$_actions = this._actions).onDirectoryClick) || void 0 === _this$_actions$onDire ? void 0 : _this$_actions$onDire.call(_this$_actions, e)
            },
            onItemExpanded: e => this._onFilesTreeViewItemExpanded(e),
            onItemCollapsed: e => this._onFilesTreeViewItemCollapsed(e),
            onItemRendered: e => this._onFilesTreeViewItemRendered(e),
            onContentReady: () => {
                var _this$_actions$onFile, _this$_actions2;
                return null === (_this$_actions$onFile = (_this$_actions2 = this._actions).onFilesTreeViewContentReady) || void 0 === _this$_actions$onFile ? void 0 : _this$_actions$onFile.call(_this$_actions2)
            }
        };
        if (this._contextMenu) {
            this._contextMenu.option("onContextMenuHidden", (() => this._onContextMenuHidden()));
            treeViewOptions.onItemContextMenu = e => this._onFilesTreeViewItemContextMenu(e);
            this._createFileActionsButton = (element, options) => this._createComponent(element, _uiFile_manager.default, options)
        }
        this._filesTreeView = this._createComponent($treeView, _tree_view.default, treeViewOptions)
    }
    _initActions() {
        this._actions = {
            onDirectoryClick: this._createActionByOption("onDirectoryClick"),
            onFilesTreeViewContentReady: this._createActionByOption("onFilesTreeViewContentReady")
        }
    }
    _render() {
        super._render();
        const that = this;
        setTimeout((() => {
            that._updateFocusedElement()
        }))
    }
    _onFilesTreeViewCreateSubDirectories(rootItem) {
        const {
            getDirectories: getDirectories
        } = this.option();
        const directoryInfo = (null === rootItem || void 0 === rootItem ? void 0 : rootItem.itemData) || null;
        return null === getDirectories || void 0 === getDirectories ? void 0 : getDirectories(directoryInfo, true)
    }
    _onFilesTreeViewItemRendered(_ref) {
        let {
            itemData: itemData
        } = _ref;
        const currentDirectory = this._getCurrentDirectory();
        if (null !== currentDirectory && void 0 !== currentDirectory && currentDirectory.fileItem.equals(itemData.fileItem)) {
            this._updateFocusedElement();
            this._restoreScrollTopPosition()
        }
    }
    _onFilesTreeViewItemExpanded(_ref2) {
        let {
            itemData: itemData
        } = _ref2;
        if (this._storeExpandedState) {
            itemData.expanded = true
        }
    }
    _onFilesTreeViewItemCollapsed(_ref3) {
        let {
            itemData: itemData
        } = _ref3;
        if (this._storeExpandedState) {
            itemData.expanded = false
        }
    }
    _createFilesTreeViewItemTemplate(itemData, itemIndex, itemElement) {
        const $itemElement = (0, _renderer.default)(itemElement);
        const $itemWrapper = $itemElement.closest(this._filesTreeViewItemSelector);
        $itemWrapper.data("item", itemData);
        const $image = (0, _icon.getImageContainer)(itemData.icon);
        const $text = (0, _renderer.default)("<span>").text(itemData.getDisplayName()).addClass("dx-filemanager-dirs-tree-item-text");
        const $button = (0, _renderer.default)("<div>");
        $itemElement.append($image, $text, $button);
        this._createFileActionsButton($button, {
            onClick: e => this._onFileItemActionButtonClick(e)
        })
    }
    _onFilesTreeViewItemContextMenu(_ref4) {
        var _this$_contextMenu;
        let {
            itemElement: itemElement,
            event: event
        } = _ref4;
        event.preventDefault();
        event.stopPropagation();
        const itemData = (0, _renderer.default)(itemElement).data("item");
        null === (_this$_contextMenu = this._contextMenu) || void 0 === _this$_contextMenu || _this$_contextMenu.showAt([itemData], itemElement, event, {
            itemData: itemData,
            itemElement: itemElement
        })
    }
    _onFileItemActionButtonClick(_ref5) {
        var _this$_contextMenu2;
        let {
            component: component,
            element: element,
            event: event
        } = _ref5;
        event.stopPropagation();
        const itemElement = component.$element().closest(this._filesTreeViewItemSelector);
        const itemData = itemElement.data("item");
        const target = {
            itemData: itemData,
            itemElement: itemElement,
            isActionButton: true
        };
        null === (_this$_contextMenu2 = this._contextMenu) || void 0 === _this$_contextMenu2 || _this$_contextMenu2.showAt([itemData], element, event, target);
        this._activeFileActionsButton = component;
        this._activeFileActionsButton.setActive(true)
    }
    _onContextMenuHidden() {
        if (this._activeFileActionsButton) {
            this._activeFileActionsButton.setActive(false)
        }
    }
    toggleNodeDisabledState(key, state) {
        var _this$_filesTreeView;
        const node = this._getNodeByKey(key);
        if (!node) {
            return
        }
        const items = null === (_this$_filesTreeView = this._filesTreeView) || void 0 === _this$_filesTreeView ? void 0 : _this$_filesTreeView.option("items");
        const itemIndex = null === items || void 0 === items ? void 0 : items.map((item => item.getInternalKey())).indexOf(node.getInternalKey());
        if (-1 !== itemIndex) {
            var _this$_filesTreeView2;
            null === (_this$_filesTreeView2 = this._filesTreeView) || void 0 === _this$_filesTreeView2 || _this$_filesTreeView2.option(`items[${itemIndex}].disabled`, state)
        }
    }
    _saveScrollTopPosition() {
        var _this$_filesTreeView3;
        if (!(0, _window.hasWindow)()) {
            return
        }
        this._scrollTopPosition = null === (_this$_filesTreeView3 = this._filesTreeView) || void 0 === _this$_filesTreeView3 ? void 0 : _this$_filesTreeView3.getScrollable().scrollTop()
    }
    _restoreScrollTopPosition() {
        if (!(0, _window.hasWindow)() || !(0, _type.isNumeric)(this._scrollTopPosition)) {
            return
        }
        setTimeout((() => {
            var _this$_filesTreeView4;
            return null === (_this$_filesTreeView4 = this._filesTreeView) || void 0 === _this$_filesTreeView4 ? void 0 : _this$_filesTreeView4.getScrollable().scrollTo(this._scrollTopPosition)
        }))
    }
    _updateFocusedElement() {
        var _this$_$focusedElemen;
        const directoryInfo = this._getCurrentDirectory();
        const $element = this._getItemElementByKey(null === directoryInfo || void 0 === directoryInfo ? void 0 : directoryInfo.getInternalKey());
        if (this._$focusedElement) {
            this._$focusedElement.toggleClass("dx-filemanager-focused-item", false)
        }
        this._$focusedElement = $element || (0, _renderer.default)();
        null === (_this$_$focusedElemen = this._$focusedElement) || void 0 === _this$_$focusedElemen || _this$_$focusedElemen.toggleClass("dx-filemanager-focused-item", true)
    }
    _getNodeByKey(key) {
        var _this$_filesTreeView5;
        return null === (_this$_filesTreeView5 = this._filesTreeView) || void 0 === _this$_filesTreeView5 ? void 0 : _this$_filesTreeView5._getNode(key)
    }
    _getPublicNode(key) {
        var _this$_filesTreeView6;
        const nodesQueue = [...null === (_this$_filesTreeView6 = this._filesTreeView) || void 0 === _this$_filesTreeView6 ? void 0 : _this$_filesTreeView6.getNodes()];
        while (nodesQueue.length) {
            const node = nodesQueue.shift();
            if (node.itemData.getInternalKey() === key) {
                return node
            }
            if (node.children.length) {
                nodesQueue.push(...node.children)
            }
        }
        return
    }
    _getItemElementByKey(key) {
        const node = this._getNodeByKey(key);
        if (node) {
            var _this$_filesTreeView7;
            const $node = null === (_this$_filesTreeView7 = this._filesTreeView) || void 0 === _this$_filesTreeView7 ? void 0 : _this$_filesTreeView7._getNodeElement(node);
            if ($node) {
                return $node.children(this._filesTreeViewItemSelector)
            }
        }
        return null
    }
    _getDefaultOptions() {
        return Object.assign({}, super._getDefaultOptions(), {
            storeExpandedState: false,
            initialFolder: void 0,
            contextMenu: void 0,
            getItems: void 0,
            getCurrentDirectory: void 0,
            onDirectoryClick: void 0
        })
    }
    _optionChanged(args) {
        const {
            name: name
        } = args;
        switch (name) {
            case "storeExpandedState":
                this._storeExpandedState = this.option(name);
                break;
            case "getItems":
            case "rootFolderDisplayName":
            case "initialFolder":
            case "contextMenu":
                this.repaint();
                break;
            case "getCurrentDirectory":
                this.getCurrentDirectory = this.option(name);
                break;
            case "onDirectoryClick":
            case "onFilesTreeViewContentReady":
                this._actions[name] = this._createActionByOption(name);
                break;
            default:
                super._optionChanged(args)
        }
    }
    get _filesTreeViewItemSelector() {
        return ".dx-treeview-item"
    }
    get _contextMenu() {
        const {
            contextMenu: contextMenu
        } = this.option();
        return contextMenu
    }
    toggleDirectoryExpandedState(directoryInfo, state) {
        var _this$_filesTreeView8;
        const deferred = new _deferred.Deferred;
        const treeViewNode = this._getPublicNode(null === directoryInfo || void 0 === directoryInfo ? void 0 : directoryInfo.getInternalKey());
        if (!treeViewNode) {
            return deferred.reject().promise()
        }
        if (treeViewNode.expanded === state || treeViewNode.itemsLoaded && !treeViewNode.itemData.fileItem.hasSubDirectories) {
            return deferred.resolve().promise()
        }
        const action = state ? "expandItem" : "collapseItem";
        return null === (_this$_filesTreeView8 = this._filesTreeView) || void 0 === _this$_filesTreeView8 ? void 0 : _this$_filesTreeView8[action](directoryInfo.getInternalKey())
    }
    refresh() {
        var _this$_filesTreeView9;
        this._$focusedElement = null;
        this._saveScrollTopPosition();
        null === (_this$_filesTreeView9 = this._filesTreeView) || void 0 === _this$_filesTreeView9 || _this$_filesTreeView9.option("dataSource", [])
    }
    updateCurrentDirectory() {
        if (this._disposed) {
            return
        }
        this._updateFocusedElement();
        if (this._storeExpandedState) {
            this._updateExpandedStateToCurrentDirectory()
        }
    }
    _updateExpandedStateToCurrentDirectory() {
        return this.toggleDirectoryExpandedStateRecursive(this._getCurrentDirectory().parentDirectory, true)
    }
    toggleDirectoryExpandedStateRecursive(directoryInfo, state) {
        const dirLine = [];
        for (let dirInfo = directoryInfo; dirInfo; dirInfo = dirInfo.parentDirectory) {
            dirLine.unshift(dirInfo)
        }
        return this.toggleDirectoryLineExpandedState(dirLine, state)
    }
    toggleDirectoryLineExpandedState(dirLine, state) {
        if (!dirLine.length) {
            return (new _deferred.Deferred).resolve().promise()
        }
        return this.toggleDirectoryExpandedState(dirLine.shift(), state).then((() => this.toggleDirectoryLineExpandedState(dirLine, state)))
    }
}
var _default = exports.default = FileManagerFilesTreeView;
