// devextreme/cjs/common/core/animation/translator.js
/**
 * DevExtreme (cjs/common/core/animation/translator.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "clearCache", {
    enumerable: true,
    get: function() {
        return _translator.clearCache
    }
});
Object.defineProperty(exports, "getTranslate", {
    enumerable: true,
    get: function() {
        return _translator.getTranslate
    }
});
Object.defineProperty(exports, "getTranslateCss", {
    enumerable: true,
    get: function() {
        return _translator.getTranslateCss
    }
});
Object.defineProperty(exports, "locate", {
    enumerable: true,
    get: function() {
        return _translator.locate
    }
});
Object.defineProperty(exports, "move", {
    enumerable: true,
    get: function() {
        return _translator.move
    }
});
Object.defineProperty(exports, "parseTranslate", {
    enumerable: true,
    get: function() {
        return _translator.parseTranslate
    }
});
Object.defineProperty(exports, "resetPosition", {
    enumerable: true,
    get: function() {
        return _translator.resetPosition
    }
});
var _translator = require("../../../__internal/common/core/animation/translator");
