// devextreme/cjs/__internal/scheduler/view_model/get_appointment_info.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/get_appointment_info.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getAppointmentInfo = exports.getAgendaAppointmentInfo = void 0;
var _m_utils_time_zone = _interopRequireDefault(require("../m_utils_time_zone"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const getAppointmentInfo = item => {
    const appointment = {
        allDay: item.allDay,
        startDate: _m_utils_time_zone.default.createDateFromUTCWithLocalOffset(new Date(item.datesBeforeSplit.startDateUTC)),
        endDate: _m_utils_time_zone.default.createDateFromUTCWithLocalOffset(new Date(item.datesBeforeSplit.endDateUTC))
    };
    const source = {
        allDay: item.allDay,
        startDate: new Date(item.source.startDate),
        endDate: new Date(item.source.endDate)
    };
    return {
        appointment: appointment,
        sourceAppointment: source
    }
};
exports.getAppointmentInfo = getAppointmentInfo;
const getAgendaAppointmentInfo = item => Object.assign({}, getAppointmentInfo(item), {
    partialDates: {
        allDay: item.allDay,
        startDate: _m_utils_time_zone.default.createDateFromUTCWithLocalOffset(new Date(item.datesAfterSplit.startDateUTC)),
        endDate: _m_utils_time_zone.default.createDateFromUTCWithLocalOffset(new Date(item.datesAfterSplit.endDateUTC))
    }
});
exports.getAgendaAppointmentInfo = getAgendaAppointmentInfo;
