// devextreme/cjs/__internal/scheduler/view_model/filtration/utils/get_filter_options/get_filter_options.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/filtration/utils/get_filter_options/get_filter_options.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getFilterOptions = void 0;
var _get_compare_options = require("../../../common/get_compare_options");
var _shift_intervals = require("../../../common/shift_intervals");
var _get_visible_date_time_intervals = require("./get_visible_date_time_intervals");
const VIEWS_WITH_ALL_DAY_PANEL = ["day", "week", "workWeek"];
const DATE_TIME_VIEWS = ["day", "week", "workWeek", "timelineDay", "timelineWeek", "timelineWorkWeek"];
const getFilterOptions = schedulerStore => {
    const compareOptions = (0, _get_compare_options.getCompareOptions)(schedulerStore);
    const viewOffset = schedulerStore.getViewOffsetMs();
    const viewType = schedulerStore.currentView.type;
    const supportAllDayPanel = VIEWS_WITH_ALL_DAY_PANEL.includes(viewType);
    const isDateTimeView = DATE_TIME_VIEWS.includes(viewType);
    return {
        allDayPanelMode: schedulerStore.getViewOption("allDayPanelMode"),
        supportAllDayPanel: supportAllDayPanel,
        isDateTimeView: isDateTimeView,
        showAllDayPanel: schedulerStore.option("showAllDayPanel"),
        resourceManager: schedulerStore.resourceManager,
        timeZone: schedulerStore.getTimeZone(),
        dataAccessor: schedulerStore._dataAccessors,
        viewOffset: viewOffset,
        firstDayOfWeek: schedulerStore.option("firstDayOfWeek"),
        allDayIntervals: (0, _shift_intervals.shiftIntervals)((0, _get_visible_date_time_intervals.getVisibleDateTimeIntervals)(compareOptions, true), viewOffset),
        regularIntervals: (0, _shift_intervals.shiftIntervals)((0, _get_visible_date_time_intervals.getVisibleDateTimeIntervals)(compareOptions, false), viewOffset)
    }
};
exports.getFilterOptions = getFilterOptions;
