// devextreme/esm/__internal/ui/__tests__/__mock__/model/tree_view.js
/**
 * DevExtreme (esm/__internal/ui/__tests__/__mock__/model/tree_view.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import TreeView from "../../../../../ui/tree_view";
import {
    CheckBoxModel
} from "./checkbox";
import {
    TextBoxModel
} from "./textbox";
const CLASSES = {
    treeView: "dx-treeview",
    searchBox: "dx-treeview-search",
    node: "dx-treeview-node",
    item: "dx-treeview-item",
    checkbox: "dx-checkbox"
};
export class TreeViewModel {
    constructor(root) {
        this.root = root
    }
    getInstance() {
        return TreeView.getInstance(this.root)
    }
    getSearchBox() {
        var _this$root;
        return new TextBoxModel(null === (_this$root = this.root) || void 0 === _this$root ? void 0 : _this$root.querySelector(`.${CLASSES.searchBox}`))
    }
    setSearchValue(value) {
        const searchBox = this.getSearchBox();
        searchBox.setValue(value)
    }
    getNodes() {
        var _this$root2;
        return (null === (_this$root2 = this.root) || void 0 === _this$root2 ? void 0 : _this$root2.querySelectorAll(`.${CLASSES.node}`)) ?? null
    }
    getNodeByText(text) {
        const nodes = this.getNodes();
        if (!nodes) {
            return null
        }
        const foundNode = Array.from(nodes).find((node => {
            const itemElement = node.querySelector(`.${CLASSES.item}`);
            const nodeText = null === itemElement || void 0 === itemElement ? void 0 : itemElement.textContent;
            return null === nodeText || void 0 === nodeText ? void 0 : nodeText.includes(text)
        })) ?? null;
        return foundNode
    }
    getCheckboxByText(text) {
        const node = this.getNodeByText(text);
        const checkboxElement = null === node || void 0 === node ? void 0 : node.querySelector(`.${CLASSES.checkbox}`);
        return checkboxElement ? new CheckBoxModel(checkboxElement) : null
    }
}
