// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/generate_grid_view_model.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/generate_grid_view_model.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.generateGridViewModel = void 0;
var _option_manager = require("./options/option_manager");
var _add_collector = require("./steps/add_collector/add_collector");
var _add_direction = require("./steps/add_direction");
var _add_emptiness = require("./steps/add_emptiness");
var _add_geometry = require("./steps/add_geometry/add_geometry");
var _add_position = require("./steps/add_position");
var _add_sorted_index = require("./steps/add_sorted_index");
var _expand_all_day = require("./steps/expand_all_day");
var _group_by_group_index = require("./steps/group_by_group_index");
var _maybe_split = require("./steps/maybe_split");
var _snap_to_cells = require("./steps/snap_to_cells");
var _sorting = require("./steps/sorting");
var _split_by_parts = require("./steps/split_by_parts/split_by_parts");
var _virtual_screen_crop = require("./steps/virtual_screen_crop");
var _virtual_screen_filter = require("./steps/virtual_screen_filter");
const generateGridViewModel = (schedulerStore, items) => {
    const optionManager = new _option_manager.OptionManager(schedulerStore);
    const {
        viewOrientation: viewOrientation,
        isMonthView: isMonthView,
        isAdaptivityEnabled: isAdaptivityEnabled,
        isTimelineView: isTimelineView,
        hasAllDayPanel: hasAllDayPanel,
        isVirtualScrolling: isVirtualScrolling,
        viewOffset: viewOffset,
        compareOptions: {
            endDayHour: endDayHour
        }
    } = optionManager.options;
    const {
        viewDataProvider: viewDataProvider
    } = schedulerStore._workSpace;
    const step2 = (0, _maybe_split.maybeSplit)(items, hasAllDayPanel, ((entities, panelName) => {
        const byGroup = (0, _group_by_group_index.groupByGroupIndex)(entities);
        const positionInsideGroup = byGroup.map((group => {
            (0, _sorting.sortByDuration)(group);
            (0, _sorting.sortByStartDate)(group);
            const innerStep0 = isMonthView || "allDayPanel" === panelName ? (0, _expand_all_day.expandAllDayAllDayPanel)(group, endDayHour, viewOffset) : (0, _expand_all_day.expandAllDayRegularPanel)(group);
            const innerStep1 = (0, _split_by_parts.splitByParts)(innerStep0, optionManager.getSplitIntervals(panelName));
            (0, _sorting.sortByDuration)(innerStep1);
            (0, _sorting.sortByStartDate)(innerStep1);
            (0, _sorting.sortByGroupIndex)(innerStep1);
            const innerStep2 = (0, _add_position.addPosition)(innerStep1, optionManager.getCells(panelName));
            const innerStep3 = isMonthView || "allDayPanel" === panelName ? (0, _snap_to_cells.snapToCells)(innerStep2, optionManager.getCells(panelName)) : innerStep2;
            const innerStep4 = (0, _add_collector.addCollector)(innerStep3, optionManager.getCollectorOptions(panelName));
            return innerStep4
        }));
        return positionInsideGroup.flat()
    }));
    const step3 = (0, _add_sorted_index.addSortedIndex)(step2);
    const step4 = (0, _virtual_screen_filter.filterByVirtualScreen)(step3, viewDataProvider, isVirtualScrolling);
    const step5 = (0, _maybe_split.maybeSplit)(step4, hasAllDayPanel, ((entities, panelName) => {
        const innerStep = (0, _add_geometry.addGeometry)(entities, optionManager.getGeometryOptions(panelName));
        return innerStep
    }));
    const step6 = (0, _virtual_screen_crop.cropByVirtualScreen)(step5, optionManager.getVirtualCropOptions());
    const step7 = (0, _add_direction.addDirection)(step6, "horizontal", viewOrientation);
    const step8 = (0, _add_emptiness.addEmptiness)(step7, {
        isTimelineView: isTimelineView,
        isAdaptivityEnabled: isAdaptivityEnabled,
        isMonthView: isMonthView
    });
    return step8
};
exports.generateGridViewModel = generateGridViewModel;
