// devextreme/cjs/__internal/ui/__tests__/__mock__/model/tree_view.js
/**
 * DevExtreme (cjs/__internal/ui/__tests__/__mock__/model/tree_view.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.TreeViewModel = void 0;
var _tree_view = _interopRequireDefault(require("../../../../../ui/tree_view"));
var _checkbox = require("./checkbox");
var _textbox = require("./textbox");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const CLASSES = {
    treeView: "dx-treeview",
    searchBox: "dx-treeview-search",
    node: "dx-treeview-node",
    item: "dx-treeview-item",
    checkbox: "dx-checkbox"
};
class TreeViewModel {
    constructor(root) {
        this.root = root
    }
    getInstance() {
        return _tree_view.default.getInstance(this.root)
    }
    getSearchBox() {
        var _this$root;
        return new _textbox.TextBoxModel(null === (_this$root = this.root) || void 0 === _this$root ? void 0 : _this$root.querySelector(`.${CLASSES.searchBox}`))
    }
    setSearchValue(value) {
        const searchBox = this.getSearchBox();
        searchBox.setValue(value)
    }
    getNodes() {
        var _this$root2;
        return (null === (_this$root2 = this.root) || void 0 === _this$root2 ? void 0 : _this$root2.querySelectorAll(`.${CLASSES.node}`)) ?? null
    }
    getNodeByText(text) {
        const nodes = this.getNodes();
        if (!nodes) {
            return null
        }
        const foundNode = Array.from(nodes).find((node => {
            const itemElement = node.querySelector(`.${CLASSES.item}`);
            const nodeText = null === itemElement || void 0 === itemElement ? void 0 : itemElement.textContent;
            return null === nodeText || void 0 === nodeText ? void 0 : nodeText.includes(text)
        })) ?? null;
        return foundNode
    }
    getCheckboxByText(text) {
        const node = this.getNodeByText(text);
        const checkboxElement = null === node || void 0 === node ? void 0 : node.querySelector(`.${CLASSES.checkbox}`);
        return checkboxElement ? new _checkbox.CheckBoxModel(checkboxElement) : null
    }
}
exports.TreeViewModel = TreeViewModel;
