// devextreme/cjs/viz/palette.js
/**
 * DevExtreme (cjs/viz/palette.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "createPalette", {
    enumerable: true,
    get: function() {
        return _palette.createPalette
    }
});
Object.defineProperty(exports, "currentPalette", {
    enumerable: true,
    get: function() {
        return _palette.currentPalette
    }
});
Object.defineProperty(exports, "generateColors", {
    enumerable: true,
    get: function() {
        return _palette.generateColors
    }
});
Object.defineProperty(exports, "getAccentColor", {
    enumerable: true,
    get: function() {
        return _palette.getAccentColor
    }
});
Object.defineProperty(exports, "getDiscretePalette", {
    enumerable: true,
    get: function() {
        return _palette.getDiscretePalette
    }
});
Object.defineProperty(exports, "getGradientPalette", {
    enumerable: true,
    get: function() {
        return _palette.getGradientPalette
    }
});
Object.defineProperty(exports, "getPalette", {
    enumerable: true,
    get: function() {
        return _palette.getPalette
    }
});
Object.defineProperty(exports, "registerPalette", {
    enumerable: true,
    get: function() {
        return _palette.registerPalette
    }
});
var _palette = require("../__internal/viz/palette");
