// devextreme/cjs/__internal/core/state_manager/prod/index.js
/**
 * DevExtreme (cjs/__internal/core/state_manager/prod/index.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "batch", {
    enumerable: true,
    get: function() {
        return _index.batch
    }
});
Object.defineProperty(exports, "computed", {
    enumerable: true,
    get: function() {
        return _index.computed
    }
});
Object.defineProperty(exports, "effect", {
    enumerable: true,
    get: function() {
        return _index.effect
    }
});
exports.setupStateManager = void 0;
Object.defineProperty(exports, "signal", {
    enumerable: true,
    get: function() {
        return _index.signal
    }
});
Object.defineProperty(exports, "untracked", {
    enumerable: true,
    get: function() {
        return _index.untracked
    }
});
var _index = require("./reactive_primitives/index");
const setupStateManager = () => {};
exports.setupStateManager = setupStateManager;
