// devextreme/cjs/__internal/grids/grid_core/context_menu/m_column_context_menu_mixin.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/context_menu/m_column_context_menu_mixin.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ColumnContextMenuMixin = void 0;
var _message = _interopRequireDefault(require("../../../../common/core/localization/message"));
var _const = require("../keyboard_navigation/const");
var _const2 = require("./const");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const ColumnContextMenuMixin = Base => class extends Base {
    constructor() {
        super(...arguments);
        this.isNeedToFocusColumn = false
    }
    getMoveColumnContextMenuItems(options) {
        var _this$isColumnReorder, _this$getKeyboardNavi;
        const {
            column: column,
            rowIndex: rowIndex
        } = options;
        const allowColumnReordering = null === (_this$isColumnReorder = this.isColumnReorderingEnabled) || void 0 === _this$isColumnReorder ? void 0 : _this$isColumnReorder.call(this, null === options || void 0 === options ? void 0 : options.column);
        const keyboardNavigationController = null === (_this$getKeyboardNavi = this.getKeyboardNavigationController) || void 0 === _this$getKeyboardNavi ? void 0 : _this$getKeyboardNavi.call(this);
        if (!allowColumnReordering || !keyboardNavigationController) {
            return []
        }
        const rtlEnabled = this.option("rtlEnabled");
        const onItemClick = e => {
            var _e$itemData;
            this.isNeedToFocusColumn = true;
            keyboardNavigationController.moveColumn(column, null === (_e$itemData = e.itemData) || void 0 === _e$itemData ? void 0 : _e$itemData.value, rowIndex)
        };
        return [{
            text: rtlEnabled ? _message.default.format("dxDataGrid-moveColumnToTheRight") : _message.default.format("dxDataGrid-moveColumnToTheLeft"),
            value: _const.Direction.Previous,
            beginGroup: true,
            disabled: !keyboardNavigationController.canReorderColumn(column, _const.Direction.Previous, rowIndex),
            icon: rtlEnabled ? _const2.CONTEXT_MENU_MOVE_NEXT_ICON_NAME : _const2.CONTEXT_MENU_MOVE_PREVIOUS_ICON_NAME,
            onItemClick: onItemClick
        }, {
            text: rtlEnabled ? _message.default.format("dxDataGrid-moveColumnToTheLeft") : _message.default.format("dxDataGrid-moveColumnToTheRight"),
            value: _const.Direction.Next,
            disabled: !keyboardNavigationController.canReorderColumn(column, _const.Direction.Next, rowIndex),
            icon: rtlEnabled ? _const2.CONTEXT_MENU_MOVE_PREVIOUS_ICON_NAME : _const2.CONTEXT_MENU_MOVE_NEXT_ICON_NAME,
            onItemClick: onItemClick
        }]
    }
};
exports.ColumnContextMenuMixin = ColumnContextMenuMixin;
