// devextreme/cjs/ui/shared/accessibility.js
/**
 * DevExtreme (cjs/ui/shared/accessibility.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "hiddenFocus", {
    enumerable: true,
    get: function() {
        return _accessibility.hiddenFocus
    }
});
Object.defineProperty(exports, "registerKeyboardAction", {
    enumerable: true,
    get: function() {
        return _accessibility.registerKeyboardAction
    }
});
Object.defineProperty(exports, "restoreFocus", {
    enumerable: true,
    get: function() {
        return _accessibility.restoreFocus
    }
});
Object.defineProperty(exports, "saveFocusedElementInfo", {
    enumerable: true,
    get: function() {
        return _accessibility.saveFocusedElementInfo
    }
});
Object.defineProperty(exports, "selectView", {
    enumerable: true,
    get: function() {
        return _accessibility.selectView
    }
});
Object.defineProperty(exports, "setTabIndex", {
    enumerable: true,
    get: function() {
        return _accessibility.setTabIndex
    }
});
Object.defineProperty(exports, "subscribeVisibilityChange", {
    enumerable: true,
    get: function() {
        return _accessibility.subscribeVisibilityChange
    }
});
Object.defineProperty(exports, "unsubscribeVisibilityChange", {
    enumerable: true,
    get: function() {
        return _accessibility.unsubscribeVisibilityChange
    }
});
var _accessibility = require("../../__internal/ui/shared/accessibility");
