// devextreme/esm/__internal/scheduler/view_model/generate_view_model/options/get_minutes_cell_intervals.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/options/get_minutes_cell_intervals.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    splitIntervalByDay
} from "../../common/split_interval_by_days";
const filterBySkippedDays = (intervals, skippedDays) => intervals.filter((item => !skippedDays.includes(new Date(item.min).getUTCDay())));
export const getMinutesCellIntervals = _ref => {
    let {
        intervals: intervals,
        startDayHour: startDayHour,
        endDayHour: endDayHour,
        durationMinutes: durationMinutes,
        skippedDays: skippedDays
    } = _ref;
    return intervals.reduce(((result, interval, rowIndex) => {
        const dayIntervals = splitIntervalByDay(Object.assign({}, interval, {
            startDayHour: startDayHour,
            endDayHour: endDayHour,
            skippedDays: skippedDays
        }));
        let columnIndex = 0;
        filterBySkippedDays(dayIntervals, skippedDays).forEach((dayInterval => {
            const date = new Date(dayInterval.min);
            while (date.getTime() < dayInterval.max) {
                const min = date.getTime();
                let max = date.setUTCMinutes(date.getUTCMinutes() + durationMinutes);
                if (date.getUTCHours() > endDayHour) {
                    date.setUTCDate(date.getUTCDate() + 1);
                    date.setUTCHours(startDayHour, 0, 0, 0);
                    max = date.getTime()
                }
                result.push({
                    min: min,
                    max: max,
                    rowIndex: rowIndex,
                    columnIndex: columnIndex,
                    cellIndex: result.length
                });
                columnIndex += 1
            }
        }));
        return result
    }), [])
};
