// devextreme/esm/__internal/scheduler/view_model/generate_view_model/options/get_group_size.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/options/get_group_size.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
export const getGroupSize = _ref => {
    let {
        cellSize: cellSize,
        cellDurationMinutes: cellDurationMinutes,
        endDayHour: endDayHour,
        startDayHour: startDayHour,
        cells: cells,
        intervals: intervals,
        viewType: viewType,
        isAllDayPanel: isAllDayPanel
    } = _ref;
    switch (viewType) {
        case "month":
        case "timelineMonth": {
            const intervalDaysCount = cells.filter((cell => 0 === cell.rowIndex)).length;
            return {
                width: cellSize.width * intervalDaysCount,
                height: cellSize.height * intervals.length
            }
        }
        case "timelineDay":
        case "timelineWeek":
        case "timelineWorkWeek":
            return {
                width: cellSize.width * cells.length, height: cellSize.height
            };
        case "day":
        case "week":
        case "workWeek":
            return {
                width: isAllDayPanel ? cellSize.width * cells.length : cellSize.width * intervals.length, height: cellSize.height * (endDayHour - startDayHour) * (60 / cellDurationMinutes)
            };
        default:
            return cellSize
    }
};
