// devextreme/esm/__internal/scheduler/view_model/generate_view_model/options/get_max_level.js
/**
 * DevExtreme (esm/__internal/scheduler/view_model/generate_view_model/options/get_max_level.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    getAbstractSizeByViewOrientation
} from "../steps/add_geometry/swap_by_view_orientation";
import {
    getDefaultAppointmentSize
} from "./get_min_appointment_size";
const ADAPTIVITY_MIN_APPOINTMENT_COUNT = 0;
const MIN_APPOINTMENT_COUNT = 1;
const CELL_MIN_Y = 30;
export const getMaxLevel = _ref => {
    let {
        maxAppointmentsPerCell: maxAppointmentsPerCell,
        cellSize: cellSize,
        collectorSize: collectorSize,
        viewOrientation: viewOrientation,
        isTimelineView: isTimelineView,
        isAdaptivityEnabled: isAdaptivityEnabled
    } = _ref;
    switch (maxAppointmentsPerCell) {
        case "auto": {
            if (isAdaptivityEnabled && "horizontal" === viewOrientation) {
                return 0
            }
            const cellSizeY = getAbstractSizeByViewOrientation(cellSize, viewOrientation).sizeY;
            const isSmallCell = cellSizeY < 30;
            if (isSmallCell) {
                return 0
            }
            const defaultAppointmentSize = getDefaultAppointmentSize({
                isTimelineView: isTimelineView,
                isAdaptivityEnabled: isAdaptivityEnabled,
                viewOrientation: viewOrientation
            });
            const minAbstractSize = getAbstractSizeByViewOrientation(defaultAppointmentSize, viewOrientation);
            const collectorSizeY = getAbstractSizeByViewOrientation(collectorSize, viewOrientation).sizeY;
            const calculated = Math.floor(Math.max(0, cellSizeY - collectorSizeY) / minAbstractSize.sizeY);
            return Math.max(calculated, isAdaptivityEnabled ? 0 : 1)
        }
        case "unlimited":
            return -1;
        default:
            return parseInt(String(maxAppointmentsPerCell), 10)
    }
};
