// devextreme/esm/__internal/ui/text_box/texteditor_button_collection/m_button.js
/**
 * DevExtreme (esm/__internal/ui/text_box/texteditor_button_collection/m_button.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import $ from "../../../../core/renderer";
import Button from "../../../../ui/button";
export const isButtonInstance = instance => instance instanceof Button;
export default class TextEditorButton {
    constructor(name, editor, options) {
        this.instance = null;
        this.$container = null;
        this.$placeMarker = null;
        this.editor = editor;
        this.name = name;
        this.options = options || {}
    }
    _addPlaceMarker($container) {
        this.$placeMarker = $("<div>").appendTo($container)
    }
    _addToContainer($element) {
        const {
            $placeMarker: $placeMarker,
            $container: $container
        } = this;
        if ($placeMarker) {
            $placeMarker.replaceWith($element)
        } else {
            $element.appendTo($container)
        }
    }
    _attachEvents(instance, $element) {
        throw "Not implemented"
    }
    _create() {
        throw "Not implemented"
    }
    _isRendered() {
        return !!this.instance
    }
    _isVisible() {
        const {
            editor: editor,
            options: options
        } = this;
        return options.visible || !(null !== editor && void 0 !== editor && editor.option("readOnly"))
    }
    _isDisabled() {
        throw "Not implemented"
    }
    _shouldRender() {
        return this._isVisible() && !this._isRendered()
    }
    dispose() {
        var _this$$placeMarker;
        const {
            instance: instance
        } = this;
        if (instance) {
            if (isButtonInstance(instance)) {
                instance.dispose();
                instance.$element().remove();
                instance._$element = null
            } else {
                instance.remove()
            }
        }
        this.instance = null;
        this.editor = null;
        this.$container = null;
        null === (_this$$placeMarker = this.$placeMarker) || void 0 === _this$$placeMarker || _this$$placeMarker.remove();
        this.$placeMarker = null
    }
    render() {
        let $container = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : this.$container;
        this.$container = $container;
        if (this._isVisible()) {
            const {
                instance: instance,
                $element: $element
            } = this._create() ?? {};
            this.instance = instance;
            this._attachEvents(instance, $element)
        } else {
            this._addPlaceMarker($container)
        }
    }
    update() {
        if (this._shouldRender()) {
            this.render()
        }
        return !!this.instance
    }
}
