// devextreme/cjs/__internal/grids/new/grid_core/keyboard_navigation/with_key_down_handler.js
/**
 * DevExtreme (cjs/__internal/grids/new/grid_core/keyboard_navigation/with_key_down_handler.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.withKeyDownHandler = void 0;
var _inferno = require("inferno");
var _index = require("../../../../grids/new/grid_core/core/events/index");
var _utils = require("./utils");
const _excluded = ["onKeyDown", "keyDownConfig", "children"];

function _objectWithoutPropertiesLoose(r, e) {
    if (null == r) {
        return {}
    }
    var t = {};
    for (var n in r) {
        if ({}.hasOwnProperty.call(r, n)) {
            if (e.includes(n)) {
                continue
            }
            t[n] = r[n]
        }
    }
    return t
}
var __decorate = (void 0, function(decorators, target, key, desc) {
    var d, c = arguments.length,
        r = c < 3 ? target : null === desc ? desc = Object.getOwnPropertyDescriptor(target, key) : desc;
    if ("object" === typeof Reflect && "function" === typeof Reflect.decorate) {
        r = Reflect.decorate(decorators, target, key, desc)
    } else {
        for (var i = decorators.length - 1; i >= 0; i--) {
            if (d = decorators[i]) {
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r
            }
        }
    }
    return c > 3 && r && Object.defineProperty(target, key, r), r
});
var __metadata = (void 0, function(k, v) {
    if ("object" === typeof Reflect && "function" === typeof Reflect.metadata) {
        return Reflect.metadata(k, v)
    }
});
const withKeyDownHandler = WrappedComponent => {
    class WithKeyDownHandler extends _inferno.Component {
        constructor() {
            super(...arguments);
            this.elementRef = (0, _inferno.createRef)()
        }
        render() {
            const _this$props = this.props,
                {
                    children: children
                } = _this$props,
                restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
            return (0, _inferno.normalizeProps)((0, _inferno.createComponentVNode)(2, WrappedComponent, Object.assign({}, restProps, {
                onKeyDown: this.onKeyDown.bind(this),
                children: children
            })))
        }
        onKeyDown(event) {
            const {
                keyDownConfig: keyDownConfig,
                onKeyDown: onKeyDown,
                caughtEventPreventDefault: caughtEventPreventDefault
            } = this.props;
            const ref = this.getActualRef();
            const fullKeyName = (0, _utils.getKeyWithModifications)(event);
            const handler = null === keyDownConfig || void 0 === keyDownConfig ? void 0 : keyDownConfig[fullKeyName];
            if (handler) {
                handler(event, ref);
                _index.eventUtils.markHandled(event)
            }
            if (handler && caughtEventPreventDefault) {
                event.preventDefault()
            }
            null === onKeyDown || void 0 === onKeyDown || onKeyDown(event)
        }
        getActualRef() {
            return this.props.elementRef ?? this.elementRef
        }
    }
    __decorate([_index.eventHandler, __metadata("design:type", Function), __metadata("design:paramtypes", [Object]), __metadata("design:returntype", void 0)], WithKeyDownHandler.prototype, "onKeyDown", null);
    return WithKeyDownHandler
};
exports.withKeyDownHandler = withKeyDownHandler;
