// devextreme/esm/__internal/grids/grid_core/ai_column/controllers/m_ai_column_cache_controller.js
/**
 * DevExtreme (esm/__internal/grids/grid_core/ai_column/controllers/m_ai_column_cache_controller.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    Controller
} from "../../m_modules";
export class AIColumnCacheController extends Controller {
    constructor() {
        super(...arguments);
        this.cache = {}
    }
    clearCache(columnName) {
        this.cache[columnName] = void 0
    }
    clearCacheByKey(columnName, key) {
        const columnCache = this.cache[columnName];
        if (columnCache) {
            columnCache[key] = void 0
        }
    }
    getCachedResponse(columnName, keys) {
        const columnCache = this.cache[columnName];
        if (!columnCache) {
            return {}
        }
        return keys.reduce(((acc, key) => {
            const value = columnCache[key];
            if (void 0 !== value && "" !== value) {
                acc[key] = value
            }
            return acc
        }), {})
    }
    setCachedResponse(columnName, data) {
        const columnCache = this.cache[columnName] ?? {};
        this.cache[columnName] = columnCache;
        Object.entries(data).forEach((_ref => {
            let [key, value] = _ref;
            columnCache[key] = value
        }))
    }
    getCachedString(columnName, key) {
        var _this$cache$columnNam;
        return null === (_this$cache$columnNam = this.cache[columnName]) || void 0 === _this$cache$columnNam ? void 0 : _this$cache$columnNam[key]
    }
    isEmptyCache(columnName) {
        return 0 === Object.keys(this.cache[columnName] ?? {}).length
    }
}
