// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_adaptivity_geometry_inside_interval.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_adaptivity_geometry_inside_interval.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.addAdaptivityGeometryInsideInterval = void 0;
var _swap_by_view_orientation = require("./swap_by_view_orientation");
const COLLECTOR_ADAPTIVE_BOTTOM_OFFSET = 40;
const addAdaptivityGeometryInsideInterval = (entity, _ref) => {
    let {
        cellSize: cellSize,
        collectorSize: collectorSize,
        collectorWithMarginsSize: collectorWithMarginsSize,
        viewOrientation: viewOrientation
    } = _ref;
    const cellAbstractSize = (0, _swap_by_view_orientation.getAbstractSizeByViewOrientation)(cellSize, viewOrientation);
    const topInsideCell = entity.isAllDayPanelOccupied || "vertical" === viewOrientation ? (cellSize.height - collectorWithMarginsSize.height) / 2 : cellSize.height - 40;
    const leftInsideCell = (cellSize.width - collectorWithMarginsSize.width) / 2;
    const abstractGeometry = (0, _swap_by_view_orientation.getAbstractSizeByViewOrientation)({
        top: topInsideCell,
        left: leftInsideCell,
        width: collectorSize.width,
        height: collectorSize.height
    }, viewOrientation);
    abstractGeometry.offsetX += entity.columnIndex * cellAbstractSize.sizeX;
    const geometry = (0, _swap_by_view_orientation.getRealSizeByViewOrientation)(abstractGeometry, viewOrientation);
    const items = entity.items.map((item => Object.assign({}, item, {
        width: cellSize.width,
        height: cellSize.height
    })));
    return Object.assign({}, entity, geometry, {
        items: items
    })
};
exports.addAdaptivityGeometryInsideInterval = addAdaptivityGeometryInsideInterval;
