// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_grouping_offset.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_grouping_offset.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.addGroupingOffset = void 0;
const addGroupingOffset = (entity, _ref) => {
    let {
        groupCount: groupCount,
        groupOrientation: groupOrientation,
        viewOrientation: viewOrientation,
        hasAllDayPanel: hasAllDayPanel,
        isGroupByDate: isGroupByDate,
        allDayPanelCellSize: allDayPanelCellSize,
        cellSize: cellSize,
        groupSize: groupSize
    } = _ref;
    if (groupCount) {
        switch (true) {
            case "horizontal" === groupOrientation && isGroupByDate:
                entity.left += (groupCount - 1) * cellSize.width * ("horizontal" === viewOrientation ? entity.columnIndex : entity.rowIndex) + cellSize.width * entity.groupIndex;
                break;
            case "horizontal" === groupOrientation:
                entity.left += entity.groupIndex * groupSize.width;
                break;
            default:
                entity.top += entity.groupIndex * groupSize.height + (entity.groupIndex + Number(!entity.isAllDayPanelOccupied)) * Number(hasAllDayPanel) * allDayPanelCellSize.height
        }
    }
};
exports.addGroupingOffset = addGroupingOffset;
