// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_geometry.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/add_geometry.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.addGeometry = void 0;
var _add_adaptivity_geometry_inside_interval = require("./add_adaptivity_geometry_inside_interval");
var _add_geometry_inside_interval = require("./add_geometry_inside_interval");
var _add_grouping_offset = require("./add_grouping_offset");
const RTLSwap = (entity, _ref) => {
    let {
        panelSize: panelSize,
        isRTLEnabled: isRTLEnabled
    } = _ref;
    if (isRTLEnabled) {
        const deltaWidth = entity.items.length ? 0 : entity.width;
        entity.left = panelSize.width - entity.left - deltaWidth
    }
};
const addPanelOffset = (entity, _ref2) => {
    let {
        cellSize: cellSize,
        viewOrientation: viewOrientation,
        isTimelineView: isTimelineView
    } = _ref2;
    switch (true) {
        case "horizontal" === viewOrientation && !isTimelineView:
            entity.top += entity.rowIndex * cellSize.height;
            break;
        case "horizontal" === viewOrientation && isTimelineView:
        case "vertical" === viewOrientation:
            entity.left += entity.rowIndex * cellSize.width
    }
};
const addGeometry = (entities, options) => entities.map((rawEntity => {
    const {
        isAdaptivityEnabled: isAdaptivityEnabled,
        maxAppointmentsPerCell: maxAppointmentsPerCell
    } = options;
    const entity = isAdaptivityEnabled && 0 === maxAppointmentsPerCell ? (0, _add_adaptivity_geometry_inside_interval.addAdaptivityGeometryInsideInterval)(rawEntity, options) : (0, _add_geometry_inside_interval.addGeometryInsideInterval)(rawEntity, options);
    addPanelOffset(entity, options);
    (0, _add_grouping_offset.addGroupingOffset)(entity, options);
    RTLSwap(entity, options);
    return entity
}));
exports.addGeometry = addGeometry;
