// devextreme/cjs/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/get_appointment_geometry.js
/**
 * DevExtreme (cjs/__internal/scheduler/view_model/generate_view_model/steps/add_geometry/get_appointment_geometry.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getAppointmentGeometry = void 0;
var _get_appointment_abstract_geometry = require("./get_appointment_abstract_geometry");
var _swap_by_view_orientation = require("./swap_by_view_orientation");
const getAppointmentGeometry = (entity, _ref) => {
    let {
        collectorPosition: collectorPosition,
        cellSize: cellSize,
        collectorWithMarginsSize: collectorWithMarginsSize,
        viewOrientation: viewOrientation,
        cells: cells
    } = _ref;
    const cellAbstractSize = (0, _swap_by_view_orientation.getAbstractSizeByViewOrientation)(cellSize, viewOrientation);
    const collectorFullAbstractSize = (0, _swap_by_view_orientation.getAbstractSizeByViewOrientation)(collectorWithMarginsSize, viewOrientation);
    const abstractGeometry = Object.assign({}, (0, _get_appointment_abstract_geometry.getAppointmentX)(entity, cellAbstractSize, cells), (0, _get_appointment_abstract_geometry.getAppointmentY)(entity, cellAbstractSize, collectorFullAbstractSize.sizeY, collectorPosition));
    const entityGeometry = (0, _swap_by_view_orientation.getRealSizeByViewOrientation)(abstractGeometry, viewOrientation);
    return entityGeometry
};
exports.getAppointmentGeometry = getAppointmentGeometry;
