// devextreme/cjs/__internal/scheduler/utils/get_targeted_appointment.js
/**
 * DevExtreme (cjs/__internal/scheduler/utils/get_targeted_appointment.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getTargetedAppointmentFromInfo = exports.getTargetedAppointment = void 0;
var _appointment_groups_utils = require("./resource_manager/appointment_groups_utils");
var _group_utils = require("./resource_manager/group_utils");
const setTargetedAppointmentResources = (rawAppointment, settings, resourceManager) => {
    const {
        groups: groups,
        resourceById: resourceById,
        groupsLeafs: groupsLeafs
    } = resourceManager;
    if (groups.length) {
        const cellGroups = (0, _group_utils.getLeafGroupValues)(groupsLeafs, settings.groupIndex);
        (0, _appointment_groups_utils.setAppointmentGroupValues)(rawAppointment, resourceById, cellGroups)
    }
};
const getTargetedAppointmentFromInfo = function(rawAppointment, settings, dataAccessor, resourceManager) {
    let usePartialDates = arguments.length > 4 && void 0 !== arguments[4] ? arguments[4] : false;
    const {
        info: info
    } = settings;
    const rawTargetedAppointment = Object.assign({}, rawAppointment);
    dataAccessor.set("startDate", rawTargetedAppointment, new Date(info.sourceAppointment.startDate));
    dataAccessor.set("endDate", rawTargetedAppointment, new Date(info.sourceAppointment.endDate));
    const displayDates = usePartialDates && "partialDates" in info ? info.partialDates : info.appointment;
    rawTargetedAppointment.displayStartDate = new Date(displayDates.startDate);
    rawTargetedAppointment.displayEndDate = new Date(displayDates.endDate);
    setTargetedAppointmentResources(rawTargetedAppointment, settings, resourceManager);
    return rawTargetedAppointment
};
exports.getTargetedAppointmentFromInfo = getTargetedAppointmentFromInfo;
const getTargetedAppointment = (rawAppointment, settings, dataAccessor, resourceManager) => {
    const startDate = dataAccessor.get("startDate", rawAppointment);
    const endDate = dataAccessor.get("endDate", rawAppointment);
    if (!("info" in settings)) {
        return Object.assign({}, rawAppointment, {
            displayStartDate: startDate,
            displayEndDate: endDate
        })
    }
    return getTargetedAppointmentFromInfo(rawAppointment, settings, dataAccessor, resourceManager)
};
exports.getTargetedAppointment = getTargetedAppointment;
