// devextreme/cjs/__internal/core/localization.js
/**
 * DevExtreme (cjs/__internal/core/localization.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "date", {
    enumerable: true,
    get: function() {
        return _date.default
    }
});
exports.disableIntl = disableIntl;
exports.locale = exports.loadMessages = exports.formatNumber = exports.formatMessage = exports.formatDate = void 0;
Object.defineProperty(exports, "message", {
    enumerable: true,
    get: function() {
        return _message.default
    }
});
Object.defineProperty(exports, "number", {
    enumerable: true,
    get: function() {
        return _number.default
    }
});
exports.parseNumber = exports.parseDate = void 0;
require("../core/localization/currency");
var _core = _interopRequireDefault(require("../core/localization/core"));
var _date = _interopRequireDefault(require("../core/localization/date"));
var _message = _interopRequireDefault(require("../core/localization/message"));
var _number = _interopRequireDefault(require("../core/localization/number"));

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
const locale = exports.locale = _core.default.locale.bind(_core.default);
const loadMessages = exports.loadMessages = _message.default.load.bind(_message.default);
const formatMessage = exports.formatMessage = _message.default.format.bind(_message.default);
const formatNumber = exports.formatNumber = _number.default.format.bind(_number.default);
const parseNumber = exports.parseNumber = _number.default.parse.bind(_number.default);
const formatDate = exports.formatDate = _date.default.format.bind(_date.default);
const parseDate = exports.parseDate = _date.default.parse.bind(_date.default);

function disableIntl() {
    if ("intl" === _number.default.engine()) {
        _number.default.resetInjection()
    }
    if ("intl" === _date.default.engine()) {
        _date.default.resetInjection()
    }
}
