// devextreme/esm/__internal/grids/new/grid_core/keyboard_navigation/controller.js
/**
 * DevExtreme (esm/__internal/grids/new/grid_core/keyboard_navigation/controller.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    getPublicElement
} from "../../../../../core/element";
import $ from "../../../../../core/renderer";
import {
    OptionsController
} from "../../../../grids/new/grid_core/options_controller/options_controller";
export class KeyboardNavigationController {
    constructor(options) {
        this.options = options;
        this.enabled = this.options.oneWay("keyboardNavigation.enabled")
    }
    setReturnFocusTo(element) {
        this.returnFocusTo = element
    }
    setFirstCardElement(element) {
        this.firstCardElement = element
    }
    returnFocus() {
        if (!this.returnFocusTo) {
            return
        }
        if (this.returnFocusTo.isConnected) {
            this.returnFocusTo.focus()
        } else {
            var _this$firstCardElemen;
            null === (_this$firstCardElemen = this.firstCardElement) || void 0 === _this$firstCardElemen || _this$firstCardElemen.focus()
        }
        this.returnFocusTo = void 0
    }
    onKeyDown(event) {
        const action = this.options.action("onKeyDown").peek();
        action({
            handled: event.dxHandled ?? false,
            event: event,
            element: getPublicElement($(event.target))
        })
    }
    onFocusedCardChanged(card, cardIdx, element) {
        const action = this.options.action("onFocusedCardChanged").peek();
        action({
            cardIndex: cardIdx,
            card: card,
            cardElement: getPublicElement($(element))
        })
    }
}
KeyboardNavigationController.dependencies = [OptionsController];
