// devextreme/cjs/__internal/grids/grid_core/ai_prompt_editor/ai_prompt_editor.js
/**
 * DevExtreme (cjs/__internal/grids/grid_core/ai_prompt_editor/ai_prompt_editor.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.AIPromptEditor = void 0;
var _message = _interopRequireDefault(require("../../../../common/core/localization/message"));
var _renderer = _interopRequireDefault(require("../../../../core/renderer"));
var _popup = _interopRequireDefault(require("../../../../ui/popup"));
var _progress_bar = _interopRequireDefault(require("../../../../ui/progress_bar"));
var _text_area = _interopRequireDefault(require("../../../../ui/text_area"));
var _const = require("./const");
var _utils = require("./utils");

function _interopRequireDefault(e) {
    return e && e.__esModule ? e : {
        default: e
    }
}
class AIPromptEditor {
    constructor(options) {
        this.options = options;
        this.currentAction = null;
        const {
            container: container,
            createComponent: createComponent
        } = options;
        container.addClass(_const.CLASSES.aiPromptEditor);
        this.prompt = (0, _utils.getPrompt)(options.prompt);
        this.popupInstance = createComponent(container, _popup.default, this.getPopupConfig())
    }
    updateButtonOption(buttonIndex, optionName, optionValue) {
        this.popupInstance.option(`toolbarItems[${buttonIndex}].options.${optionName}`, optionValue)
    }
    updateToolbarItemVisibility(itemIndex, visible) {
        this.popupInstance.option(`toolbarItems[${itemIndex}].visible`, visible)
    }
    getTextAreaConfig() {
        return Object.assign({
            value: this.prompt,
            height: 110,
            stylingMode: "outlined",
            onValueChanged: e => {
                this.updateButtonOption(_const.APPLY_BUTTON_INDEX, "disabled", !e.value || e.value === this.prompt);
                this.updateButtonOption(_const.REGENERATE_DATA_BUTTON_INDEX, "disabled", !e.value || e.value !== this.prompt)
            },
            placeholder: _message.default.format("dxDataGrid-aiPromptEditorPlaceholder"),
            valueChangeEvent: "input change keyup"
        }, this.options.editorOptions)
    }
    getPopupConfig() {
        return Object.assign({}, _const.DEFAULT_POPUP_OPTIONS, {
            shading: false,
            shadingColor: "transparent",
            dragEnabled: true,
            hideOnOutsideClick: true,
            showCloseButton: true,
            title: _message.default.format("dxDataGrid-aiPromptEditorTitle"),
            wrapperAttr: {
                class: `${_const.CLASSES.aiPromptEditor} ${_const.CLASSES.aiDialog}`
            },
            contentTemplate: $container => {
                const $editorContainer = (0, _renderer.default)("<div>").addClass(_const.CLASSES.aiPromptEditorTextArea).appendTo($container);
                const $progressContainer = (0, _renderer.default)("<div>").addClass(_const.CLASSES.aiPromptEditorProgressBar).appendTo($container);
                this.editorInstance = this.options.createComponent($editorContainer, _text_area.default, this.getTextAreaConfig());
                this.progressBar = this.options.createComponent($progressContainer, _progress_bar.default, {
                    value: false,
                    visible: false,
                    showStatus: false,
                    width: "100%"
                })
            },
            toolbarItems: [{
                toolbar: "bottom",
                location: "before",
                widget: "dxButton",
                options: this.getRegenerateDataButtonConfig()
            }, {
                toolbar: "bottom",
                location: "after",
                widget: "dxButton",
                options: this.getApplyButtonConfig()
            }, {
                toolbar: "bottom",
                location: "after",
                widget: "dxButton",
                visible: false,
                options: this.getStopButtonConfig()
            }]
        }, this.options.popupOptions)
    }
    getApplyButtonConfig() {
        return {
            type: "default",
            icon: "arrowright",
            stylingMode: "contained",
            text: _message.default.format("dxDataGrid-aiPromptEditorApplyButton"),
            disabled: !this.editorInstance || !(0, _utils.isPromptChanged)(this.prompt, this.editorInstance.option("value")),
            elementAttr: {
                class: _const.CLASSES.aiPromptEditorApplyButton
            },
            onClick: this.options.onSubmit
        }
    }
    getRegenerateDataButtonConfig() {
        return {
            icon: "refresh",
            stylingMode: "outlined",
            text: _message.default.format("dxDataGrid-aiPromptEditorRegenerateButton"),
            disabled: !this.prompt,
            elementAttr: {
                class: _const.CLASSES.aiPromptEditorRefreshButton
            },
            onClick: this.options.onRefresh
        }
    }
    getStopButtonConfig() {
        return {
            type: "default",
            icon: "square",
            stylingMode: "contained",
            text: _message.default.format("dxDataGrid-aiPromptEditorStopButton"),
            elementAttr: {
                class: _const.CLASSES.aiPromptEditorStopButton
            },
            onClick: this.options.onStop
        }
    }
    setPrompt(prompt) {
        this.prompt = (0, _utils.getPrompt)(prompt)
    }
    toggleDisableState(disabled) {
        const editorValue = this.getEditorValue();
        this.updateButtonOption(_const.REGENERATE_DATA_BUTTON_INDEX, "disabled", disabled ? true : (0, _utils.isPromptChanged)(this.prompt, editorValue));
        this.updateButtonOption(_const.APPLY_BUTTON_INDEX, "disabled", disabled ? true : !(0, _utils.isPromptChanged)(this.prompt, editorValue));
        this.editorInstance.option("disabled", disabled);
        this.popupInstance.option("shading", disabled);
        this.popupInstance.option("hideOnOutsideClick", !disabled)
    }
    getEditorValue() {
        return this.editorInstance.option("value")
    }
    show() {
        return this.popupInstance.show()
    }
    hide() {
        return this.popupInstance.hide()
    }
    isVisible() {
        return true === this.popupInstance.option("visible")
    }
    toggleApplyButtonVisibility(visible) {
        this.updateToolbarItemVisibility(_const.APPLY_BUTTON_INDEX, visible);
        this.updateToolbarItemVisibility(_const.STOP_BUTTON_INDEX, !visible)
    }
    setLoading(isLoading) {
        this.progressBar.option("visible", isLoading)
    }
    updatePrompt(prompt) {
        this.setPrompt(prompt);
        this.editorInstance.option("value", prompt)
    }
    updateStateOnAction(action) {
        this.currentAction = action;
        switch (action) {
            case "apply":
            case "regenerate":
                this.setLoading(true);
                this.toggleDisableState(true);
                this.toggleApplyButtonVisibility(false);
                break;
            case "stop":
                this.setLoading(false);
                this.toggleDisableState(false);
                this.toggleApplyButtonVisibility(true)
        }
    }
    updateOptions(options) {
        this.options = options;
        this.updatePrompt((0, _utils.getPrompt)(options.prompt));
        this.popupInstance.option(this.getPopupConfig())
    }
    isApplying() {
        return "apply" === this.currentAction
    }
}
exports.AIPromptEditor = AIPromptEditor;
