// devextreme/esm/__internal/core/localization/default_date_names.js
/**
 * DevExtreme (esm/__internal/core/localization/default_date_names.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    map
} from "../../core/utils/m_iterator";
const MONTHS = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
const DAYS = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];
const PERIODS = ["AM", "PM"];
const QUARTERS = ["Q1", "Q2", "Q3", "Q4"];
const cutCaptions = (captions, format) => {
    const lengthByFormat = {
        abbreviated: 3,
        short: 2,
        narrow: 1
    };
    return map(captions, (caption => caption.substr(0, lengthByFormat[format])))
};
export default {
    getMonthNames: format => cutCaptions(MONTHS, format),
    getDayNames: format => cutCaptions(DAYS, format),
    getQuarterNames: _format => QUARTERS,
    getPeriodNames: _format => PERIODS
};
