// devextreme/esm/__internal/core/localization/message.js
/**
 * DevExtreme (esm/__internal/core/localization/message.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import coreLocalization from "../../core/localization/core";
import {
    defaultMessages
} from "../../core/localization/default_messages";
import {
    injector as dependencyInjector
} from "../../core/utils/m_dependency_injector";
import {
    extend
} from "../../core/utils/m_extend";
import {
    humanize
} from "../../core/utils/m_inflector";
import {
    format as stringFormat
} from "../../core/utils/m_string";
const baseDictionary = extend(true, {}, defaultMessages);
const getDataByLocale = (localeData, locale) => {
    var _Object$entries$find;
    return localeData[locale] || (null === locale || void 0 === locale ? void 0 : locale.toLowerCase) && (null === (_Object$entries$find = Object.entries(localeData).find((_ref => {
        let [key] = _ref;
        return key.toLowerCase() === locale.toLowerCase()
    }))) || void 0 === _Object$entries$find ? void 0 : _Object$entries$find[1]) || {}
};
const newMessages = {};
const messageLocalization = dependencyInjector({
    engine: () => "base",
    _dictionary: baseDictionary,
    load(messages) {
        extend(true, this._dictionary, messages)
    },
    _localizablePrefix: "@",
    setup(localizablePrefix) {
        this._localizablePrefix = localizablePrefix
    },
    localizeString(text) {
        const that = this;
        const regex = new RegExp(`(^|[^a-zA-Z_0-9${that._localizablePrefix}-]+)(${that._localizablePrefix}{1,2})([a-zA-Z_0-9-]+)`, "g");
        const escapeString = that._localizablePrefix + that._localizablePrefix;
        return text.replace(regex, ((_, prefix, escape, localizationKey) => {
            const defaultResult = that._localizablePrefix + localizationKey;
            let result;
            if (escape !== escapeString) {
                result = that.format(localizationKey)
            }
            if (!result) {
                newMessages[localizationKey] = humanize(localizationKey)
            }
            return prefix + (result || defaultResult)
        }))
    },
    getMessagesByLocales() {
        return this._dictionary
    },
    getDictionary(onlyNew) {
        if (onlyNew) {
            return newMessages
        }
        return extend({}, newMessages, this.getMessagesByLocales()[coreLocalization.locale()])
    },
    getFormatter(key) {
        return this._getFormatterBase(key) || this._getFormatterBase(key, "en")
    },
    _getFormatterBase(key, _locale) {
        const message = coreLocalization.getValueByClosestLocale((locale => getDataByLocale(this._dictionary, locale)[key]));
        if (message) {
            return function() {
                const args = 1 === arguments.length && Array.isArray(arguments[0]) ? arguments[0].slice(0) : Array.prototype.slice.call(arguments, 0);
                args.unshift(message);
                return stringFormat.apply(this, args)
            }
        }
        return
    },
    format(key) {
        const formatter = this.getFormatter(key);
        const values = Array.prototype.slice.call(arguments, 1);
        return (null === formatter || void 0 === formatter ? void 0 : formatter.apply(this, values)) || ""
    }
});
export default messageLocalization;
