// devextreme/cjs/core/utils/ajax_utils.js
/**
 * DevExtreme (cjs/core/utils/ajax_utils.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "evalCrossDomainScript", {
    enumerable: true,
    get: function() {
        return _m_ajax_utils.evalCrossDomainScript
    }
});
Object.defineProperty(exports, "evalScript", {
    enumerable: true,
    get: function() {
        return _m_ajax_utils.evalScript
    }
});
Object.defineProperty(exports, "getAcceptHeader", {
    enumerable: true,
    get: function() {
        return _m_ajax_utils.getAcceptHeader
    }
});
Object.defineProperty(exports, "getJsonpCallbackName", {
    enumerable: true,
    get: function() {
        return _m_ajax_utils.getJsonpCallbackName
    }
});
Object.defineProperty(exports, "getMethod", {
    enumerable: true,
    get: function() {
        return _m_ajax_utils.getMethod
    }
});
Object.defineProperty(exports, "getRequestHeaders", {
    enumerable: true,
    get: function() {
        return _m_ajax_utils.getRequestHeaders
    }
});
Object.defineProperty(exports, "getRequestOptions", {
    enumerable: true,
    get: function() {
        return _m_ajax_utils.getRequestOptions
    }
});
Object.defineProperty(exports, "isCrossDomain", {
    enumerable: true,
    get: function() {
        return _m_ajax_utils.isCrossDomain
    }
});
var _m_ajax_utils = require("../../__internal/core/utils/m_ajax_utils");
