// devextreme/cjs/core/utils/array.js
/**
 * DevExtreme (cjs/core/utils/array.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "getIntersection", {
    enumerable: true,
    get: function() {
        return _m_array.getIntersection
    }
});
Object.defineProperty(exports, "getUniqueValues", {
    enumerable: true,
    get: function() {
        return _m_array.getUniqueValues
    }
});
Object.defineProperty(exports, "groupBy", {
    enumerable: true,
    get: function() {
        return _m_array.groupBy
    }
});
Object.defineProperty(exports, "normalizeIndexes", {
    enumerable: true,
    get: function() {
        return _m_array.normalizeIndexes
    }
});
Object.defineProperty(exports, "removeDuplicates", {
    enumerable: true,
    get: function() {
        return _m_array.removeDuplicates
    }
});
Object.defineProperty(exports, "wrapToArray", {
    enumerable: true,
    get: function() {
        return _m_array.wrapToArray
    }
});
var _m_array = require("../../__internal/core/utils/m_array");
