// devextreme/cjs/core/utils/common.js
/**
 * DevExtreme (cjs/core/utils/common.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "applyServerDecimalSeparator", {
    enumerable: true,
    get: function() {
        return _m_common.applyServerDecimalSeparator
    }
});
Object.defineProperty(exports, "asyncNoop", {
    enumerable: true,
    get: function() {
        return _m_common.asyncNoop
    }
});
Object.defineProperty(exports, "deferRender", {
    enumerable: true,
    get: function() {
        return _m_common.deferRender
    }
});
Object.defineProperty(exports, "deferRenderer", {
    enumerable: true,
    get: function() {
        return _m_common.deferRenderer
    }
});
Object.defineProperty(exports, "deferUpdate", {
    enumerable: true,
    get: function() {
        return _m_common.deferUpdate
    }
});
Object.defineProperty(exports, "deferUpdater", {
    enumerable: true,
    get: function() {
        return _m_common.deferUpdater
    }
});
Object.defineProperty(exports, "denormalizeKey", {
    enumerable: true,
    get: function() {
        return _m_common.denormalizeKey
    }
});
Object.defineProperty(exports, "ensureDefined", {
    enumerable: true,
    get: function() {
        return _m_common.ensureDefined
    }
});
Object.defineProperty(exports, "equalByValue", {
    enumerable: true,
    get: function() {
        return _m_common.equalByValue
    }
});
Object.defineProperty(exports, "escapeRegExp", {
    enumerable: true,
    get: function() {
        return _m_common.escapeRegExp
    }
});
Object.defineProperty(exports, "executeAsync", {
    enumerable: true,
    get: function() {
        return _m_common.executeAsync
    }
});
Object.defineProperty(exports, "findBestMatches", {
    enumerable: true,
    get: function() {
        return _m_common.findBestMatches
    }
});
Object.defineProperty(exports, "getKeyHash", {
    enumerable: true,
    get: function() {
        return _m_common.getKeyHash
    }
});
Object.defineProperty(exports, "grep", {
    enumerable: true,
    get: function() {
        return _m_common.grep
    }
});
Object.defineProperty(exports, "noop", {
    enumerable: true,
    get: function() {
        return _m_common.noop
    }
});
Object.defineProperty(exports, "normalizeKey", {
    enumerable: true,
    get: function() {
        return _m_common.normalizeKey
    }
});
Object.defineProperty(exports, "pairToObject", {
    enumerable: true,
    get: function() {
        return _m_common.pairToObject
    }
});
Object.defineProperty(exports, "splitPair", {
    enumerable: true,
    get: function() {
        return _m_common.splitPair
    }
});
var _m_common = require("../../__internal/core/utils/m_common");
