// devextreme/cjs/core/utils/dom.js
/**
 * DevExtreme (cjs/core/utils/dom.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "clearSelection", {
    enumerable: true,
    get: function() {
        return _m_dom.clearSelection
    }
});
Object.defineProperty(exports, "clipboardText", {
    enumerable: true,
    get: function() {
        return _m_dom.clipboardText
    }
});
Object.defineProperty(exports, "closestCommonParent", {
    enumerable: true,
    get: function() {
        return _m_dom.closestCommonParent
    }
});
Object.defineProperty(exports, "contains", {
    enumerable: true,
    get: function() {
        return _m_dom.contains
    }
});
Object.defineProperty(exports, "createTextElementHiddenCopy", {
    enumerable: true,
    get: function() {
        return _m_dom.createTextElementHiddenCopy
    }
});
Object.defineProperty(exports, "extractTemplateMarkup", {
    enumerable: true,
    get: function() {
        return _m_dom.extractTemplateMarkup
    }
});
Object.defineProperty(exports, "insertBefore", {
    enumerable: true,
    get: function() {
        return _m_dom.insertBefore
    }
});
Object.defineProperty(exports, "isElementInDom", {
    enumerable: true,
    get: function() {
        return _m_dom.isElementInDom
    }
});
Object.defineProperty(exports, "normalizeTemplateElement", {
    enumerable: true,
    get: function() {
        return _m_dom.normalizeTemplateElement
    }
});
Object.defineProperty(exports, "replaceWith", {
    enumerable: true,
    get: function() {
        return _m_dom.replaceWith
    }
});
Object.defineProperty(exports, "resetActiveElement", {
    enumerable: true,
    get: function() {
        return _m_dom.resetActiveElement
    }
});
var _m_dom = require("../../__internal/core/utils/m_dom");
