// devextreme/cjs/core/utils/inflector.js
/**
 * DevExtreme (cjs/core/utils/inflector.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "camelize", {
    enumerable: true,
    get: function() {
        return _m_inflector.camelize
    }
});
Object.defineProperty(exports, "captionize", {
    enumerable: true,
    get: function() {
        return _m_inflector.captionize
    }
});
Object.defineProperty(exports, "dasherize", {
    enumerable: true,
    get: function() {
        return _m_inflector.dasherize
    }
});
Object.defineProperty(exports, "humanize", {
    enumerable: true,
    get: function() {
        return _m_inflector.humanize
    }
});
Object.defineProperty(exports, "titleize", {
    enumerable: true,
    get: function() {
        return _m_inflector.titleize
    }
});
Object.defineProperty(exports, "underscore", {
    enumerable: true,
    get: function() {
        return _m_inflector.underscore
    }
});
var _m_inflector = require("../../__internal/core/utils/m_inflector");
