// devextreme/cjs/core/utils/math.js
/**
 * DevExtreme (cjs/core/utils/math.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "adjust", {
    enumerable: true,
    get: function() {
        return _m_math.adjust
    }
});
Object.defineProperty(exports, "fitIntoRange", {
    enumerable: true,
    get: function() {
        return _m_math.fitIntoRange
    }
});
Object.defineProperty(exports, "getExponent", {
    enumerable: true,
    get: function() {
        return _m_math.getExponent
    }
});
Object.defineProperty(exports, "getExponentLength", {
    enumerable: true,
    get: function() {
        return _m_math.getExponentLength
    }
});
Object.defineProperty(exports, "getPrecision", {
    enumerable: true,
    get: function() {
        return _m_math.getPrecision
    }
});
Object.defineProperty(exports, "getRemainderByDivision", {
    enumerable: true,
    get: function() {
        return _m_math.getRemainderByDivision
    }
});
Object.defineProperty(exports, "getRoot", {
    enumerable: true,
    get: function() {
        return _m_math.getRoot
    }
});
Object.defineProperty(exports, "inRange", {
    enumerable: true,
    get: function() {
        return _m_math.inRange
    }
});
Object.defineProperty(exports, "multiplyInExponentialForm", {
    enumerable: true,
    get: function() {
        return _m_math.multiplyInExponentialForm
    }
});
Object.defineProperty(exports, "roundFloatPart", {
    enumerable: true,
    get: function() {
        return _m_math.roundFloatPart
    }
});
Object.defineProperty(exports, "sign", {
    enumerable: true,
    get: function() {
        return _m_math.sign
    }
});
Object.defineProperty(exports, "solveCubicEquation", {
    enumerable: true,
    get: function() {
        return _m_math.solveCubicEquation
    }
});
Object.defineProperty(exports, "trunc", {
    enumerable: true,
    get: function() {
        return _m_math.trunc
    }
});
var _m_math = require("../../__internal/core/utils/m_math");
