// devextreme/cjs/core/utils/size.js
/**
 * DevExtreme (cjs/core/utils/size.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "addOffsetToMaxHeight", {
    enumerable: true,
    get: function() {
        return _m_size.addOffsetToMaxHeight
    }
});
Object.defineProperty(exports, "addOffsetToMinHeight", {
    enumerable: true,
    get: function() {
        return _m_size.addOffsetToMinHeight
    }
});
Object.defineProperty(exports, "getElementBoxParams", {
    enumerable: true,
    get: function() {
        return _m_size.getElementBoxParams
    }
});
Object.defineProperty(exports, "getHeight", {
    enumerable: true,
    get: function() {
        return _m_size.getHeight
    }
});
Object.defineProperty(exports, "getInnerHeight", {
    enumerable: true,
    get: function() {
        return _m_size.getInnerHeight
    }
});
Object.defineProperty(exports, "getInnerWidth", {
    enumerable: true,
    get: function() {
        return _m_size.getInnerWidth
    }
});
Object.defineProperty(exports, "getOffset", {
    enumerable: true,
    get: function() {
        return _m_size.getOffset
    }
});
Object.defineProperty(exports, "getOuterHeight", {
    enumerable: true,
    get: function() {
        return _m_size.getOuterHeight
    }
});
Object.defineProperty(exports, "getOuterWidth", {
    enumerable: true,
    get: function() {
        return _m_size.getOuterWidth
    }
});
Object.defineProperty(exports, "getSize", {
    enumerable: true,
    get: function() {
        return _m_size.getSize
    }
});
Object.defineProperty(exports, "getVerticalOffsets", {
    enumerable: true,
    get: function() {
        return _m_size.getVerticalOffsets
    }
});
Object.defineProperty(exports, "getVisibleHeight", {
    enumerable: true,
    get: function() {
        return _m_size.getVisibleHeight
    }
});
Object.defineProperty(exports, "getWidth", {
    enumerable: true,
    get: function() {
        return _m_size.getWidth
    }
});
Object.defineProperty(exports, "getWindowByElement", {
    enumerable: true,
    get: function() {
        return _m_size.getWindowByElement
    }
});
Object.defineProperty(exports, "implementationsMap", {
    enumerable: true,
    get: function() {
        return _m_size.implementationsMap
    }
});
Object.defineProperty(exports, "parseHeight", {
    enumerable: true,
    get: function() {
        return _m_size.parseHeight
    }
});
Object.defineProperty(exports, "setHeight", {
    enumerable: true,
    get: function() {
        return _m_size.setHeight
    }
});
Object.defineProperty(exports, "setInnerHeight", {
    enumerable: true,
    get: function() {
        return _m_size.setInnerHeight
    }
});
Object.defineProperty(exports, "setInnerWidth", {
    enumerable: true,
    get: function() {
        return _m_size.setInnerWidth
    }
});
Object.defineProperty(exports, "setOuterHeight", {
    enumerable: true,
    get: function() {
        return _m_size.setOuterHeight
    }
});
Object.defineProperty(exports, "setOuterWidth", {
    enumerable: true,
    get: function() {
        return _m_size.setOuterWidth
    }
});
Object.defineProperty(exports, "setWidth", {
    enumerable: true,
    get: function() {
        return _m_size.setWidth
    }
});
var _m_size = require("../../__internal/core/utils/m_size");
