// devextreme/cjs/core/utils/style.js
/**
 * DevExtreme (cjs/core/utils/style.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "normalizeStyleProp", {
    enumerable: true,
    get: function() {
        return _m_style.normalizeStyleProp
    }
});
Object.defineProperty(exports, "parsePixelValue", {
    enumerable: true,
    get: function() {
        return _m_style.parsePixelValue
    }
});
Object.defineProperty(exports, "setHeight", {
    enumerable: true,
    get: function() {
        return _m_style.setHeight
    }
});
Object.defineProperty(exports, "setStyle", {
    enumerable: true,
    get: function() {
        return _m_style.setStyle
    }
});
Object.defineProperty(exports, "setWidth", {
    enumerable: true,
    get: function() {
        return _m_style.setWidth
    }
});
Object.defineProperty(exports, "styleProp", {
    enumerable: true,
    get: function() {
        return _m_style.styleProp
    }
});
Object.defineProperty(exports, "stylePropPrefix", {
    enumerable: true,
    get: function() {
        return _m_style.stylePropPrefix
    }
});
var _m_style = require("../../__internal/core/utils/m_style");
