// devextreme/cjs/core/utils/support.js
/**
 * DevExtreme (cjs/core/utils/support.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "animation", {
    enumerable: true,
    get: function() {
        return _m_support.animation
    }
});
Object.defineProperty(exports, "inputType", {
    enumerable: true,
    get: function() {
        return _m_support.inputType
    }
});
Object.defineProperty(exports, "nativeScrolling", {
    enumerable: true,
    get: function() {
        return _m_support.nativeScrolling
    }
});
Object.defineProperty(exports, "pointerEvents", {
    enumerable: true,
    get: function() {
        return _m_support.pointerEvents
    }
});
Object.defineProperty(exports, "styleProp", {
    enumerable: true,
    get: function() {
        return _m_support.styleProp
    }
});
Object.defineProperty(exports, "stylePropPrefix", {
    enumerable: true,
    get: function() {
        return _m_support.stylePropPrefix
    }
});
Object.defineProperty(exports, "supportProp", {
    enumerable: true,
    get: function() {
        return _m_support.supportProp
    }
});
Object.defineProperty(exports, "touch", {
    enumerable: true,
    get: function() {
        return _m_support.touch
    }
});
Object.defineProperty(exports, "touchEvents", {
    enumerable: true,
    get: function() {
        return _m_support.touchEvents
    }
});
Object.defineProperty(exports, "transition", {
    enumerable: true,
    get: function() {
        return _m_support.transition
    }
});
Object.defineProperty(exports, "transitionEndEventName", {
    enumerable: true,
    get: function() {
        return _m_support.transitionEndEventName
    }
});
var _m_support = require("../../__internal/core/utils/m_support");
