// devextreme/cjs/core/utils/template_manager.js
/**
 * DevExtreme (cjs/core/utils/template_manager.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "acquireIntegrationTemplate", {
    enumerable: true,
    get: function() {
        return _m_template_manager.acquireIntegrationTemplate
    }
});
Object.defineProperty(exports, "acquireTemplate", {
    enumerable: true,
    get: function() {
        return _m_template_manager.acquireTemplate
    }
});
Object.defineProperty(exports, "addOneRenderedCall", {
    enumerable: true,
    get: function() {
        return _m_template_manager.addOneRenderedCall
    }
});
Object.defineProperty(exports, "addPublicElementNormalization", {
    enumerable: true,
    get: function() {
        return _m_template_manager.addPublicElementNormalization
    }
});
Object.defineProperty(exports, "defaultCreateElement", {
    enumerable: true,
    get: function() {
        return _m_template_manager.defaultCreateElement
    }
});
Object.defineProperty(exports, "findTemplates", {
    enumerable: true,
    get: function() {
        return _m_template_manager.findTemplates
    }
});
Object.defineProperty(exports, "getNormalizedTemplateArgs", {
    enumerable: true,
    get: function() {
        return _m_template_manager.getNormalizedTemplateArgs
    }
});
Object.defineProperty(exports, "suitableTemplatesByName", {
    enumerable: true,
    get: function() {
        return _m_template_manager.suitableTemplatesByName
    }
});
Object.defineProperty(exports, "templateKey", {
    enumerable: true,
    get: function() {
        return _m_template_manager.templateKey
    }
});
Object.defineProperty(exports, "validateTemplateSource", {
    enumerable: true,
    get: function() {
        return _m_template_manager.validateTemplateSource
    }
});
var _m_template_manager = require("../../__internal/core/utils/m_template_manager");
