// devextreme/cjs/core/utils/type.js
/**
 * DevExtreme (cjs/core/utils/type.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "isBoolean", {
    enumerable: true,
    get: function() {
        return _m_type.isBoolean
    }
});
Object.defineProperty(exports, "isDate", {
    enumerable: true,
    get: function() {
        return _m_type.isDate
    }
});
Object.defineProperty(exports, "isDeferred", {
    enumerable: true,
    get: function() {
        return _m_type.isDeferred
    }
});
Object.defineProperty(exports, "isDefined", {
    enumerable: true,
    get: function() {
        return _m_type.isDefined
    }
});
Object.defineProperty(exports, "isEmptyObject", {
    enumerable: true,
    get: function() {
        return _m_type.isEmptyObject
    }
});
Object.defineProperty(exports, "isEvent", {
    enumerable: true,
    get: function() {
        return _m_type.isEvent
    }
});
Object.defineProperty(exports, "isExponential", {
    enumerable: true,
    get: function() {
        return _m_type.isExponential
    }
});
Object.defineProperty(exports, "isFunction", {
    enumerable: true,
    get: function() {
        return _m_type.isFunction
    }
});
Object.defineProperty(exports, "isNumeric", {
    enumerable: true,
    get: function() {
        return _m_type.isNumeric
    }
});
Object.defineProperty(exports, "isObject", {
    enumerable: true,
    get: function() {
        return _m_type.isObject
    }
});
Object.defineProperty(exports, "isPlainObject", {
    enumerable: true,
    get: function() {
        return _m_type.isPlainObject
    }
});
Object.defineProperty(exports, "isPrimitive", {
    enumerable: true,
    get: function() {
        return _m_type.isPrimitive
    }
});
Object.defineProperty(exports, "isPromise", {
    enumerable: true,
    get: function() {
        return _m_type.isPromise
    }
});
Object.defineProperty(exports, "isRenderer", {
    enumerable: true,
    get: function() {
        return _m_type.isRenderer
    }
});
Object.defineProperty(exports, "isString", {
    enumerable: true,
    get: function() {
        return _m_type.isString
    }
});
Object.defineProperty(exports, "isWindow", {
    enumerable: true,
    get: function() {
        return _m_type.isWindow
    }
});
Object.defineProperty(exports, "type", {
    enumerable: true,
    get: function() {
        return _m_type.type
    }
});
var _m_type = require("../../__internal/core/utils/m_type");
