// devextreme/cjs/core/utils/window.js
/**
 * DevExtreme (cjs/core/utils/window.js)
 * Version: 25.2.4
 * Build date: Mon Jan 26 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "defaultScreenFactorFunc", {
    enumerable: true,
    get: function() {
        return _m_window.defaultScreenFactorFunc
    }
});
Object.defineProperty(exports, "getCurrentScreenFactor", {
    enumerable: true,
    get: function() {
        return _m_window.getCurrentScreenFactor
    }
});
Object.defineProperty(exports, "getNavigator", {
    enumerable: true,
    get: function() {
        return _m_window.getNavigator
    }
});
Object.defineProperty(exports, "getWindow", {
    enumerable: true,
    get: function() {
        return _m_window.getWindow
    }
});
Object.defineProperty(exports, "hasProperty", {
    enumerable: true,
    get: function() {
        return _m_window.hasProperty
    }
});
Object.defineProperty(exports, "hasWindow", {
    enumerable: true,
    get: function() {
        return _m_window.hasWindow
    }
});
Object.defineProperty(exports, "setWindow", {
    enumerable: true,
    get: function() {
        return _m_window.setWindow
    }
});
var _m_window = require("../../__internal/core/utils/m_window");
